################################################################################
# SPECIES

# Species 'aedes_aegypti' (id = 7159)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (7159, 7159, 'aedes_aegypti', 'aaegypti, aedes, aedesaegypti');

# Species 'acyrthosiphon_pisum' (id = 7029)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (7029, 7029, 'acyrthosiphon_pisum', 'apisum, acyrthosiphon, acyrthosiphonpisum, pea aphid');

# Species 'ailuropoda_melanoleuca' (id = 9646)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (9646, 9646, 'ailuropoda_melanoleuca', 'Panda, panda, giant panda, ailuropodamelanoleuca');

# Species 'anas_platyrhynchos' (id = 8839)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (8839, 8839, 'anas_platyrhynchos', 'duck, mallard, mallard duck, mallard ducks, anas platyrhynchos, anas boschas, anas domesticus');

# Species 'anolis_carolinensis' (id = 28377)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (28377, 28377, 'anolis_carolinensis', 'anole, green anole lizard, green anole, american anole, redthroated anole, anoliscarolinensis, anolis_carolinensis');

# Species 'anopheles_darlingi' (id = 43151)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (43151, 43151, 'anopheles_darlingi', 'adarlingi, anophelesdarlingi, American malaria mosquito');

# Species 'anopheles_gambiae' (id = 7165)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (7165, 7165, 'anopheles_gambiae', 'agambiae, anopheles, anophelesgambiae, mosquito');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (7165, 180454, 'anopheles_gambiae', 'agambiae, anopheles, anophelesgambiae, mosquito');

# Species 'apis_mellifera' (id = 7460)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (7460, 7460, 'apis_mellifera', 'amellifera, apismellifera, honeybee');

# Species 'aspergillus_clavatus' (id = 5057)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (5057, 5057, 'aspergillus_clavatus', 'aspergillusclavatus, aclavatus, Aspergillus clavatus, Aspergillus clavatus NRRL 1');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (5057, 344612, 'aspergillus_clavatus', 'aspergillusclavatus, aclavatus, Aspergillus clavatus, Aspergillus clavatus NRRL 1');

# Species 'aspergillus_flavus' (id = 5059)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (5059, 5059, 'aspergillus_flavus', 'aspergillusflavus, aflavus, Aspergillus flavus, Aspergillus flavus NRRL3357');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (5059, 332952, 'aspergillus_flavus', 'aspergillusflavus, aflavus, Aspergillus flavus, Aspergillus flavus NRRL3357');

# Species 'aspergillus_fumigatus' (id = 330879)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (330879, 330879, 'aspergillus_fumigatus', 'aspergillusfumigatus, afumigatus, Aspergillus fumigatus, Aspergillus fumigatus Af293,Sartorya fumigata,sartoryafumigata,sfumigata');

# Species 'afumigatusa1163' (id = 451804)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (451804, 451804, 'afumigatusa1163', 'aspergillusfumigatus, afumigatus, Aspergillus fumigatus, Aspergillus fumigatus A1163,Aspergillus fumigatus CEA10,Sartorya fumigata,sartoryafumigata,sfumigata');

# Species 'aspergillus_nidulans' (id = 162425)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (162425, 162425, 'aspergillus_nidulans', 'aspergillusnidulans, anidulans, Aspergillus nidulellus, aspergillusnidulellus, anidulellus, Emericella nidulans, emericellanidulans, enidulans, Aspergillus nidulans FGSC A4');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (162425, 227321, 'aspergillus_nidulans', 'aspergillusnidulans, anidulans, Aspergillus nidulellus, aspergillusnidulellus, anidulellus, Emericella nidulans, emericellanidulans, enidulans, Aspergillus nidulans FGSC A4');

# Species 'aspergillus_niger' (id = 425011)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (425011, 425011, 'aspergillus_niger', 'aspergillusniger, aniger, Aspergillus niger, Aspergillus niger CBS 513.88');

# Species 'aspergillus_oryzae' (id = 5062)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (5062, 5062, 'aspergillus_oryzae', 'aspergillusoryzae, aoryzae, Aspergillus oryzae, Aspergillus oryzae RIB40');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (5062, 510516, 'aspergillus_oryzae', 'aspergillusoryzae, aoryzae, Aspergillus oryzae, Aspergillus oryzae RIB40');

# Species 'aspergillus_terreus' (id = 33178)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (33178, 33178, 'aspergillus_terreus', 'aspergillusterreus, aterreus, Aspergillus terreus, Aspergillus terreus NIH2624');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (33178, 341663, 'aspergillus_terreus', 'aspergillusterreus, aterreus, Aspergillus terreus, Aspergillus terreus NIH2624');

# Species 'astyanax_mexicanus' (id = 7994)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (7994, 7994, 'astyanax_mexicanus', 'cave fish, Mexican tetra');

# Species 'neosartorya_fischeri' (id = 36630)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (36630, 36630, 'neosartorya_fischeri', 'neosartoryafischeri, nfischeri, Neosartorya fischeri, Neosartorya fischeri NRRL 181');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (36630, 331117, 'neosartorya_fischeri', 'neosartoryafischeri, nfischeri, Neosartorya fischeri, Neosartorya fischeri NRRL 181');

# Species 'puccinia_graministritici' (id = 418459)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (418459, 418459, 'puccinia_graministritici', 'pucciniagraministritici, pgraminis, Puccinia graminis f. sp. tritici, Puccinia graminis f. sp. tritici CRL 75-36-700-3');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (418459, 56615, 'puccinia_graministritici', 'pucciniagraministritici, pgraminis, Puccinia graminis f. sp. tritici, Puccinia graminis f. sp. tritici CRL 75-36-700-3');

# Species 'mycosphaerella_graminicola' (id = 336722)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (336722, 336722, 'mycosphaerella_graminicola', 'mycosphaerellagraminicola, mgraminicola, Mycosphaerella graminicola, Mycosphaerella graminicola IPO323');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (336722, 54734, 'mycosphaerella_graminicola', 'mycosphaerellagraminicola, mgraminicola, Mycosphaerella graminicola, Mycosphaerella graminicola IPO323');

# Species 'trichoplax_adhaerens' (id = 10228)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (10228, 10228, 'trichoplax_adhaerens', 'trichoplaxadhaerens, tadhaerens, Trichoplax adhaerens, Trichoplax adhaerens Grell-BS-1999');

# Species 'schistosoma_mansoni' (id = 6183)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (6183, 6183, 'schistosoma_mansoni', 'schistosomamansoni, smansoni, Schistosoma mansoni, Schistosoma mansoni Puerto Rico');

# Species 'leishmania_major' (id = 5664)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (5664, 5664, 'leishmania_major', 'leishmaniamajor, Leishmania major, Leishmania major Freidlin, Leishmania major strain Freidlin, lmajor');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (5664, 347515, 'leishmania_major', 'leishmaniamajor, Leishmania major, Leishmania major Freidlin, Leishmania major strain Freidlin, lmajor');

# Species 'pythium_ultimum' (id = 65071)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (65071, 65071, 'pythium_ultimum', 'pythium_ultimum');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (65071, 431595, 'pythium_ultimum', 'pythium_ultimum');

# Species 'gibberella_moniliformis' (id = 117187)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (117187, 117187, 'gibberella_moniliformis', 'gibberella_moniliformis');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (117187, 334819, 'gibberella_moniliformis', 'gibberella_moniliformis');

# Species 'hyaloperonospora_arabidopsidis' (id = 559515)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (559515, 559515, 'hyaloperonospora_arabidopsidis', 'Peronospora arabidopsidis , Hyaloperonospora arabidopsidis , Hyaloperonospora arabidopsidis Emoy2 ,Hyaloperonospora arabidopsidis East Malling Oystese 2, Hyaloperonospora parasitica');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (559515, 272952, 'hyaloperonospora_arabidopsidis', 'Peronospora arabidopsidis , Hyaloperonospora arabidopsidis , Hyaloperonospora arabidopsidis Emoy2 ,Hyaloperonospora arabidopsidis East Malling Oystese 2, Hyaloperonospora parasitica');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (559515, 123356, 'hyaloperonospora_arabidopsidis', 'Peronospora arabidopsidis , Hyaloperonospora arabidopsidis , Hyaloperonospora arabidopsidis Emoy2 ,Hyaloperonospora arabidopsidis East Malling Oystese 2, Hyaloperonospora parasitica');

# Species 'neurospora_crassa' (id = 5141)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (5141, 5141, 'neurospora_crassa', 'neurosporacrassa, ncrassa, Neurospora crassa, Neurospora crassa OR74A');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (5141, 367110, 'neurospora_crassa', 'neurosporacrassa, ncrassa, Neurospora crassa, Neurospora crassa OR74A');

# Species 'ustilago_maydis' (id = 237631)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (237631, 237631, 'ustilago_maydis', 'ustilagomaydis, umaydis, Ustilago maydis, Ustilago maydis 521');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (237631, 5270, 'ustilago_maydis', 'ustilagomaydis, umaydis, Ustilago maydis, Ustilago maydis 521');

# Species 'puccinia_triticina' (id = 208348)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (208348, 208348, 'puccinia_triticina', 'pucciniatriticina, ptriticina, Puccinia triticina, Puccinia triticina 1-1 BBBD Race 1');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (208348, 630390, 'puccinia_triticina', 'pucciniatriticina, ptriticina, Puccinia triticina, Puccinia triticina 1-1 BBBD Race 1');

# Species 'bos_taurus' (id = 9913)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (9913, 9913, 'bos_taurus', 'bostaurus, btaurus, cow');

# Species 'callithrix_jacchus' (id = 9483)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (9483, 9483, 'callithrix_jacchus', 'Marmoset, marmoset, white-tufted-ear marmoset, callithrixjacchus, callithrix_jacchus');

# Species 'canis_familiaris' (id = 9615)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (9615, 9615, 'canis_familiaris', 'canisfamiliaris, cfamiliaris, dog, doggy');

# Species 'cavia_porcellus' (id = 10141)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (10141, 10141, 'cavia_porcellus', 'cavia porcellus,guinea_pig,guinea pig');

# Species 'chlorocebus_sabaeus' (id = 60711)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (60711, 60711, 'chlorocebus_sabaeus', 'Chlorocebus sabeus,vervet monkey,African Green Monkey,AGM,vervet,chlorocebus_sabeus,Chlorocebus sabaeus');

# Species 'choloepus_hoffmanni' (id = 9358)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (9358, 9358, 'choloepus_hoffmanni', 'sloth');

# Species 'ciona_intestinalis' (id = 7719)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (7719, 7719, 'ciona_intestinalis', 'cintestinalis, cionaintestinalis, seasquirt');

# Species 'ciona_savignyi' (id = 51511)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (51511, 51511, 'ciona_savignyi', 'ciona_savi, csav, savi');

# Species 'culex_quinquefasciatus' (id = 7176)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (7176, 7176, 'culex_quinquefasciatus', 'cpipiens, culex, culexpipiens, culex quinquefasciatus, cquinquefasciatus, culexquinquefasciatus');

# Species 'danio_rerio' (id = 7955)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (7955, 7955, 'danio_rerio', 'danio, daniorerio, drerio, zebrafish');

# Species 'dasypus_novemcinctus' (id = 9361)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (9361, 9361, 'dasypus_novemcinctus', 'armadillo, dasypusnovemcinctus, dnovemcinctus');

# Species 'dictyostelium_discoideum' (id = 366501)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (366501, 366501, 'dictyostelium_discoideum', 'dictyostelium, dictyostelium_discoideum,Slime_mold');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (366501, 44689, 'dictyostelium_discoideum', 'dictyostelium, dictyostelium_discoideum,Slime_mold');

# Species 'dipodomys_ordii' (id = 10020)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (10020, 10020, 'dipodomys_ordii', 'kangaroorat, roorat');

# Species 'drosophila_melanogaster' (id = 7227)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (7227, 7227, 'drosophila_melanogaster', 'dmelanogaster, drosophila, drosophilamelanogaster');

# Species 'drosophila_pseudoobscura' (id = 46245)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (46245, 46245, 'drosophila_pseudoobscura', 'dpseudoobscura, drosophila, drosophilapseudoobscura');

# Species 'drosophila_ananassae' (id = 7217)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (7217, 7217, 'drosophila_ananassae', 'dananassae, drosophila, drosophilaananassae');

# Species 'drosophila_erecta' (id = 7220)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (7220, 7220, 'drosophila_erecta', 'derecta, drosophila, drosophilaerecta');

# Species 'drosophila_grimshawi' (id = 7222)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (7222, 7222, 'drosophila_grimshawi', 'dgrimshawi, drosophila, drosophilagrimshawi');

# Species 'drosophila_mojavensis' (id = 7230)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (7230, 7230, 'drosophila_mojavensis', 'dmojavensis, drosophila, drosophilamojavensis');

# Species 'drosophila_persimilis' (id = 7234)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (7234, 7234, 'drosophila_persimilis', 'dpersimilis, drosophila, drosophilapersimilis');

# Species 'drosophila_sechellia' (id = 7238)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (7238, 7238, 'drosophila_sechellia', 'dsechellia, drosophila, drosophilasechellia');

# Species 'drosophila_simulans' (id = 7240)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (7240, 7240, 'drosophila_simulans', 'dsimulans, drosophila, drosophilasimulans');

# Species 'drosophila_virilis' (id = 7244)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (7244, 7244, 'drosophila_virilis', 'dvirilis, drosophila, drosophilavirilis');

# Species 'drosophila_willistoni' (id = 7260)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (7260, 7260, 'drosophila_willistoni', 'dwillistoni, drosophila, drosophilawillistoni');

# Species 'drosophila_yakuba' (id = 7245)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (7245, 7245, 'drosophila_yakuba', 'dyakuba, drosophila, drosophilayakuba');

# Species 'daphnia_pulex' (id = 6669)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (6669, 6669, 'daphnia_pulex', 'daphniapulex, dpulex, common water flea');

# Species 'echinops_telfairi' (id = 9371)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (9371, 9371, 'echinops_telfairi', 'echinopstelfairi, etelfairi, tenrec');

# Species 'erinaceus_europaeus' (id = 9365)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (9365, 9365, 'erinaceus_europaeus', 'hedgehog');

# Species 'equus_caballus' (id = 9796)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (9796, 9796, 'equus_caballus', 'horse');

# Species 'felis_catus' (id = 9685)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (9685, 9685, 'felis_catus', 'cat, fcatus, feliscatus, moggy, tiddles');

# Species 'ficedula_albicollis' (id = 59894)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (59894, 59894, 'ficedula_albicollis', 'flycatcher, falbicollis, falb');

# Species 'gadus_morhua' (id = 8049)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (8049, 8049, 'gadus_morhua', 'cod,gmorhua,gadus_morhua,gmor,atlantic_cod,gadmor');

# Species 'gallus_gallus' (id = 9031)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (9031, 9031, 'gallus_gallus', 'chick, chicken, gallusgallus, ggallus');

# Species 'gasterosteus_aculeatus' (id = 69293)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (69293, 69293, 'gasterosteus_aculeatus', 'gaculeatus, gasterosteusaculeatus, stickleback');

# Species 'gorilla_gorilla' (id = 9595)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (9595, 9595, 'gorilla_gorilla', 'gorilla');

# Species 'homo_sapiens' (id = 9606)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (9606, 9606, 'homo_sapiens', 'homosapiens, hsapiens, human');

# Species 'ictidomys_tridecemlineatus' (id = 43179)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (43179, 43179, 'ictidomys_tridecemlineatus', 'squirrel, thirteen-lined ground squirrel,');

# Species 'ixodes_scapularis' (id = 6945)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (6945, 6945, 'ixodes_scapularis', 'ixodes, ixodesscapularis, ixodes scapularis');

# Species 'latimeria_chalumnae' (id = 7897)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (7897, 7897, 'latimeria_chalumnae', 'coelacanth, lchalumnae, living fossil, lamineria');

# Species 'lepisosteus_oculatus' (id = 7918)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (7918, 7918, 'lepisosteus_oculatus', 'spotted gar');

# Species 'loxodonta_africana' (id = 9785)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (9785, 9785, 'loxodonta_africana', 'elephant, lafricana, loxodontaafricana');

# Species 'macaca_mulatta' (id = 9544)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (9544, 9544, 'macaca_mulatta', 'macaque, rhesus, rhesus macaque, rmacaque');

# Species 'macropus_eugenii' (id = 9315)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (9315, 9315, 'macropus_eugenii', 'wallaby, macropus, tammar wallaby, macropuseugenii');

# Species 'meleagris_gallopavo' (id = 9103)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (9103, 9103, 'meleagris_gallopavo', 'turkey, common turkey, meleagris gallopavo, mgallopavo');

# Species 'microcebus_murinus' (id = 30608)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (30608, 30608, 'microcebus_murinus', 'mouse lemur, mmur, mmurinus, mouse_lemur, microcebus murinus');

# Species 'monodelphis_domestica' (id = 13616)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (13616, 13616, 'monodelphis_domestica', 'mdomestica, mono, monodelphis, monodelphisdomestica, opossum, possum');

# Species 'mus_musculus' (id = 10090)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (10090, 10090, 'mus_musculus', 'mmusculus, mouse, musmusculus');

# Species 'mustela_putorius_furo' (id = 9669)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (9669, 9669, 'mustela_putorius_furo', 'mputorius, ferret, mfuro');

# Species 'myotis_lucifugus' (id = 59463)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (59463, 59463, 'myotis_lucifugus', 'little brown bat, microbat');

# Species 'ochotona_princeps' (id = 9978)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (9978, 9978, 'ochotona_princeps', 'pika, American pika');

# Species 'oreochromis_niloticus' (id = 8128)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (8128, 8128, 'oreochromis_niloticus', 'oreochromis_niloticus, tilapia, nile tilapia, Tilapia, Nile tilapia');

# Species 'ornithorhynchus_anatinus' (id = 9258)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (9258, 9258, 'ornithorhynchus_anatinus', 'duckbilled_platypus, duckbilledplatypus, ornithorhynchusanatinus, platypus');

# Species 'oryctolagus_cuniculus' (id = 9986)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (9986, 9986, 'oryctolagus_cuniculus', 'ocuniculus, oryctolaguscuniculus, rabbit');

# Species 'oryzias_latipes' (id = 8090)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (8090, 8090, 'oryzias_latipes', 'japanese medaka, japanese ricefish, medaka');

# Species 'otolemur_garnettii' (id = 30611)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (30611, 30611, 'otolemur_garnettii', 'bushbaby, ogarnetti, otolemur, otogar, galago');

# Species 'ovis_aries' (id = 9940)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (9940, 9940, 'ovis_aries', 'sheep, ovisaries, domestic sheep, oar, ovis aries');

# Species 'pan_troglodytes' (id = 9598)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (9598, 9598, 'pan_troglodytes', 'pantroglodytes, ptroglodytes, chimpanzee, chimp');

# Species 'papio_anubis' (id = 9555)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (9555, 9555, 'papio_anubis', 'papio_anubis, panubis, panu, papanu');

# Species 'pediculus_humanus' (id = 121225)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (121225, 121225, 'pediculus_humanus', 'pediculushumanus, pediculus, phumanus');

# Species 'pelodiscus_sinensis' (id = 13735)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (13735, 13735, 'pelodiscus_sinensis', 'psinensis, softshell turtle, chinese softshell turtle, chinese soft shell turtle, p_sinensis');

# Species 'petromyzon_marinus' (id = 7757)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (7757, 7757, 'petromyzon_marinus', 'lamprey, sea lamprey, pmarinus');

# Species 'physeter_macrocephalus' (id = 9755)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (9755, 9755, 'physeter_macrocephalus', 'PhyMac,physeter_macrocephalus,sperm whale,pmac,pmacrocephalus,physeter macrocephalus,phymac');

# Species 'poecilia_formosa' (id = 48698)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (48698, 48698, 'poecilia_formosa', 'poecilia_formosa, pfor, pformosa, poefor');

# Species 'pongo_abelii' (id = 9601)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (9601, 9601, 'pongo_abelii', 'pongopygmaeus, pongo pygmaeus, ppygmaeus, orangutan, orang, pongo_abelii, pabelii');

# Species 'procavia_capensis' (id = 9813)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (9813, 9813, 'procavia_capensis', 'rockhyrax, hyrax, procap');

# Species 'pteropus_vampyrus' (id = 132908)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (132908, 132908, 'pteropus_vampyrus', 'megabat, pvampyrus, largeflyingfox');

# Species 'rattus_norvegicus' (id = 10116)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (10116, 10116, 'rattus_norvegicus', 'rat, rattusnorvegicus, rnorvegicus');

# Species 'saccharomyces_cerevisiae' (id = 4932)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (4932, 4932, 'saccharomyces_cerevisiae', 'saccharomyces, saccharomycescerevisiae, scerevisiae, yeast');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (4932, 559292, 'saccharomyces_cerevisiae', 'saccharomyces, saccharomycescerevisiae, scerevisiae, yeast');

# Species 'sarcophilus_harrisii' (id = 9305)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (9305, 9305, 'sarcophilus_harrisii', 'devil, Sarcophilus_harrisii, sarcophilus_harrisii, tasmanian_devil, taz');

# Species 'schizosaccharomyces_pombe' (id = 4896)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (4896, 4896, 'schizosaccharomyces_pombe', 'schizosaccharomyces, schizosaccharomycespombe, spombe, fission yeast');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (4896, 284812, 'schizosaccharomyces_pombe', 'schizosaccharomyces, schizosaccharomycespombe, spombe, fission yeast');

# Species 'sorex_araneus' (id = 42254)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (42254, 42254, 'sorex_araneus', 'sorex_araneus, ground_shrew, sorex araneus, saraneus, shrew');

# Species 'phaeodactylum_tricornutum' (id = 2850)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (2850, 2850, 'phaeodactylum_tricornutum', 'diatom,phaeodactlyum_tricornutum,phaeodactlyum,');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (2850, 556484, 'phaeodactylum_tricornutum', 'diatom,phaeodactlyum_tricornutum,phaeodactlyum,');

# Species 'plasmodium_vivax' (id = 5855)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (5855, 5855, 'plasmodium_vivax', 'plasmodium_vivax,');

# Species 'plasmodium_falciparum' (id = 5833)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (5833, 5833, 'plasmodium_falciparum', 'plasmodium_falciparum,plasmodium_falciparum_3D7,');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (5833, 36329, 'plasmodium_falciparum', 'plasmodium_falciparum,plasmodium_falciparum_3D7,');

# Species 'plasmodium_knowlesi' (id = 5851)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (5851, 5851, 'plasmodium_knowlesi', 'plasmodium_knowlesi,');

# Species 'plasmodium_chabaudi' (id = 31271)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (31271, 31271, 'plasmodium_chabaudi', 'plasmodium_chabaudi');

# Species 'plasmodium_berghei' (id = 5823)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (5823, 5823, 'plasmodium_berghei', 'plasmodium_berghei,plasmodium_berghei_ANKA');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (5823, 5821, 'plasmodium_berghei', 'plasmodium_berghei,plasmodium_berghei_ANKA');

# Species 'thalassiosira_pseudonana' (id = 35128)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (35128, 35128, 'thalassiosira_pseudonana', 'diatom,thalassiosira_pseudonana,thalassiosira');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (35128, 296543, 'thalassiosira_pseudonana', 'diatom,thalassiosira_pseudonana,thalassiosira');

# Species 'taeniopygia_guttata' (id = 59729)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (59729, 59729, 'taeniopygia_guttata', 'zebrafinch, zebra_finch, taeniopygiaguttata, tguttata, poephila_guttata, taenopygia_guttata, poephilaguttata, taenopygiaguttata');

# Species 'takifugu_rubripes' (id = 31033)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (31033, 31033, 'takifugu_rubripes', 'frubripes, fugu, fugu_rubripes, fugurubripes, pufferfish');

# Species 'tarsius_syrichta' (id = 9478)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (9478, 9478, 'tarsius_syrichta', 'tarsier, bonyfingers');

# Species 'tetraodon_nigroviridis' (id = 99883)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (99883, 99883, 'tetraodon_nigroviridis', 'tetraodon, tnigroviridis, tnig, tetraodon_nigroviridis');

# Species 'tupaia_belangeri' (id = 37347)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (37347, 37347, 'tupaia_belangeri', 'treeshrew');

# Species 'tursiops_truncatus' (id = 9739)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (9739, 9739, 'tursiops_truncatus', 'dolphin, bottlenose_dolphi');

# Species 'vicugna_pacos' (id = 30538)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (30538, 30538, 'vicugna_pacos', 'alpaca, lama_pacos');

# Species 'xenopus_tropicalis' (id = 8364)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (8364, 8364, 'xenopus_tropicalis', 'pipid, pipidfrog, xenopus, xenopustropicalis, xtropicalis');

# Species 'xiphophorus_maculatus' (id = 8083)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (8083, 8083, 'xiphophorus_maculatus', 'xiphophorus, xmaculatus, platyfish');

# Species 'magnaporthe_oryzae' (id = 242507)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (242507, 242507, 'magnaporthe_oryzae', 'rice blast, magnaporthe, magnaportheoryzae, moryzae');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (242507, 148305, 'magnaporthe_oryzae', 'rice blast, magnaporthe, magnaportheoryzae, moryzae');

# Species 'phytophthora_infestans' (id = 403677)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (403677, 403677, 'phytophthora_infestans', 'phytophthora_infestans');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (403677, 4787, 'phytophthora_infestans', 'phytophthora_infestans');

# Species 'phytophthora_sojae' (id = 67593)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (67593, 67593, 'phytophthora_sojae', 'phytophthora_sojae');

# Species 'phytophthora_ramorum' (id = 164328)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (164328, 164328, 'phytophthora_ramorum', 'phytophthora_ramorum');

# Species 'nectria_haematococca' (id = 140110)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (140110, 140110, 'nectria_haematococca', 'nectria_haematococca');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (140110, 660122, 'nectria_haematococca', 'nectria_haematococca');

# Species 'fusarium_oxysporum' (id = 426428)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (426428, 426428, 'fusarium_oxysporum', 'fusarium_oxysporum, fusarium oxysporum, Fusarium oxysporum f. sp. lycopersici 4287, Fusarium oxysporum f. sp. lycopersici');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (426428, 59765, 'fusarium_oxysporum', 'fusarium_oxysporum, fusarium oxysporum, Fusarium oxysporum f. sp. lycopersici 4287, Fusarium oxysporum f. sp. lycopersici');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (426428, 5507, 'fusarium_oxysporum', 'fusarium_oxysporum, fusarium oxysporum, Fusarium oxysporum f. sp. lycopersici 4287, Fusarium oxysporum f. sp. lycopersici');

# Species 'tuber_melanosporum' (id = 656061)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (656061, 656061, 'tuber_melanosporum', 'tuber_melanosporum');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (656061, 39416, 'tuber_melanosporum', 'tuber_melanosporum');

# Species 'phaeosphaeria_nodorum' (id = 321614)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (321614, 321614, 'phaeosphaeria_nodorum', 'phaeosphaeria_nodorum');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (321614, 13684, 'phaeosphaeria_nodorum', 'phaeosphaeria_nodorum');

# Species 'ashbya_gossypii' (id = 284811)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (284811, 284811, 'ashbya_gossypii', 'ashbya_gossypii');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (284811, 33169, 'ashbya_gossypii', 'ashbya_gossypii');

# Species 'gibberella_zeae' (id = 229533)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (229533, 229533, 'gibberella_zeae', 'gibberella_zeae, gibberella zeae, gibberella zeae PH1');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (229533, 5518, 'gibberella_zeae', 'gibberella_zeae, gibberella zeae, gibberella zeae PH1');

# Species 'nomascus_leucogenys' (id = 61853)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (61853, 61853, 'nomascus_leucogenys', 'nomascus_leucogenys, gibbon, Gibbon');

# Species 'sus_scrofa' (id = 9823)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (9823, 9823, 'sus_scrofa', 'pig, sscrofa, susscr, sscr, sus scrofa');

# Species 'aegilops_tauschii' (id = 37682)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (37682, 37682, 'aegilops_tauschii', 'Aegilops tauschii, goatgrass');

# Species 'vitis_vinifera' (id = 29760)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (29760, 29760, 'vitis_vinifera', 'grape, vitis vinifera, vvinifera');

# Species 'glycine_max' (id = 3847)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (3847, 3847, 'glycine_max', 'soybean, glycine max, glycinemax');

# Species 'oryza_glaberrima' (id = 4538)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (4538, 4538, 'oryza_glaberrima', 'african rice, oryza glaberrima, oryzaglaberrima');

# Species 'brassica_rapa' (id = 51351)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (51351, 51351, 'brassica_rapa', 'Brassica rapa, Brassica rapa subsp. pekinensis');

# Species 'oryza_brachyantha' (id = 4533)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (4533, 4533, 'oryza_brachyantha', 'Oryza brachyantha');

# Species 'setaria_italica' (id = 4555)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (4555, 4555, 'setaria_italica', 'Setaria italica');

# Species 'solanum_lycopersicum' (id = 4081)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (4081, 4081, 'solanum_lycopersicum', 'Solanum lycopersicum, Lycopersicon esculentum');

# Species 'solanum_tuberosum' (id = 4113)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (4113, 4113, 'solanum_tuberosum', 'Solanum tuberosum');

# Species 'medicago_truncatula' (id = 3880)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (3880, 3880, 'medicago_truncatula', 'Medicago truncatula, barrel medic, Medicago tribuloides');

# Species 'musa_acuminata' (id = 214687)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (214687, 214687, 'musa_acuminata', 'Musa acuminata');

# Species 'hordeum_vulgare' (id = 112509)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (112509, 112509, 'hordeum_vulgare', 'Hordeum vulgare, barley');

# Species 'arabidopsis_lyrata' (id = 81972)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (81972, 81972, 'arabidopsis_lyrata', 'Arabidopsis lyrata,alyrata,arabidopsislyrata,lyrate rockcress');

# Species 'arabidopsis_thaliana' (id = 3702)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (3702, 3702, 'arabidopsis_thaliana', 'Arabidopsis thaliana,athaliana,arabidopsisthaliana,thale cress,thale-cress,mouse-ear cress');

# Species 'brachypodium_distachyon' (id = 15368)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (15368, 15368, 'brachypodium_distachyon', 'Brachypodium distachyon,bdistachyon,brachypodiumdistachyon');

# Species 'chlamydomonas_reinhardtii' (id = 3055)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (3055, 3055, 'chlamydomonas_reinhardtii', 'Chlamydomonas reinhardtii,creinhardtii,chlamydomonasreinhardtii,green algae');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (3055, 906914, 'chlamydomonas_reinhardtii', 'Chlamydomonas reinhardtii,creinhardtii,chlamydomonasreinhardtii,green algae');

# Species 'cyanidioschyzon_merolae' (id = 45157)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (45157, 45157, 'cyanidioschyzon_merolae', 'Cyanidioschyzon merolae,cmerolae,cyanidioschyzonmerolae,red algae');

# Species 'leersia_perrieri' (id = 77586)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (77586, 77586, 'leersia_perrieri', 'Leersia perrieri,lperrieri,leersiaperrieri');

# Species 'oryza_barthii' (id = 65489)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (65489, 65489, 'oryza_barthii', 'Oryza barthii,obarthii,oryzabarthii,African wild rice');

# Species 'oryza_glumaepatula' (id = 40148)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (40148, 40148, 'oryza_glumaepatula', 'Oryza glumaepatula,oglumaepatula,oryzaglumaepatula');

# Species 'oryza_granulata' (id = 110450)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (110450, 110450, 'oryza_granulata', 'Oryza granulata,ogranulata,oryzagranulata');

# Species 'oryza_longistaminata' (id = 4528)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (4528, 4528, 'oryza_longistaminata', 'Oryza longistaminata,olongistaminata,oryzalongistaminata,red rice');

# Species 'oryza_meridionalis' (id = 40149)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (40149, 40149, 'oryza_meridionalis', 'Oryza meridionalis,omeridionalis,oryzameridionalis');

# Species 'oryza_nivara' (id = 4536)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (4536, 4536, 'oryza_nivara', 'Oryza nivara,onivara,oryzanivara');

# Species 'oryza_officinalis' (id = 4535)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (4535, 4535, 'oryza_officinalis', 'Oryza officinalis,oofficinalis,oryzaofficinalis');

# Species 'oryza_punctata' (id = 4537)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (4537, 4537, 'oryza_punctata', 'Oryza punctata,opunctata,oryzapunctata');

# Species 'oryza_indica' (id = 39946)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (39946, 39946, 'oryza_indica', 'Oryza sativa,osativa,oryzasativa,rice,Indian rice');

# Species 'oryza_sativa' (id = 39947)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (39947, 39947, 'oryza_sativa', 'Oryza sativa,osativa,oryzasativa,rice,Japanese rice');

# Species 'physcomitrella_patens' (id = 145481)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (145481, 145481, 'physcomitrella_patens', 'Physcomitrella patens,ppatens,physcomitrellapatens,moss,Physcomitrella patens subsp. patens');

# Species 'populus_trichocarpa' (id = 3694)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (3694, 3694, 'populus_trichocarpa', 'Populus trichocarpa,ptrichocarpa,populustrichocarpa,poplar,black cottonwood');

# Species 'selaginella_moellendorffii' (id = 88036)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (88036, 88036, 'selaginella_moellendorffii', 'Selaginella moellendorffii,smoellendorffii,selaginellamoellendorffii,club moss');

# Species 'sorghum_bicolor' (id = 4558)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (4558, 4558, 'sorghum_bicolor', 'Sorghum bicolor,sbicolor,sorghumbicolor,sorghum,milo,broomcorn,shattercane,chicken cornSorghum bicolor subsp. verticilliflorum,Sorghum bicolor subsp. x drummondii');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (4558, 91525, 'sorghum_bicolor', 'Sorghum bicolor,sbicolor,sorghumbicolor,sorghum,milo,broomcorn,shattercane,chicken cornSorghum bicolor subsp. verticilliflorum,Sorghum bicolor subsp. x drummondii');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (4558, 171959, 'sorghum_bicolor', 'Sorghum bicolor,sbicolor,sorghumbicolor,sorghum,milo,broomcorn,shattercane,chicken cornSorghum bicolor subsp. verticilliflorum,Sorghum bicolor subsp. x drummondii');

# Species 'triticum_urartu' (id = 4572)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (4572, 4572, 'triticum_urartu', 'Triticum urartu, wheat');

# Species 'triticum_aestivum' (id = 4565)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (4565, 4565, 'triticum_aestivum', 'Triticum aestivum, bread wheat, common wheat');

# Species 'zea_mays' (id = 4577)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (4577, 4577, 'zea_mays', 'Zea mays,zmays,zeamays,corn,maize');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (4577, 112001, 'zea_mays', 'Zea mays,zmays,zeamays,corn,maize');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (4577, 381124, 'zea_mays', 'Zea mays,zmays,zeamays,corn,maize');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (4577, 334825, 'zea_mays', 'Zea mays,zmays,zeamays,corn,maize');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (4577, 4579, 'zea_mays', 'Zea mays,zmays,zeamays,corn,maize');
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (4577, 76912, 'zea_mays', 'Zea mays,zmays,zeamays,corn,maize');

# Species 'amphimedon_queenslandica' (id = 400682)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (400682, 400682, 'amphimedon_queenslandica', 'sponge, aqueenslandica, amphimedon_queenslandica');

# Species 'atta_cephalotes' (id = 12957)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (12957, 12957, 'atta_cephalotes', 'acephalotes, atta_cephalotes, Atta cephalotes');

# Species 'bombyx_mori' (id = 7091)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (7091, 7091, 'bombyx_mori', 'silkworm, silk moth, bmori, Bombyx mori');

# Species 'nematostella_vectensis' (id = 45351)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (45351, 45351, 'nematostella_vectensis', 'anemone, sea anemone, nvectensis, nematostella_vectensis');

# Species 'strongylocentrotus_purpuratus' (id = 7668)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (7668, 7668, 'strongylocentrotus_purpuratus', 'sea urchin, spurpuratus, strongylocentrotus_purpuratus');

# Species 'tribolium_castaneum' (id = 7070)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (7070, 7070, 'tribolium_castaneum', 'red flour beetle, tcastaneum, tribolium_castaneum');

# Species 'gaeumannomyces_graminis' (id = 644352)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (644352, 644352, 'gaeumannomyces_graminis', 'ggraminis, gaeumannomyces_graminis, gaeumannomyces graminis, gaeumannomyces graminis var. tritici R3-111a-1');

# Species 'trichinella_spiralis' (id = 6334)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (6334, 6334, 'trichinella_spiralis', 'trichinella_spiralis');

# Species 'nasonia_vitripennis' (id = 7425)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (7425, 7425, 'nasonia_vitripennis', 'nasoniavitripennis, nvitripennis, Nasonia vitripennis, jewel wasp');

# Species 'danaus_plexippus' (id = 13037)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (13037, 13037, 'danaus_plexippus', 'dplexippus, danaus plexippus, monarch butterfly');

# Species 'heliconius_melpomene' (id = 34740)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (34740, 34740, 'heliconius_melpomene', 'hmelpomene, heliconius melpomene, postman butterfly');

# Species 'strigamia_maritima' (id = 126957)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (126957, 126957, 'strigamia_maritima', 'strigamiamaritima, smaritima, Strigamia maritima');

# Species 'megaselia_scalaris' (id = 36166)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (36166, 36166, 'megaselia_scalaris', 'megaseliascalaris, mscalaris, Megaselia scalaris');

# Species 'capitella_teleta' (id = 283909)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (283909, 283909, 'capitella_teleta', 'capitellateleta, cteleta, Capitella teleta');

# Species 'crassostrea_gigas' (id = 29159)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (29159, 29159, 'crassostrea_gigas', 'crassostreagigas, cgigas, Crassostrea gigas');

# Species 'helobdella_robusta' (id = 6412)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (6412, 6412, 'helobdella_robusta', 'helobdellarobusta, hrobusta, Helobdella robusta');

# Species 'lottia_gigantea' (id = 225164)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (225164, 225164, 'lottia_gigantea', 'lottiagigantea, lgigantea, Lottia gigantea');

# Species 'rhodnius_prolixus' (id = 13249)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (13249, 13249, 'rhodnius_prolixus', 'rhodniusprolixus, rprolixus, Rhodnius prolixus');

# Species 'tetranychus_urticae' (id = 32264)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (32264, 32264, 'tetranychus_urticae', 'tetranychusurticae, turticae, Tetranychus urticae');

# Species 'dendroctonus_ponderosae' (id = 77166)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (77166, 77166, 'dendroctonus_ponderosae', 'dendroctonusponderosae, dponderosae, Dendroctonus ponderosae');

# Species 'solenopsis_invicta' (id = 13686)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (13686, 13686, 'solenopsis_invicta', 'solenopsisinvicta, sinvicta, Solenopsis invicta');

# Species 'mnemiopsis_leidyi' (id = 27923)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (27923, 27923, 'mnemiopsis_leidyi', 'mnemiopsis leidyi, mleidyi, warty comb jelly, sea walnut');

# Species 'melitaea_cinxia' (id = 113334)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (113334, 113334, 'melitaea_cinxia', 'Glanville fritillary, melitaeacinxia, mcinxia, Melitaea cinxia');

# Species 'caenorhabditis_elegans' (id = 6239)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (6239, 6239, 'caenorhabditis_elegans', 'caenorhabditiselegans, celegans, elegans');

# Species 'caenorhabditis_briggsae' (id = 6238)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (6238, 6238, 'caenorhabditis_briggsae', 'caenorhabditis_briggae_prjna10731');

# Species 'caenorhabditis_remanei' (id = 31234)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (31234, 31234, 'caenorhabditis_remanei', 'caenorhabditis_remanei_prjna53967');

# Species 'caenorhabditis_brenneri' (id = 135651)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (135651, 135651, 'caenorhabditis_brenneri', 'caenorhabditis_brenneri_prjna20035');

# Species 'caenorhabditis_japonica' (id = 281687)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (281687, 281687, 'caenorhabditis_japonica', 'caenorhabditis_japonica_prjna12591');

# Species 'brugia_malayi' (id = 6279)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (6279, 6279, 'brugia_malayi', 'brugia_malayi_prjna10729');

# Species 'onchocerca_volvulus' (id = 6282)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (6282, 6282, 'onchocerca_volvulus', 'onchocerca_volvulus_prjeb513');

# Species 'pristionchus_pacificus' (id = 54126)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (54126, 54126, 'pristionchus_pacificus', 'pristionchus_pacificus_prjna12644');

# Species 'strongyloides_ratti' (id = 34506)
INSERT INTO species (species_id, taxonomy_id, name, aliases)
VALUES (34506, 34506, 'strongyloides_ratti', 'strongyloides_ratti_prjeb125');

################################################################################
# SOURCES

# Source 'AFFY_DrosGenome1::drosophila_melanogaster' (id = 1)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('AFFY_DrosGenome1', '1', 'N', 50, 1, '', 'NOIDEA');

# Source 'AedesGenBank::aedes_aegypti' (id = 2)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('AedesGenBank', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'AnoXcel::anopheles_gambiae' (id = 3)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('AnoXcel', '1', 'Y', 40, 1, '', 'NOIDEA');

# Source 'ArrayExpress::EG' (id = 4)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('ArrayExpress', '1', 'Y', 50, 1, '', 'NOIDEA');

# Source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('ArrayExpress', '1', 'Y', 50, 1, '', 'NOIDEA');

# Source 'AspGD::MULTI' (id = 6)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('AspGD', '1', 'N', 70, 1, '', 'NOIDEA');

# Source 'BDGP_insitu_expr::drosophila_melanogaster' (id = 7)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('BDGP_insitu_expr', '1', 'N', 50, 1, 'BDGP in situ gene expression', 'NOIDEA');

# Source 'BROAD_F_oxysporum::MULTI' (id = 8)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('BROAD_F_oxysporum', '1', 'N', 70, 1, '', 'NOIDEA');

# Source 'BROAD_G_moniliformis::MULTI' (id = 9)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('BROAD_G_moniliformis', '1', 'N', 70, 1, '', 'NOIDEA');

# Source 'BROAD_Magnaporthe_DB::MULTI' (id = 10)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('BROAD_Magnaporthe_DB', '1', 'Y', 70, 1, '', 'NOIDEA');

# Source 'BROAD_P_infestans::MULTI' (id = 11)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('BROAD_P_infestans', '1', 'N', 70, 1, '', 'NOIDEA');

# Source 'BROAD_P_infestans_GENE::MULTI' (id = 12)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('BROAD_P_infestans_GENE', '1', 'N', 70, 1, '', 'NOIDEA');

# Source 'BROAD_P_infestans_TRANSCRIPT::MULTI' (id = 13)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('BROAD_P_infestans_TRANSCRIPT', '1', 'N', 70, 1, '', 'NOIDEA');

# Source 'BROAD_U_maydis::MULTI' (id = 14)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('BROAD_U_maydis', '1', 'N', 70, 1, '', 'NOIDEA');

# Source 'BROAD_g_zeae::MULTI' (id = 15)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('BROAD_G_zeae', '1', 'N', 70, 1, '', 'NOIDEA');

# Source 'BioGRID::drosophila_melanogaster' (id = 16)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('BioGRID', '1', 'N', 50, 1, '', 'NOIDEA');

# Source 'CADRE::MULTI' (id = 17)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('CADRE', '1', 'N', 70, 1, '', 'NOIDEA');

# Source 'CADRE_AFum_A1163::MULTI' (id = 18)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('CADRE_Afum_A1163', '1', 'N', 70, 1, '', 'NOIDEA');

# Source 'CCDS::homo_sapiens' (id = 19)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('CCDS', '1', 'Y', 10, 1, '', 'NOIDEA');

# Source 'CCDS::mus_musculus' (id = 20)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('CCDS', '1', 'Y', 10, 1, '', 'NOIDEA');

# Source 'CGNC::gallus_gallus' (id = 21)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('CGNC', '1', 'Y', 30, 1, 'cgnc', 'NOIDEA');

# Source 'Celera_Gene::anopheles_gambiae' (id = 22)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Celera_Gene', '1', 'N', 40, 1, '', 'NOIDEA');

# Source 'Celera_Pep::anopheles_gambiae' (id = 23)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Celera_Pep', '1', 'Y', 40, 1, '', 'NOIDEA');

# Source 'Celera_Trans::anopheles_gambiae' (id = 24)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Celera_Trans', '1', 'Y', 40, 1, '', 'NOIDEA');

# Source 'Clone_based_ensembl_gene::homo_sapiens' (id = 25)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Clone_based_ensembl_gene', '1', 'N', 50, 1, '', 'NOIDEA');

# Source 'Clone_based_ensembl_transcript::homo_sapiens' (id = 26)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Clone_based_ensembl_transcript', '1', 'N', 50, 1, '', 'NOIDEA');

# Source 'Clone_based_vega_gene::homo_sapiens' (id = 27)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Clone_based_vega_gene', '1', 'N', 50, 1, '', 'NOIDEA');

# Source 'Clone_based_vega_transcript::homo_sapiens' (id = 28)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Clone_based_vega_transcript', '1', 'N', 50, 1, '', 'NOIDEA');

# Source 'Codelink::homo_sapiens' (id = 29)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Codelink', '1', 'Y', 50, 1, '', 'NOIDEA');

# Source 'Codelink::mus_musculus' (id = 30)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Codelink', '1', 'Y', 50, 1, '', 'NOIDEA');

# Source 'Codelink::rattus_norvegicus' (id = 31)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Codelink', '1', 'Y', 50, 1, '', 'NOIDEA');

# Source 'CommunityGO::arabidopsis_thaliana' (id = 32)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('GO', '1', 'Y', 85, 1, 'main', 'NOIDEA');

# adding source dependency that source CommunityGO::arabidopsis_thaliana needs TAIR_TRANSLATION loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (32, 'TAIR_TRANSLATION');

# Source 'DBASS3::homo_sapiens' (id = 33)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('DBASS3', '1', 'Y', 50, 1, 'Database of aberrant 3\' splice sites.', 'NOIDEA');

# Source 'DBASS5::homo_sapiens' (id = 34)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('DBASS5', '1', 'Y', 50, 1, 'Database of aberrant 5\' splice sites.', 'NOIDEA');

# Source 'DEDb::drosophila_melanogaster' (id = 35)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('DEDb', '1', 'N', 50, 1, 'Drosophila melanogaster Exon DataBase', 'NOIDEA');

# Source 'EC_NUMBER::saccharomyces_cerevisiae' (id = 36)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('EC_NUMBER', '1', 'N', 50, 70, '', 'NOIDEA');

# Source 'ENA_GENE::MULTI' (id = 37)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('ENA_GENE', '1', 'N', 70, 1, '', 'NOIDEA');

# Source 'EPD::drosophila_melanogaster' (id = 38)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('EPD', '1', 'N', 50, 1, '', 'NOIDEA');

# Source 'EPD::drosophila_mojavensis' (id = 39)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('EPD', '1', 'N', 50, 1, '', 'NOIDEA');

# Source 'EPD::drosophila_pseudoobscura' (id = 40)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('EPD', '1', 'N', 50, 1, '', 'NOIDEA');

# Source 'EPD::drosophila_sechellia' (id = 41)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('EPD', '1', 'N', 50, 1, '', 'NOIDEA');

# Source 'EPD::drosophila_simulans' (id = 42)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('EPD', '1', 'N', 50, 1, '', 'NOIDEA');

# Source 'EPD::drosophila_virilis' (id = 43)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('EPD', '1', 'N', 50, 1, '', 'NOIDEA');

# Source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('EntrezGene', '1', 'Y', 10, 1, '', 'NOIDEA');

# Source 'EntrezGene_trans_name' (id = 45)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('EntrezGene_trans_name', '1', 'N', 70, 1, '', 'NOIDEA');

# Source 'FlyAtlas::drosophila_melanogaster' (id = 46)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('FlyAtlas', '1', 'N', 50, 1, '', 'NOIDEA');

# Source 'FlyAtlas::drosophila_pseudoobscura' (id = 47)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('FlyAtlas', '1', 'N', 50, 1, '', 'NOIDEA');

# Source 'FlyBaseCGID_gene::drosophila_melanogaster' (id = 48)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('FlyBaseCGID_gene', '1', 'N', 50, 1, 'FlyBase_Annotation_IDs', 'NOIDEA');

# Source 'FlyBaseCGID_transcript::drosophila_melanogaster' (id = 49)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('FlyBaseCGID_transcript', '1', 'N', 50, 1, 'FlyBase_Annotation_IDs', 'NOIDEA');

# Source 'FlyBaseCGID_translation::drosophila_melanogaster' (id = 50)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('FlyBaseCGID_translation', '1', 'N', 50, 1, 'FlyBase_Annotation_IDs', 'NOIDEA');

# Source 'FlyBaseName_gene::drosophila_melanogaster' (id = 51)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('FlyBaseName_gene', '1', 'N', 50, 1, 'Name assigned to gene in FlyBase gff', 'NOIDEA');

# Source 'FlyBaseName_transcript::drosophila_melanogaster' (id = 52)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('FlyBaseName_transcript', '1', 'N', 50, 1, 'Transcript name in FlyBase gff', 'NOIDEA');

# Source 'FlyBaseName_translations::drosophila_melanogaster' (id = 53)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('FlyBaseName_translation', '1', 'N', 50, 1, 'Translation name in FlyBase gff', 'NOIDEA');

# Source 'FlyExpress::drosophila_melanogaster' (id = 54)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('FlyExpress', '1', 'N', 50, 1, '', 'NOIDEA');

# Source 'FlyReactome::drosophila_melanogaster' (id = 55)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('FlyReactome', '1', 'N', 50, 1, '', 'NOIDEA');

# Source 'GO::MULTI' (id = 56)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('GO', '1', 'Y', 80, 1, 'main', 'NOIDEA');

# adding source dependency that source GO::MULTI needs Uniprot/SPTREMBL loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (56, 'Uniprot/SPTREMBL');

# adding source dependency that source GO::MULTI needs Uniprot/SWISSPROT loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (56, 'Uniprot/SWISSPROT');

# adding source dependency that source GO::MULTI needs RefSeq_dna loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (56, 'RefSeq_dna');

# adding source dependency that source GO::MULTI needs RefSeq_peptide loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (56, 'RefSeq_peptide');

# adding source dependency that source GO::MULTI needs SGD loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (56, 'SGD');

# Source 'GO::caenorhabditis_elegans' (id = 57)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('GO', '1', 'Y', 85, 1, 'main', 'NOIDEA');

# adding source dependency that source GO::caenorhabditis_elegans needs Uniprot/SPTREMBL loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (57, 'Uniprot/SPTREMBL');

# adding source dependency that source GO::caenorhabditis_elegans needs Uniprot/SWISSPROT loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (57, 'Uniprot/SWISSPROT');

# adding source dependency that source GO::caenorhabditis_elegans needs RefSeq_dna loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (57, 'RefSeq_dna');

# adding source dependency that source GO::caenorhabditis_elegans needs RefSeq_peptide loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (57, 'RefSeq_peptide');

# adding source dependency that source GO::caenorhabditis_elegans needs wormbase_all loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (57, 'wormbase_all');

# Source 'GO::caenorhabditis_elegans_gene' (id = 58)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('GO_to_gene', '1', 'Y', 85, 1, 'main', 'NOIDEA');

# Source 'GO::danio_rerio' (id = 59)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('GO', '1', 'Y', 85, 1, 'main', 'NOIDEA');

# adding source dependency that source GO::danio_rerio needs Uniprot/SPTREMBL loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (59, 'Uniprot/SPTREMBL');

# adding source dependency that source GO::danio_rerio needs Uniprot/SWISSPROT loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (59, 'Uniprot/SWISSPROT');

# adding source dependency that source GO::danio_rerio needs RefSeq_dna loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (59, 'RefSeq_dna');

# adding source dependency that source GO::danio_rerio needs RefSeq_peptide loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (59, 'RefSeq_peptide');

# Source 'GO::drosophila_melanogaster' (id = 60)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('GO', '1', 'Y', 85, 1, 'flybasego', 'NOIDEA');

# adding source dependency that source GO::drosophila_melanogaster needs Uniprot/SPTREMBL loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (60, 'Uniprot/SPTREMBL');

# adding source dependency that source GO::drosophila_melanogaster needs Uniprot/SWISSPROT loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (60, 'Uniprot/SWISSPROT');

# adding source dependency that source GO::drosophila_melanogaster needs RefSeq_dna loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (60, 'RefSeq_dna');

# adding source dependency that source GO::drosophila_melanogaster needs RefSeq_peptide loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (60, 'RefSeq_peptide');

# adding source dependency that source GO::drosophila_melanogaster needs flybase_gene_id loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (60, 'flybase_gene_id');

# Source 'GO::homo_sapiens' (id = 61)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('GO', '1', 'Y', 85, 1, 'main', 'NOIDEA');

# adding source dependency that source GO::homo_sapiens needs Uniprot/SPTREMBL loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (61, 'Uniprot/SPTREMBL');

# adding source dependency that source GO::homo_sapiens needs Uniprot/SWISSPROT loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (61, 'Uniprot/SWISSPROT');

# adding source dependency that source GO::homo_sapiens needs RefSeq_dna loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (61, 'RefSeq_dna');

# adding source dependency that source GO::homo_sapiens needs RefSeq_peptide loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (61, 'RefSeq_peptide');

# Source 'GO::mus_musculus' (id = 62)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('GO', '1', 'Y', 85, 1, 'main', 'NOIDEA');

# adding source dependency that source GO::mus_musculus needs Uniprot/SPTREMBL loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (62, 'Uniprot/SPTREMBL');

# adding source dependency that source GO::mus_musculus needs Uniprot/SWISSPROT loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (62, 'Uniprot/SWISSPROT');

# adding source dependency that source GO::mus_musculus needs RefSeq_dna loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (62, 'RefSeq_dna');

# adding source dependency that source GO::mus_musculus needs RefSeq_peptide loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (62, 'RefSeq_peptide');

# Source 'GO::rattus_norvegicus' (id = 63)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('GO', '1', 'Y', 85, 1, 'main', 'NOIDEA');

# adding source dependency that source GO::rattus_norvegicus needs Uniprot/SPTREMBL loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (63, 'Uniprot/SPTREMBL');

# adding source dependency that source GO::rattus_norvegicus needs Uniprot/SWISSPROT loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (63, 'Uniprot/SWISSPROT');

# adding source dependency that source GO::rattus_norvegicus needs RefSeq_dna loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (63, 'RefSeq_dna');

# adding source dependency that source GO::rattus_norvegicus needs RefSeq_peptide loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (63, 'RefSeq_peptide');

# Source 'GO::saccharomyces_cerevisiae' (id = 64)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('GO', '1', 'Y', 85, 1, 'main', 'NOIDEA');

# adding source dependency that source GO::saccharomyces_cerevisiae needs Uniprot/SPTREMBL loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (64, 'Uniprot/SPTREMBL');

# adding source dependency that source GO::saccharomyces_cerevisiae needs Uniprot/SWISSPROT loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (64, 'Uniprot/SWISSPROT');

# adding source dependency that source GO::saccharomyces_cerevisiae needs RefSeq_dna loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (64, 'RefSeq_dna');

# adding source dependency that source GO::saccharomyces_cerevisiae needs RefSeq_peptide loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (64, 'RefSeq_peptide');

# adding source dependency that source GO::saccharomyces_cerevisiae needs SGD loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (64, 'SGD');

# Source 'GeneDB::MULTI' (id = 65)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('GeneDB', '1', 'N', 70, 1, '', 'NOIDEA');

# Source 'GeneDB_Spombe::MULTI' (id = 66)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('GeneDB_Spombe', '1', 'N', 70, 1, '', 'NOIDEA');

# Source 'GenomeRNAi::drosophila_melanogaster' (id = 67)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('GenomeRNAi', '1', 'N', 50, 1, '', 'NOIDEA');

# Source 'Genoscope_annotated_gene::tetraodon_nigroviridis' (id = 68)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Genoscope_annotated_gene', '1', 'Y', 40, 1, '', 'NOIDEA');

# Source 'Genoscope_pred_gene::tetraodon_nigroviridis' (id = 69)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Genoscope_pred_gene', '1', 'Y', 40, 1, '', 'NOIDEA');

# Source 'Genoscope_pred_transcript::tetraodon_nigroviridis' (id = 70)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Genoscope_tnig_transcript', '1', 'Y', 40, 1, '', 'NOIDEA');

# Source 'Genoscope_pred_translation::tetraodon_nigroviridis' (id = 71)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Genoscope_pred_translation', '1', 'Y', 40, 1, '', 'NOIDEA');

# Source 'Gramene_Pathway::arabidopsis_thaliana' (id = 72)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Gramene_Pathway', '1', 'Y', 50, 1, '', 'NOIDEA');

# Source 'Gramene_Pathway::brachypodium_distachyon' (id = 73)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Gramene_Pathway', '1', 'Y', 50, 1, '', 'NOIDEA');

# Source 'Gramene_Pathway::oryza_sativa' (id = 74)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Gramene_Pathway', '1', 'Y', 50, 1, '', 'NOIDEA');

# Source 'Gramene_Pathway::populus_trichocarpa' (id = 75)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Gramene_Pathway', '1', 'Y', 50, 1, '', 'NOIDEA');

# Source 'Gramene_Pathway::solanum_lycopersicum' (id = 76)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Gramene_Pathway', '1', 'Y', 50, 1, '', 'NOIDEA');

# Source 'Gramene_Pathway::solanum_tuberosum' (id = 77)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Gramene_Pathway', '1', 'Y', 50, 1, '', 'NOIDEA');

# Source 'Gramene_Pathway::sorghum_bicolor' (id = 78)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Gramene_Pathway', '1', 'Y', 50, 1, '', 'NOIDEA');

# Source 'Gramene_Pathway::zea_mays' (id = 79)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Gramene_Pathway', '1', 'Y', 50, 1, '', 'NOIDEA');

# Source 'HGNC::homo_sapiens' (id = 80)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('HGNC', '1', 'Y', 30, 3, 'vega', 'NOIDEA');

# Source 'HGNC::homo_sapiens#01' (id = 81)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('HGNC', '1', 'Y', 30, 2, 'ccds', 'NOIDEA');

# Source 'HGNC::homo_sapiens#02' (id = 82)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('HGNC', '1', 'Y', 29, 4, 'entrezgene_manual', 'NOIDEA');

# adding source dependency that source HGNC::homo_sapiens#02 needs EntrezGene loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (82, 'EntrezGene');

# adding source dependency that source HGNC::homo_sapiens#02 needs Uniprot/SWISSPROT loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (82, 'Uniprot/SWISSPROT');

# adding source dependency that source HGNC::homo_sapiens#02 needs RefSeq_dna loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (82, 'RefSeq_dna');

# adding source dependency that source HGNC::homo_sapiens#02 needs RefSeq_peptide loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (82, 'RefSeq_peptide');

# Source 'HGNC::homo_sapiens#03' (id = 83)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('HGNC', '1', 'N', 30, 5, 'refseq_manual', 'NOIDEA');

# Source 'HGNC::homo_sapiens#04' (id = 84)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('HGNC', '1', 'N', 30, 6, 'entrezgene_mapped', 'NOIDEA');

# Source 'HGNC::homo_sapiens#05' (id = 85)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('HGNC', '1', 'N', 30, 7, 'refseq_mapped', 'NOIDEA');

# Source 'HGNC::homo_sapiens#06' (id = 86)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('HGNC', '1', 'N', 30, 8, 'uniprot', 'NOIDEA');

# Source 'HGNC::homo_sapiens#07' (id = 87)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('HGNC', '1', 'N', 30, 1, 'ensembl_manual', 'NOIDEA');

# Source 'HGNC::homo_sapiens#08' (id = 88)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('HGNC', '1', 'N', 30, 100, 'desc_only', 'NOIDEA');

# Source 'HGNC::homo_sapiens#09' (id = 89)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('HGNC', '1', 'N', 30, 5, 'swissprot_manual', 'NOIDEA');

# Source 'HGNC_curated_transcript::homo_sapiens' (id = 90)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('HGNC_curated_transcript_notransfer', '1', 'Y', 49, 1, 'curated with transcript number postfix', 'NOIDEA');

# Source 'HGNC_gene_name' (id = 91)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('HGNC_gene_name', '1', 'N', 49, 1, '', 'NOIDEA');

# Source 'HGNC_trans_name' (id = 92)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('HGNC_trans_name', '1', 'N', 49, 1, '', 'NOIDEA');

# Source 'HPA::homo_sapiens' (id = 93)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('HPA', '1', 'Y', 50, 1, 'Human Protein Atlas (HPA) database', 'NOIDEA');

# Source 'IKMCs_ES_cells_available::mus_musculus' (id = 94)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('IKMCs_ES_cells_available', '1', 'N', 40, 1, '', 'NOIDEA');

# Source 'IKMCs_KOs::mus_musculus' (id = 95)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('IKMCs_KOs', '1', 'Y', 40, 1, '', 'NOIDEA');

# Source 'IKMCs_Mice_available::mus_musculus' (id = 96)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('IKMCs_Mice_available', '1', 'N', 40, 1, '', 'NOIDEA');

# Source 'IKMCs_No_products_available_yet::mus_musculus' (id = 97)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('IKMCs_No_products_available_yet', '1', 'N', 40, 1, '', 'NOIDEA');

# Source 'IKMCs_Vector_available::mus_musculus' (id = 98)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('IKMCs_Vector_available', '1', 'N', 40, 1, '', 'NOIDEA');

# Source 'IMGT/GENE_DB::homo_sapiens' (id = 99)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('IMGT/GENE_DB', '1', 'Y', 40, 1, '', 'NOIDEA');

# Source 'IMGT/GENE_DB::mus_musculus' (id = 100)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('IMGT/GENE_DB', '1', 'Y', 40, 1, '', 'NOIDEA');

# Source 'IMGT/LIGM_DB::homo_sapiens' (id = 101)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('IMGT/LIGM_DB', '1', 'N', 40, 1, '', 'NOIDEA');

# Source 'IMGT/LIGM_DB::mus_musculus' (id = 102)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('IMGT/LIGM_DB', '1', 'N', 40, 1, '', 'NOIDEA');

# Source 'ImmunoDB::aedes_aegypti' (id = 103)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('ImmunoDB', '1', 'Y', 40, 1, '', 'NOIDEA');

# Source 'ImmunoDB::anopheles_gambiae' (id = 104)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('ImmunoDB', '1', 'Y', 40, 1, '', 'NOIDEA');

# Source 'ImmunoDB::culex_quinquefasciatus' (id = 105)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('ImmunoDB', '1', 'Y', 40, 1, '', 'NOIDEA');

# Source 'InteractiveFly::drosophila_melanogaster' (id = 106)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('InteractiveFly', '1', 'N', 50, 1, '', 'NOIDEA');

# Source 'Interpro::CORE' (id = 107)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Interpro', '1', 'Y', 70, 1, '', 'NOIDEA');

# Source 'Interpro::MULTI' (id = 108)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Interpro', '1', 'Y', 70, 1, '', 'NOIDEA');

# Source 'InterproGO::MULTI' (id = 109)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('GO', '1', 'Y', 85, 2, 'interpro', 'NOIDEA');

# adding source dependency that source InterproGO::MULTI needs GO loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (109, 'GO');

# Source 'LRG_HGNC_notransfer' (id = 110)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('LRG_HGNC_notransfer', '1', 'N', 30, 5, '', 'NOIDEA');

# Source 'MEROPS::drosophila_grimshawi' (id = 111)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('MEROPS', '1', 'N', 50, 1, '', 'NOIDEA');

# Source 'MEROPS::drosophila_melanogaster' (id = 112)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('MEROPS', '1', 'N', 50, 1, '', 'NOIDEA');

# Source 'MEROPS::drosophila_virilis' (id = 113)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('MEROPS', '1', 'N', 50, 1, '', 'NOIDEA');

# Source 'MGI::mus_musculus#01' (id = 114)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('MGI', '1', 'Y', 30, 1, 'official', 'NOIDEA');

# Source 'MGI::mus_musculus#02' (id = 115)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('MGI', '1', 'Y', 45, 2, 'vega', 'NOIDEA');

# Source 'MGI::mus_musculus#03' (id = 116)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('MGI', '1', 'Y', 35, 3, 'ccds', 'NOIDEA');

# Source 'MGI::mus_musculus#04' (id = 117)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('MGI', '1', 'N', 30, 4, 'uniprot', 'NOIDEA');

# Source 'MGI::mus_musculus#05' (id = 118)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('MGI', '1', 'Y', 1, 10, 'descriptions', 'NOIDEA');

# Source 'MGI_automatic_transcript::mus_musculus' (id = 119)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('MGI_automatic_transcript_notransfer', '1', 'N', 49, 1, '', 'NOIDEA');

# Source 'MGI_curated_transcript::mus_musculus' (id = 120)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('MGI_curated_transcript_notransfer', '1', 'Y', 49, 1, '', 'NOIDEA');

# Source 'MGI_gene_name' (id = 121)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('MGI_gene_name', '1', 'N', 70, 1, '', 'NOIDEA');

# Source 'MGI_trans_name' (id = 122)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('MGI_trans_name', '1', 'N', 70, 1, '', 'NOIDEA');

# Source 'MIM2GENE::homo_sapiens' (id = 123)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('MIM2GENE', '1', 'Y', 60, 1, '', 'NOIDEA');

# adding source dependency that source MIM2GENE::homo_sapiens needs MIM loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (123, 'MIM');

# adding source dependency that source MIM2GENE::homo_sapiens needs EntrezGene loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (123, 'EntrezGene');

# Source 'MIM::homo_sapiens' (id = 124)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('MIM', '1', 'Y', 10, 1, '', 'NOIDEA');

# Source 'MIM_GENE::homo_sapiens' (id = 125)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('MIM_GENE', '1', 'N', 40, 1, '', 'NOIDEA');

# Source 'MIM_MORBID::homo_sapiens' (id = 126)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('MIM_MORBID', '1', 'N', 40, 1, '', 'NOIDEA');

# Source 'MitoDrome::drosophila_melanogaster' (id = 127)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('MitoDrome', '1', 'N', 50, 1, '', 'NOIDEA');

# Source 'NASC_GENE_ID::arabidopsis_thaliana' (id = 128)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('NASC_GENE_ID', '1', 'N', 1, 1, '', 'NOIDEA');

# Source 'OTTT::homo_sapiens' (id = 129)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('OTTT', '1', 'N', 50, 1, '', 'NOIDEA');

# Source 'OTTT::mus_musculus' (id = 130)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('OTTT', '1', 'N', 50, 1, '', 'NOIDEA');

# Source 'Orphanet::homo_sapiens' (id = 131)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Orphanet', '1', 'Y', 50, 1, '', 'NOIDEA');

# adding source dependency that source Orphanet::homo_sapiens needs HGNC loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (131, 'HGNC');

# Source 'Oxford_FGU_Oa_gene::ornithorhynchus_anatinus' (id = 132)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Oxford_FGU_Oa_gene', '1', 'Y', 40, 1, '', 'NOIDEA');

# Source 'Oxford_FGU_Oa_tscript::ornithorhynchus_anatinus' (id = 133)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Oxford_FGU_Oa_tscript', '1', 'Y', 40, 1, '', 'NOIDEA');

# Source 'PGD_GENE::MULTI' (id = 134)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('PGD_GENE', '1', 'N', 70, 1, '', 'NOIDEA');

# Source 'PGSC_GENE::solanum_tuberosum' (id = 135)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('PGSC_GENE', '1', 'Y', 10, 1, '', 'NOIDEA');

# Source 'PHIbase::MULTI' (id = 136)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('PHIbase', '1', 'Y', 50, 1, '', 'NOIDEA');

# adding source dependency that source PHIbase::MULTI needs Uniprot/SWISSPROT loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (136, 'Uniprot/SWISSPROT');

# adding source dependency that source PHIbase::MULTI needs Uniprot/SPTREMBL loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (136, 'Uniprot/SPTREMBL');

# Source 'PHYTOZOME_GMAX_GENE::glycine_max' (id = 137)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('PHYTOZOME_GMAX_GENE', '1', 'Y', 10, 1, '', 'NOIDEA');

# Source 'PIGGY::sus_scrofa' (id = 138)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('PIGGY', '1', 'Y', 30, 1, 'vega', 'NOIDEA');

# Source 'PIGGY_curated_transcript::sus_scrofa' (id = 139)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('PIGGY_curated_transcript_notransfer', '1', 'Y', 49, 1, 'curated with transcript number postfix', 'NOIDEA');

# Source 'PIGGY_gene_name' (id = 140)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('PIGGY_gene_name', '1', 'N', 49, 1, '', 'NOIDEA');

# Source 'PIGGY_trans_name' (id = 141)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('PIGGY_trans_name', '1', 'N', 49, 1, '', 'NOIDEA');

# Source 'PO_GROW::arabidopsis_thaliana' (id = 142)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('PO', '1', 'Y', 85, 1, 'main', 'NOIDEA');

# adding source dependency that source PO_GROW::arabidopsis_thaliana needs TAIR_TRANSLATION loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (142, 'TAIR_TRANSLATION');

# Source 'PO_STRU::arabidopsis_thaliana' (id = 143)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('PO', '1', 'Y', 85, 1, 'main', 'NOIDEA');

# adding source dependency that source PO_STRU::arabidopsis_thaliana needs TAIR_TRANSLATION loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (143, 'TAIR_TRANSLATION');

# Source 'PUBMED::aedes_aegypti' (id = 144)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('PUBMED', '1', 'Y', 40, 1, '', 'NOIDEA');

# Source 'PUBMED::anopheles_gambiae' (id = 145)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('PUBMED', '1', 'Y', 40, 1, '', 'NOIDEA');

# Source 'PUBMED::culex_quinquefasciatus' (id = 146)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('PUBMED', '1', 'Y', 40, 1, '', 'NOIDEA');

# Source 'PUBMED::ixodes_scapularis' (id = 147)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('PUBMED', '1', 'Y', 40, 1, '', 'NOIDEA');

# Source 'Platypus_olfactory_receptor::ornithorhynchus_anatinus' (id = 148)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Platypus_olfactory_receptor', '1', 'Y', 40, 1, '', 'NOIDEA');

# Source 'PomBase::schizosaccharomyces_pombe' (id = 149)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('PomBase', '1', 'Y', 10, 1, '', 'NOIDEA');

# Source 'PomBase_GENE::schizosaccharomyces_pombe' (id = 150)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('PomBase_GENE', '1', 'N', 10, 1, '', 'NOIDEA');

# Source 'PomBase_TRANSCRIPT::schizosaccharomyces_pombe' (id = 151)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('PomBase_TRANSCRIPT', '1', 'N', 10, 1, '', 'NOIDEA');

# Source 'RFAM::EG' (id = 152)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RFAM', '1', 'Y', 70, 1, '', 'NOIDEA');

# Source 'RFAM::MULTI' (id = 153)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RFAM', '1', 'Y', 70, 1, '', 'NOIDEA');

# Source 'RFAM_gene_name' (id = 154)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RFAM_gene_name', '1', 'N', 70, 1, '', 'NOIDEA');

# Source 'RFAM_trans_name' (id = 155)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RFAM_trans_name', '1', 'N', 70, 1, '', 'NOIDEA');

# Source 'RGD::rattus_norvegicus' (id = 156)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RGD', '1', 'Y', 30, 2, '', 'NOIDEA');

# adding source dependency that source RGD::rattus_norvegicus needs RefSeq_dna loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (156, 'RefSeq_dna');

# adding source dependency that source RGD::rattus_norvegicus needs RefSeq_peptide loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (156, 'RefSeq_peptide');

# Source 'RGD::rattus_norvegicus#01' (id = 157)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RGD', '1', 'Y', 30, 3, 'vega', 'NOIDEA');

# Source 'RGD::rattus_norvegicus#02' (id = 158)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RGD', '1', 'N', 30, 1, 'direct_xref', 'NOIDEA');

# Source 'RGD_curated_transcript::mus_musculus' (id = 159)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RGD_curated_transcript_notransfer', '1', 'Y', 49, 1, '', 'NOIDEA');

# Source 'RGD_trans_name' (id = 160)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RGD_trans_name', '1', 'N', 49, 1, '', 'NOIDEA');

# Source 'RNACentral::MULTI' (id = 161)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RNACentral', '1', 'Y', 1, 1, '', 'NOIDEA');

# Source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RNAMMER', '1', 'Y', 70, 1, '', 'NOIDEA');

# Source 'Reactome::MULTI' (id = 163)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Reactome', '1', 'Y', 80, 1, 'direct', 'NOIDEA');

# Source 'Reactome::MULTI-Uniprot' (id = 164)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Reactome', '1', 'N', 20, 1, 'uniprot', 'NOIDEA');

# Source 'Reactome_gene::MULTI' (id = 165)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Reactome_gene', '1', 'N', 20, 1, 'gene', 'NOIDEA');

# Source 'Reactome_transcript::MULTI' (id = 166)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Reactome_transcript', '1', 'N', 20, 1, 'transcript', 'NOIDEA');

# Source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::MULTI-complete' (id = 168)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::MULTI-fungi' (id = 169)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 1, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::MULTI-predicted' (id = 171)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna_predicted', '1', 'N', 20, 1, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::MULTI-protozoa' (id = 172)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::MULTI-vertebrate_mammalian' (id = 173)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::MULTI-vertebrate_other' (id = 174)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::ailuropoda_melanoleuca' (id = 175)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::amphimedon_queenslandica' (id = 176)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::anas_platyrhynchos' (id = 177)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::anolis_carolinensis' (id = 178)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::apis_mellifera' (id = 179)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::astyanax_mexicanus' (id = 180)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::bos_taurus' (id = 181)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::caenorhabditis_elegans' (id = 182)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::callithrix_jacchus' (id = 183)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::canis_familiaris' (id = 184)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::cavia_porcellus' (id = 185)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::ciona_intestinalis' (id = 186)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::danio_rerio' (id = 187)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::dasypus_novemcinctus' (id = 188)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::drosophila_melanogaster' (id = 189)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::echinops_telfairi' (id = 190)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::equus_caballus' (id = 191)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::felis_catus' (id = 192)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::gallus_gallus' (id = 193)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::gorilla_gorilla' (id = 194)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::homo_sapiens' (id = 195)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 25, 2, 'ccds', 'NOIDEA');

# Source 'RefSeq_dna::homo_sapiens#01' (id = 196)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::ictidomys_tridecemlineatus' (id = 197)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::lepisosteus_oculatus' (id = 198)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::loxodonta_africana' (id = 199)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::macaca_mulatta' (id = 200)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::meleagris_gallopavo' (id = 201)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::monodelphis_domestica' (id = 202)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::mus_musculus' (id = 203)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 25, 2, 'ccds', 'NOIDEA');

# Source 'RefSeq_dna::mus_musculus#01' (id = 204)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::mustela_putorius_furo' (id = 205)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::ochotona_princeps' (id = 206)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::oreochromis_niloticus' (id = 207)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::ornithorhynchus_anatinus' (id = 208)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::oryctolagus_cuniculus' (id = 209)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::oryzias_latipes' (id = 210)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::otolemur_garnettii' (id = 211)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::ovis_aries' (id = 212)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::pan_troglodytes' (id = 213)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::papio_anubis' (id = 214)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::pongo_abelii' (id = 215)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::rattus_norvegicus' (id = 216)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::sorex_araneus' (id = 217)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::strongylocentrotus_purpuratus' (id = 218)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::sus_scrofa' (id = 219)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::taeniopygia_guttata' (id = 220)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::takifugu_rubripes' (id = 221)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::tribolium_castaneum' (id = 222)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_dna::tursiops_truncatus' (id = 223)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_dna', '1', 'Y', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_import', '1', 'Y', 20, 1, 'otherfeatures', 'NOIDEA');

# Source 'RefSeq_mRNA::CCDS' (id = 225)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_mRNA', '1', 'N', 20, 2, 'ccds', 'NOIDEA');

# Source 'RefSeq_mRNA::MULTI' (id = 226)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_mRNA', '1', 'N', 20, 3, 'refseq', 'NOIDEA');

# Source 'RefSeq_mRNA::otherfeatures' (id = 227)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_mRNA', '1', 'N', 20, 1, 'otherfeatures', 'NOIDEA');

# Source 'RefSeq_mRNA_predicted::CCDS' (id = 228)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_mRNA_predicted', '1', 'N', 20, 1, 'ccds', 'NOIDEA');

# Source 'RefSeq_mRNA_predicted::MULTI' (id = 229)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_mRNA_predicted', '1', 'N', 20, 2, 'refseq', 'NOIDEA');

# Source 'RefSeq_mRNA_predicted::otherfeatures' (id = 230)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_mRNA_predicted', '1', 'N', 20, 1, 'otherfeatures', 'NOIDEA');

# Source 'RefSeq_ncRNA::MULTI' (id = 231)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_ncRNA', '1', 'N', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_ncRNA::otherfeatures' (id = 232)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_ncRNA', '1', 'N', 20, 1, 'otherfeatures', 'NOIDEA');

# Source 'RefSeq_ncRNA_predicted::MULTI' (id = 233)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_ncRNA_predicted', '1', 'N', 20, 1, '', 'NOIDEA');

# Source 'RefSeq_ncRNA_predicted::otherfeatures' (id = 234)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_ncRNA_predicted', '1', 'N', 20, 1, 'otherfeatures', 'NOIDEA');

# Source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::MULTI-complete' (id = 236)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::MULTI-fungi' (id = 237)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::MULTI-predicted' (id = 239)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide_predicted', '1', 'N', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::MULTI-protozoa' (id = 240)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::MULTI-vertebrate_mammalian' (id = 241)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::MULTI-vertebrate_other' (id = 242)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::ailuropoda_melanoleuca' (id = 243)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::amphimedon_queenslandica' (id = 244)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::anas_platyrhynchos' (id = 245)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::anolis_carolinensis' (id = 246)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::apis_mellifera' (id = 247)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::astyanax_mexicanus' (id = 248)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::bos_taurus' (id = 249)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::caenorhabditis_elegans' (id = 250)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::callithrix_jacchus' (id = 251)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::canis_familiaris' (id = 252)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::cavia_porcellus' (id = 253)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::ciona_intestinalis' (id = 254)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::danio_rerio' (id = 255)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::dasypus_novemcinctus' (id = 256)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::drosophila_melanogaster' (id = 257)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::echinops_telfairi' (id = 258)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::equus_caballus' (id = 259)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::felis_catus' (id = 260)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::gallus_gallus' (id = 261)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::gorilla_gorilla' (id = 262)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::homo_sapiens' (id = 263)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::ictidomys_tridecemlineatus' (id = 264)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::lepisosteus_oculatus' (id = 265)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::loxodonta_africana' (id = 266)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::macaca_mulatta' (id = 267)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::meleagris_gallopavo' (id = 268)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::monodelphis_domestica' (id = 269)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::mus_musculus' (id = 270)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::mustela_putorius_furo' (id = 271)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::ochotona_princeps' (id = 272)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::oreochromis_niloticus' (id = 273)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::ornithorhynchus_anatinus' (id = 274)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::oryctolagus_cuniculus' (id = 275)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::oryzias_latipes' (id = 276)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::otherfeatures' (id = 277)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'N', 20, 1, 'otherfeatures', 'NOIDEA');

# Source 'RefSeq_peptide::otolemur_garnettii' (id = 278)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::ovis_aries' (id = 279)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::pan_troglodytes' (id = 280)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::papio_anubis' (id = 281)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::pongo_abelii' (id = 282)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::rattus_norvegicus' (id = 283)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::sorex_araneus' (id = 284)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::strongylocentrotus_purpuratus' (id = 285)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::sus_scrofa' (id = 286)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::taeniopygia_guttata' (id = 287)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::takifugu_rubripes' (id = 288)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::tribolium_castaneum' (id = 289)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide::tursiops_truncatus' (id = 290)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide', '1', 'Y', 20, 2, '', 'NOIDEA');

# Source 'RefSeq_peptide_predicted::otherfeatures' (id = 291)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('RefSeq_peptide_predicted', '1', 'N', 20, 1, 'otherfeatures', 'NOIDEA');

# Source 'Ribosomal_Protein_Gene_DB::anopheles_gambiae' (id = 292)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Ribosomal_Protein_Gene_DB', '1', 'Y', 40, 1, '', 'NOIDEA');

# Source 'SCHISTODB::MULTI' (id = 293)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('SCHISTODB', '1', 'N', 70, 1, '', 'NOIDEA');

# Source 'SGD::saccharomyces_cerevisiae' (id = 294)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('SGD', '1', 'Y', 10, 1, '', 'NOIDEA');

# Source 'SGD_GENE::saccharomyces_cerevisiae' (id = 295)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('SGD_GENE', '1', 'N', 10, 1, '', 'NOIDEA');

# Source 'SGD_TRANSCRIPT::saccharomyces_cerevisiae' (id = 296)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('SGD_TRANSCRIPT', '1', 'N', 10, 1, '', 'NOIDEA');

# Source 'SGD_TRANSLATION::saccharomyces_cerevisiae' (id = 297)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('SGD_TRANSLATION', '1', 'N', 10, 1, '', 'NOIDEA');

# Source 'TAIR_LOCUS::arabidopsis_thaliana' (id = 298)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('TAIR_LOCUS', '1', 'Y', 1, 1, '', 'NOIDEA');

# Source 'TAIR_LOCUS_MODEL::arabidopsis_thaliana' (id = 299)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('TAIR_LOCUS_MODEL', '1', 'N', 1, 1, '', 'NOIDEA');

# Source 'TAIR_SYMBOL::arabidopsis_thaliana' (id = 300)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('TAIR_SYMBOL', '1', 'N', 1, 1, '', 'NOIDEA');

# Source 'TAIR_TRANSLATION::arabidopsis_thaliana' (id = 301)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('TAIR_TRANSLATION', '1', 'N', 1, 1, '', 'NOIDEA');

# Source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('TRNASCAN_SE', '1', 'Y', 70, 1, '', 'NOIDEA');

# Source 'TransFac::drosophila_melanogaster' (id = 303)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('TransFac', '1', 'N', 50, 1, '', 'NOIDEA');

# Source 'TransFac::drosophila_pseudoobscura' (id = 304)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('TransFac', '1', 'N', 50, 1, '', 'NOIDEA');

# Source 'UCSC::MULTI' (id = 305)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UCSC', '1', 'Y', 70, 1, '', 'NOIDEA');

# Source 'UCSC::homo_sapiens' (id = 306)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UCSC_hg38', '1', 'Y', 70, 1, '', 'NOIDEA');

# Source 'UCSC::mus_musculus' (id = 307)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UCSC_mm10', '1', 'Y', 70, 1, '', 'NOIDEA');

# Source 'UniGene::acyrthosiphon_pisum' (id = 308)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::aedes_aegypti' (id = 309)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::anolis_carolinensis' (id = 310)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::anopheles_gambiae' (id = 311)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::apis_mellifera' (id = 312)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::arabidopsis_thaliana' (id = 313)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::bombyx_mori' (id = 314)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::bos_taurus' (id = 315)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::brachypodium_distachyon' (id = 316)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::brassica_rapa' (id = 317)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::caenorhabditis_elegans' (id = 318)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::canis_familiaris' (id = 319)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::chlamydomonas_reinhardtii' (id = 320)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::ciona_intestinalis' (id = 321)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::ciona_savignyi' (id = 322)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::culex_quinquefasciatus' (id = 323)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::danio_rerio' (id = 324)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::equus_caballus' (id = 325)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::gadus_morhua' (id = 326)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::gallus_gallus' (id = 327)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::gasterosteus_aculeatus' (id = 328)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::glycine_max' (id = 329)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::homo_sapiens' (id = 330)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::hordeum_vulgare' (id = 331)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::ixodes_scapularis' (id = 332)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::macaca_mulatta' (id = 333)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::meleagris_gallopavo' (id = 334)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::monodelphis_domestica' (id = 335)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::mus_musculus' (id = 336)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::nematostella_vectensis' (id = 337)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::oreochromis_niloticus' (id = 338)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::ornithorhynchus_anatinus' (id = 339)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::oryctolagus_cuniculus' (id = 340)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::oryzias_latipes' (id = 341)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::ovis_aries' (id = 342)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::pan_troglodytes' (id = 343)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::papio_anubis' (id = 344)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::petromyzon_marinus' (id = 345)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::physcomitrella_patens' (id = 346)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::pongo_abelii' (id = 347)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::populus_trichocarpa' (id = 348)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::rattus_norvegicus' (id = 349)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::selaginella_moellendorffii' (id = 350)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::solanum_lycopersicum' (id = 351)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::solanum_tuberosum' (id = 352)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::sorghum_bicolor' (id = 353)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::strongylocentrotus_purpuratus' (id = 354)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::sus_scrofa' (id = 355)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::taeniopygia_guttata' (id = 356)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::takifugu_rubripes' (id = 357)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::tribolium_castaneum' (id = 358)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::vitis_vinifera' (id = 359)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::xenopus_tropicalis' (id = 360)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniGene::zea_mays' (id = 361)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniGene', '1', 'Y', 30, 1, '', 'NOIDEA');

# Source 'UniParc::MULTI' (id = 362)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('UniParc', '1', 'Y', 20, 1, '', 'NOIDEA');

# Source 'UniProt::PDB' (id = 363)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('PDB', '1', 'N', 20, 1, '', 'NOIDEA');

# Source 'UniProt::protein_id' (id = 364)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('protein_id', '1', 'N', 20, 1, '', 'NOIDEA');

# Source 'UniProt::protein_id-predicted' (id = 365)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('protein_id_predicted', '1', 'N', 20, 1, '', 'NOIDEA');

# Source 'Uniprot/SPTREMBL::DIRECT' (id = 366)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Uniprot/SPTREMBL', '1', 'N', 22, 1, 'direct', 'NOIDEA');

# Source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Uniprot/SPTREMBL', '1', 'Y', 20, 3, 'sequence_mapped', 'NOIDEA');

# adding source dependency that source Uniprot/SPTREMBL::MULTI needs MIM loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (367, 'MIM');

# Source 'Uniprot/SPTREMBL::MULTI-evidence_gt_2' (id = 368)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Uniprot/SPTREMBL', '1', 'N', 20, 10, 'protein_evidence_gt_2', 'LOWEVIDENCE');

# Source 'Uniprot/SPTREMBL::MULTI-invertebrate' (id = 369)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Uniprot/SPTREMBL', '1', 'Y', 20, 3, '', 'NOIDEA');

# adding source dependency that source Uniprot/SPTREMBL::MULTI-invertebrate needs MIM loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (369, 'MIM');

# Source 'Uniprot/SPTREMBL::MULTI-predicted' (id = 370)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Uniprot/SPTREMBL_predicted', '1', 'N', 20, 1, '', 'NOIDEA');

# Source 'Uniprot/SPTREMBL::drosophila_melanogaster' (id = 371)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Uniprot/SPTREMBL', '1', 'Y', 20, 1, '', 'NOIDEA');

# Source 'Uniprot/SWISSPROT::DIRECT' (id = 372)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Uniprot/SWISSPROT', '1', 'N', 22, 1, 'direct', 'NOIDEA');

# Source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Uniprot/SWISSPROT', '1', 'Y', 20, 3, 'sequence_mapped', 'NOIDEA');

# adding source dependency that source Uniprot/SWISSPROT::MULTI needs MIM loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (373, 'MIM');

# Source 'Uniprot/SWISSPROT::MULTI-invertebrate' (id = 374)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Uniprot/SWISSPROT', '1', 'Y', 20, 3, 'sequence_mapped', 'NOIDEA');

# adding source dependency that source Uniprot/SWISSPROT::MULTI-invertebrate needs MIM loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (374, 'MIM');

# Source 'Uniprot/SWISSPROT::MULTI-predicted' (id = 375)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Uniprot/SWISSPROT_predicted', '1', 'N', 20, 1, '', 'NOIDEA');

# Source 'Uniprot::ChEMBL' (id = 376)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('ChEMBL', '1', 'N', 20, 1, '', 'NOIDEA');

# Source 'Uniprot::EMBL' (id = 377)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('EMBL', '1', 'N', 20, 1, '', 'NOIDEA');

# Source 'Uniprot::EMBL-predicted' (id = 378)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('EMBL_predicted', '1', 'N', 20, 1, '', 'NOIDEA');

# Source 'Uniprot_gn' (id = 379)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Uniprot_gn', '1', 'N', 20, 1, '', 'NOIDEA');

# Source 'Uniprot_gn_gene_name' (id = 380)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Uniprot_gn_gene_name', '1', 'N', 70, 1, '', 'NOIDEA');

# Source 'Uniprot_gn_trans_name' (id = 381)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Uniprot_gn_trans_name', '1', 'N', 70, 1, '', 'NOIDEA');

# Source 'VB_Community_Annotation::aedes_aegypti' (id = 382)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('VB_Community_Annotation', '1', 'Y', 40, 1, '', 'NOIDEA');

# Source 'VB_Community_Annotation::anopheles_gambiae' (id = 383)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('VB_Community_Annotation', '1', 'Y', 40, 1, '', 'NOIDEA');

# Source 'VB_Community_Annotation::culex_quinquefasciatus' (id = 384)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('VB_Community_Annotation', '1', 'Y', 40, 1, '', 'NOIDEA');

# Source 'VB_Community_Annotation::ixodes_scapularis' (id = 385)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('VB_Community_Annotation', '1', 'Y', 40, 1, '', 'NOIDEA');

# Source 'VB_Community_Annotation_Synonym::aedes_aegypti' (id = 386)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('VB_Community_Annotation_Synonym', '1', 'Y', 40, 1, '', 'NOIDEA');

# adding source dependency that source VB_Community_Annotation_Synonym::aedes_aegypti needs VB_Community_Annotation loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (386, 'VB_Community_Annotation');

# Source 'VB_Community_Annotation_Synonym::anopheles_gambiae' (id = 387)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('VB_Community_Annotation_Synonym', '1', 'Y', 40, 1, '', 'NOIDEA');

# adding source dependency that source VB_Community_Annotation_Synonym::anopheles_gambiae needs VB_Community_Annotation loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (387, 'VB_Community_Annotation');

# Source 'VB_Community_Annotation_Synonym::culex_quinquefasciatus' (id = 388)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('VB_Community_Annotation_Synonym', '1', 'Y', 40, 1, '', 'NOIDEA');

# adding source dependency that source VB_Community_Annotation_Synonym::culex_quinquefasciatus needs VB_Community_Annotation loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (388, 'VB_Community_Annotation');

# Source 'VB_External_Description::aedes_aegypti' (id = 389)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('VB_External_Description', '1', 'Y', 40, 1, '', 'NOIDEA');

# Source 'VB_External_Description::culex_quinquefasciatus' (id = 390)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('VB_External_Description', '1', 'Y', 40, 1, '', 'NOIDEA');

# Source 'VB_External_Description::ixodes_scapularis' (id = 391)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('VB_External_Description', '1', 'Y', 40, 1, '', 'NOIDEA');

# Source 'VB_RNA_Description::aedes_aegypti' (id = 392)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('VB_RNA_Description', '1', 'Y', 40, 1, '', 'NOIDEA');

# Source 'VB_RNA_Description::anopheles_gambiae' (id = 393)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('VB_RNA_Description', '1', 'Y', 40, 1, '', 'NOIDEA');

# Source 'VB_RNA_Description::culex_quinquefasciatus' (id = 394)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('VB_RNA_Description', '1', 'Y', 40, 1, '', 'NOIDEA');

# Source 'WikiGene::MULTI' (id = 395)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('WikiGene', '1', 'N', 100, 1, '', 'NOIDEA');

# Source 'Xenopus_Jamboree::xenopus_tropicalis' (id = 396)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('Xenbase', '1', 'Y', 20, 1, '', 'NOIDEA');

# Source 'ZFIN_ID::danio_rerio#01' (id = 397)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('ZFIN_ID', '1', 'Y', 31, 1, 'uniprot/refseq', 'NOIDEA');

# Source 'ZFIN_ID::danio_rerio#02' (id = 398)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('ZFIN_ID', '1', 'Y', 30, 2, 'vega', 'NOIDEA');

# Source 'ZFIN_ID::danio_rerio#03' (id = 399)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('ZFIN_ID', '1', 'Y', 1, 10, 'description_only', 'NOIDEA');

# Source 'ZFIN_ID_curated_transcript::danio_rerio' (id = 400)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('ZFIN_ID_curated_transcript_notransfer', '1', 'Y', 49, 1, 'curated with transcript number postfix', 'NOIDEA');

# Source 'ZFIN_ID_gene_name' (id = 401)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('ZFIN_ID_gene_name', '1', 'N', 49, 1, '', 'NOIDEA');

# Source 'ZFIN_ID_trans_name' (id = 402)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('ZFIN_ID_trans_name', '1', 'N', 49, 1, '', 'NOIDEA');

# Source 'cint_aniseed_v1::ciona_intestinalis' (id = 403)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('cint_aniseed_v1', '1', 'Y', 50, 1, '', 'NOIDEA');

# Source 'cint_jgi_v1::ciona_intestinalis' (id = 404)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('cint_jgi_v1', '1', 'Y', 50, 1, '', 'NOIDEA');

# Source 'flybase_annotation_id::drosophila_melanogaster' (id = 405)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('flybase_annotation_id', '1', 'N', 50, 1, 'Annotation ID assigned by FlyBase', 'NOIDEA');

# Source 'flybase_gene_id::drosophila_ananassae' (id = 406)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('flybase_gene_id', '1', 'Y', 100, 1, 'ID assigned by FlyBase', 'NOIDEA');

# adding source dependency that source flybase_gene_id::drosophila_ananassae needs Uniprot/SPTREMBL loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (406, 'Uniprot/SPTREMBL');

# adding source dependency that source flybase_gene_id::drosophila_ananassae needs Uniprot/SWISSPROT loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (406, 'Uniprot/SWISSPROT');

# adding source dependency that source flybase_gene_id::drosophila_ananassae needs Interpro loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (406, 'Interpro');

# Source 'flybase_gene_id::drosophila_erecta' (id = 407)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('flybase_gene_id', '1', 'Y', 100, 1, 'ID assigned by FlyBase', 'NOIDEA');

# adding source dependency that source flybase_gene_id::drosophila_erecta needs Uniprot/SPTREMBL loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (407, 'Uniprot/SPTREMBL');

# adding source dependency that source flybase_gene_id::drosophila_erecta needs Uniprot/SWISSPROT loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (407, 'Uniprot/SWISSPROT');

# adding source dependency that source flybase_gene_id::drosophila_erecta needs Interpro loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (407, 'Interpro');

# Source 'flybase_gene_id::drosophila_grimshawi' (id = 408)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('flybase_gene_id', '1', 'Y', 100, 1, 'ID assigned by FlyBase', 'NOIDEA');

# adding source dependency that source flybase_gene_id::drosophila_grimshawi needs Uniprot/SPTREMBL loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (408, 'Uniprot/SPTREMBL');

# adding source dependency that source flybase_gene_id::drosophila_grimshawi needs Uniprot/SWISSPROT loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (408, 'Uniprot/SWISSPROT');

# adding source dependency that source flybase_gene_id::drosophila_grimshawi needs Interpro loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (408, 'Interpro');

# Source 'flybase_gene_id::drosophila_melanogaster' (id = 409)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('flybase_gene_id', '1', 'Y', 75, 1, 'ID assigned by FlyBase', 'NOIDEA');

# adding source dependency that source flybase_gene_id::drosophila_melanogaster needs Uniprot/SPTREMBL loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (409, 'Uniprot/SPTREMBL');

# adding source dependency that source flybase_gene_id::drosophila_melanogaster needs Uniprot/SWISSPROT loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (409, 'Uniprot/SWISSPROT');

# adding source dependency that source flybase_gene_id::drosophila_melanogaster needs Interpro loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (409, 'Interpro');

# Source 'flybase_gene_id::drosophila_mojavensis' (id = 410)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('flybase_gene_id', '1', 'Y', 100, 1, 'ID assigned by FlyBase', 'NOIDEA');

# adding source dependency that source flybase_gene_id::drosophila_mojavensis needs Uniprot/SPTREMBL loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (410, 'Uniprot/SPTREMBL');

# adding source dependency that source flybase_gene_id::drosophila_mojavensis needs Uniprot/SWISSPROT loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (410, 'Uniprot/SWISSPROT');

# adding source dependency that source flybase_gene_id::drosophila_mojavensis needs Interpro loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (410, 'Interpro');

# Source 'flybase_gene_id::drosophila_persimilis' (id = 411)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('flybase_gene_id', '1', 'Y', 100, 1, 'ID assigned by FlyBase', 'NOIDEA');

# adding source dependency that source flybase_gene_id::drosophila_persimilis needs Uniprot/SPTREMBL loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (411, 'Uniprot/SPTREMBL');

# adding source dependency that source flybase_gene_id::drosophila_persimilis needs Uniprot/SWISSPROT loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (411, 'Uniprot/SWISSPROT');

# adding source dependency that source flybase_gene_id::drosophila_persimilis needs Interpro loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (411, 'Interpro');

# Source 'flybase_gene_id::drosophila_pseudoobscura' (id = 412)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('flybase_gene_id', '1', 'Y', 100, 1, 'ID assigned by FlyBase', 'NOIDEA');

# adding source dependency that source flybase_gene_id::drosophila_pseudoobscura needs Uniprot/SPTREMBL loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (412, 'Uniprot/SPTREMBL');

# adding source dependency that source flybase_gene_id::drosophila_pseudoobscura needs Uniprot/SWISSPROT loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (412, 'Uniprot/SWISSPROT');

# adding source dependency that source flybase_gene_id::drosophila_pseudoobscura needs Interpro loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (412, 'Interpro');

# Source 'flybase_gene_id::drosophila_sechellia' (id = 413)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('flybase_gene_id', '1', 'Y', 100, 1, 'ID assigned by FlyBase', 'NOIDEA');

# adding source dependency that source flybase_gene_id::drosophila_sechellia needs Uniprot/SPTREMBL loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (413, 'Uniprot/SPTREMBL');

# adding source dependency that source flybase_gene_id::drosophila_sechellia needs Uniprot/SWISSPROT loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (413, 'Uniprot/SWISSPROT');

# adding source dependency that source flybase_gene_id::drosophila_sechellia needs Interpro loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (413, 'Interpro');

# Source 'flybase_gene_id::drosophila_simulans' (id = 414)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('flybase_gene_id', '1', 'Y', 100, 1, 'ID assigned by FlyBase', 'NOIDEA');

# adding source dependency that source flybase_gene_id::drosophila_simulans needs Uniprot/SPTREMBL loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (414, 'Uniprot/SPTREMBL');

# adding source dependency that source flybase_gene_id::drosophila_simulans needs Uniprot/SWISSPROT loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (414, 'Uniprot/SWISSPROT');

# adding source dependency that source flybase_gene_id::drosophila_simulans needs Interpro loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (414, 'Interpro');

# Source 'flybase_gene_id::drosophila_virilis' (id = 415)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('flybase_gene_id', '1', 'Y', 100, 1, 'ID assigned by FlyBase', 'NOIDEA');

# adding source dependency that source flybase_gene_id::drosophila_virilis needs Uniprot/SPTREMBL loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (415, 'Uniprot/SPTREMBL');

# adding source dependency that source flybase_gene_id::drosophila_virilis needs Uniprot/SWISSPROT loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (415, 'Uniprot/SWISSPROT');

# adding source dependency that source flybase_gene_id::drosophila_virilis needs Interpro loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (415, 'Interpro');

# Source 'flybase_gene_id::drosophila_willistoni' (id = 416)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('flybase_gene_id', '1', 'Y', 100, 1, 'ID assigned by FlyBase', 'NOIDEA');

# adding source dependency that source flybase_gene_id::drosophila_willistoni needs Uniprot/SPTREMBL loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (416, 'Uniprot/SPTREMBL');

# adding source dependency that source flybase_gene_id::drosophila_willistoni needs Uniprot/SWISSPROT loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (416, 'Uniprot/SWISSPROT');

# adding source dependency that source flybase_gene_id::drosophila_willistoni needs Interpro loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (416, 'Interpro');

# Source 'flybase_gene_id::drosophila_yakuba' (id = 417)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('flybase_gene_id', '1', 'Y', 100, 1, 'ID assigned by FlyBase', 'NOIDEA');

# adding source dependency that source flybase_gene_id::drosophila_yakuba needs Uniprot/SPTREMBL loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (417, 'Uniprot/SPTREMBL');

# adding source dependency that source flybase_gene_id::drosophila_yakuba needs Uniprot/SWISSPROT loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (417, 'Uniprot/SWISSPROT');

# adding source dependency that source flybase_gene_id::drosophila_yakuba needs Interpro loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (417, 'Interpro');

# Source 'flybase_transcript_id::drosophila_melanogaster' (id = 418)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('flybase_transcript_id', '1', 'N', 50, 1, 'ID assigned by FlyBase', 'NOIDEA');

# Source 'flybase_translation_id::drosophila_melanogaster' (id = 419)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('flybase_translation_id', '1', 'N', 50, 1, 'ID assigned by FlyBase', 'NOIDEA');

# Source 'goslim_goa::EG' (id = 420)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('goslim_goa', '1', 'Y', 120, 1, 'main', 'NOIDEA');

# adding source dependency that source goslim_goa::EG needs GO loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (420, 'GO');

# Source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('goslim_goa', '1', 'Y', 120, 1, 'main', 'NOIDEA');

# adding source dependency that source goslim_goa::MULTI needs GO loaded first
INSERT IGNORE INTO dependent_source (master_source_id, dependent_name)
VALUES (421, 'GO');

# Source 'miRBase::MULTI' (id = 422)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('miRBase', '1', 'Y', 70, 1, '', 'NOIDEA');

# Source 'miRBase::drosophila_melanogaster' (id = 423)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('miRBase', '1', 'N', 50, 1, '', 'NOIDEA');

# Source 'miRBase::drosophila_pseudoobscura' (id = 424)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('miRBase', '1', 'N', 50, 1, '', 'NOIDEA');

# Source 'miRBase_gene_name' (id = 425)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('miRBase_gene_name', '1', 'N', 70, 1, '', 'NOIDEA');

# Source 'miRBase_trans_name' (id = 426)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('miRBase_trans_name', '1', 'N', 70, 1, '', 'NOIDEA');

# Source 'misc_EG::EG' (id = 427)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('misc_EG', '1', 'Y', 70, 1, '', 'NOIDEA');

# Source 'ncRNA_EG::EG' (id = 428)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('ncRNA_EG', '1', 'Y', 70, 1, '', 'NOIDEA');

# Source 'phatr_jgi_v2::MULTI' (id = 429)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('phatr_jgi_v2', '1', 'N', 70, 1, '', 'NOIDEA');

# Source 'phatr_jgi_v2_bd::MULTI' (id = 430)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('phatr_jgi_v2_bd', '1', 'N', 70, 1, '', 'NOIDEA');

# Source 'phyra_jgi_v1.1::MULTI' (id = 431)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('phyra_jgi_v1.1', '1', 'N', 70, 1, '', 'NOIDEA');

# Source 'phyra_jgi_v1.1_gene::MULTI' (id = 432)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('phyra_jgi_v1.1_gene', '1', 'N', 70, 1, '', 'NOIDEA');

# Source 'phyra_jgi_v1.1_transcript::MULTI' (id = 433)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('phyra_jgi_v1.1_transcript', '1', 'N', 70, 1, '', 'NOIDEA');

# Source 'physo1_jgi_v1.1::MULTI' (id = 434)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('physo1_jgi_v1.1', '1', 'N', 70, 1, '', 'NOIDEA');

# Source 'physo1_jgi_v1.1_gene::MULTI' (id = 435)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('physo1_jgi_v1.1_gene', '1', 'N', 70, 1, '', 'NOIDEA');

# Source 'physo1_jgi_v1.1_transcript::MULTI' (id = 436)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('physo1_jgi_v1.1_transcript', '1', 'N', 70, 1, '', 'NOIDEA');

# Source 'thaps_jgi_v2::MULTI' (id = 437)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('thaps_jgi_v2', '1', 'N', 70, 1, '', 'NOIDEA');

# Source 'thaps_jgi_v2_bd::MULTI' (id = 438)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('thaps_jgi_v2_bd', '1', 'N', 70, 1, '', 'NOIDEA');

# Source 'triad_jgi_v1.0::MULTI' (id = 439)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('triad_jgi_v1.0', '1', 'N', 70, 1, '', 'NOIDEA');

# Source 'wormbase::bmalayi' (id = 440)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('wormbase_all', '1', 'Y', 50, 1, '', 'NOIDEA');

# Source 'wormbase::cbrenneri' (id = 441)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('wormbase_all', '1', 'Y', 50, 1, '', 'NOIDEA');

# Source 'wormbase::cbriggsae' (id = 442)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('wormbase_all', '1', 'Y', 50, 1, '', 'NOIDEA');

# Source 'wormbase::celegans' (id = 443)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('wormbase_all', '1', 'Y', 50, 1, '', 'NOIDEA');

# Source 'wormbase::cjaponica' (id = 444)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('wormbase_all', '1', 'Y', 50, 1, '', 'NOIDEA');

# Source 'wormbase::cremanei' (id = 445)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('wormbase_all', '1', 'Y', 50, 1, '', 'NOIDEA');

# Source 'wormbase::ovolvulus' (id = 446)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('wormbase_all', '1', 'Y', 50, 1, '', 'NOIDEA');

# Source 'wormbase::ppacificus' (id = 447)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('wormbase_all', '1', 'Y', 50, 1, '', 'NOIDEA');

# Source 'wormbase::sratti' (id = 448)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('wormbase_all', '1', 'Y', 50, 1, '', 'NOIDEA');

# Source 'wormbase_gene::wormbase' (id = 449)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('wormbase_gene', '1', 'N', 50, 1, '', 'NOIDEA');

# Source 'wormbase_gseqname::wormbase' (id = 450)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('wormbase_gseqname', '1', 'N', 50, 1, '', 'NOIDEA');

# Source 'wormbase_locus::wormbase' (id = 451)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('wormbase_locus', '1', 'N', 50, 1, '', 'NOIDEA');

# Source 'wormbase_transcript::wormbase' (id = 452)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('wormbase_transcript', '1', 'N', 50, 1, '', 'NOIDEA');

# Source 'wormpep_id::wormbase' (id = 453)
INSERT INTO source (name, source_release, download, ordered, priority, priority_description, status)
VALUES ('wormpep_id', '1', 'N', 50, 1, '', 'NOIDEA');

################################################################################
# DATA FILES

#-------------------------------------------------------------------------------
# Data for species 'acyrthosiphon_pisum' (id = 7029)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 7029, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 7029, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 7029, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::CORE' (id = 107)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (107, 7029, 'script:project=>ensemblgenomes,', \N, now(), now(), 'InterproFromCoreParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 7029, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 7029, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 7029, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::acyrthosiphon_pisum' (id = 308)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (308, 7029, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Acyrthosiphon_pisum/Aps.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Acyrthosiphon_pisum/Aps.data.gz', \N, now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 7029, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 7029, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 7029, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 7029, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'aedes_aegypti' (id = 7159)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 7159, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 7159, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 7159, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'ImmunoDB::aedes_aegypti' (id = 103)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (103, 7159, 'file:ImmunoDB/Aedes_ImmunoDB.txt', \N, now(), now(), 'ImmunoDBParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 7159, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 7159, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'PUBMED::aedes_aegypti' (id = 144)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (144, 7159, 'file:PubMed/PubMed_citations_Aedes_aegypti_Liverpool.txt', \N, now(), now(), 'VBPubMedParser');

# Data from source 'UniGene::aedes_aegypti' (id = 309)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (309, 7159, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Aedes_aegypti/Aae.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Aedes_aegypti/Aae.data.gz', \N, now(), now(), 'UniGeneParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 7159, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 7159, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'VB_Community_Annotation::aedes_aegypti' (id = 382)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (382, 7159, 'file:VB_Community_Annotation/Aedes_VB_Community_Annotation.fa', \N, now(), now(), 'VBCommunitySymbolParser');

# Data from source 'VB_Community_Annotation_Synonym::aedes_aegypti' (id = 386)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (386, 7159, 'file:VB_Community_Annotation/Aedes_aegypti_synonym.txt', \N, now(), now(), 'VBCommunitySynonymParser');

# Data from source 'VB_External_Description::aedes_aegypti' (id = 389)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (389, 7159, 'file:VB_External_Description/AaegL1-2_gene_description.txt', \N, now(), now(), 'VBExternalDescriptionParser');

# Data from source 'VB_RNA_Description::aedes_aegypti' (id = 392)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (392, 7159, 'file:VB_RNA_Description/AaegL1-2_RNA_description.txt', \N, now(), now(), 'VBRNADescriptionParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 7159, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'aegilops_tauschii' (id = 37682)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 37682, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 37682, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 37682, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 37682, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 37682, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (167, 37682, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (235, 37682, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 37682, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 37682, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 37682, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 37682, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'ncRNA_EG::EG' (id = 428)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (428, 37682, 'script:logic_name=>ncrna_eg,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

#-------------------------------------------------------------------------------
# Data for species 'afumigatusa1163' (id = 451804)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 451804, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 451804, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 451804, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 451804, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 451804, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 451804, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RefSeq_dna::MULTI-fungi' (id = 169)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (169, 451804, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-fungi' (id = 237)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (237, 451804, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 451804, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 451804, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 451804, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 451804, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'ailuropoda_melanoleuca' (id = 9646)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 9646, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 9646, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 9646, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 9646, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 9646, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 9646, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 9646, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 9646, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::ailuropoda_melanoleuca' (id = 175)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (175, 9646, 'ftp://ftp.ncbi.nih.gov/genomes/Ailuropoda_melanoleuca/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 9646, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::ailuropoda_melanoleuca' (id = 243)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (243, 9646, 'ftp://ftp.ncbi.nih.gov/genomes/Ailuropoda_melanoleuca/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 9646, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 9646, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 9646, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 9646, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 9646, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'amphimedon_queenslandica' (id = 400682)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 400682, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 400682, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 400682, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 400682, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 400682, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::amphimedon_queenslandica' (id = 176)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (176, 400682, 'ftp://ftp.ncbi.nih.gov/genomes/Amphimedon_queenslandica/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_peptide::amphimedon_queenslandica' (id = 244)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (244, 400682, 'ftp://ftp.ncbi.nih.gov/genomes/Amphimedon_queenslandica/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 400682, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 400682, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 400682, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'anas_platyrhynchos' (id = 8839)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 8839, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 8839, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 8839, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 8839, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 8839, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 8839, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 8839, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 8839, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::anas_platyrhynchos' (id = 177)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (177, 8839, 'ftp://ftp.ncbi.nih.gov/genomes/Anas_platyrhynchos/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 8839, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::anas_platyrhynchos' (id = 245)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (245, 8839, 'ftp://ftp.ncbi.nih.gov/genomes/Anas_platyrhynchos/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 8839, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 8839, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 8839, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 8839, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 8839, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'anolis_carolinensis' (id = 28377)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 28377, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 28377, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 28377, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 28377, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 28377, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 28377, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 28377, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 28377, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::anolis_carolinensis' (id = 178)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (178, 28377, 'ftp://ftp.ncbi.nih.gov/genomes/Anolis_carolinensis/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 28377, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::anolis_carolinensis' (id = 246)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (246, 28377, 'ftp://ftp.ncbi.nih.gov/genomes/Anolis_carolinensis/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::anolis_carolinensis' (id = 310)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (310, 28377, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Anolis_carolinensis/Acr.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Anolis_carolinensis/Acr.data.gz', \N, now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 28377, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 28377, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 28377, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 28377, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 28377, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'anopheles_darlingi' (id = 43151)
#-------------------------------------------------------------------------------

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 43151, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 43151, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::CORE' (id = 107)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (107, 43151, 'script:project=>ensemblgenomes,', \N, now(), now(), 'InterproFromCoreParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 43151, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 43151, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 43151, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 43151, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 43151, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 43151, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 43151, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'anopheles_gambiae' (id = 7165)
#-------------------------------------------------------------------------------

# Data from source 'AnoXcel::anopheles_gambiae' (id = 3)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (3, 7165, 'file:AnoXcel/anoXcel3.6-Xrefs.txt', \N, now(), now(), 'AnoXcelParser');

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 7165, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'Celera_Gene::anopheles_gambiae' (id = 22)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (22, 7165, '', \N, now(), now(), 'CeleraProteinParser');

# Data from source 'Celera_Pep::anopheles_gambiae' (id = 23)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (23, 7165, 'file:Celera/Anopheles_Celera_consensus-proteins_xref.fasta', \N, now(), now(), 'CeleraProteinParser');

# Data from source 'Celera_Trans::anopheles_gambiae' (id = 24)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (24, 7165, 'file:Celera/Anopheles_Celera_consensus-transcripts_xref.fasta', \N, now(), now(), 'CeleraTranscriptParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 7165, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 7165, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'ImmunoDB::anopheles_gambiae' (id = 104)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (104, 7165, 'file:ImmunoDB/Anopheles_ImmunoDB.txt', \N, now(), now(), 'ImmunoDBParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 7165, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 7165, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'PUBMED::anopheles_gambiae' (id = 145)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (145, 7165, 'file:PubMed/PubMed_citations_Anopheles_gambiae_PEST.txt', \N, now(), now(), 'VBPubMedParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 7165, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_mRNA_predicted::MULTI' (id = 229)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (229, 7165, '', \N, now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 7165, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'Ribosomal_Protein_Gene_DB::anopheles_gambiae' (id = 292)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (292, 7165, 'file:Ribosomal_Protein_Gene_Database/RPG_Anopheles_gambiae_PEST.txt', \N, now(), now(), 'VBribosomalParser');

# Data from source 'UniGene::anopheles_gambiae' (id = 311)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (311, 7165, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Anopheles_gambiae/Aga.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Anopheles_gambiae/Aga.data.gz', \N, now(), now(), 'UniGeneParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 7165, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 7165, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'VB_Community_Annotation::anopheles_gambiae' (id = 383)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (383, 7165, 'file:VB_Community_Annotation/Anopheles_gambiae_symbols.txt', \N, now(), now(), 'VBCommunitySymbolParser');

# Data from source 'VB_Community_Annotation_Synonym::anopheles_gambiae' (id = 387)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (387, 7165, 'file:VB_Community_Annotation/Anopheles_gambiae_synonyms.txt', \N, now(), now(), 'VBCommunitySynonymParser');

# Data from source 'VB_RNA_Description::anopheles_gambiae' (id = 393)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (393, 7165, 'file:VB_RNA_Description/An_AgamP3-5_ncRNA_description_lookup.txt', \N, now(), now(), 'VBRNADescriptionParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 7165, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'apis_mellifera' (id = 7460)
#-------------------------------------------------------------------------------

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 7460, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 7460, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::CORE' (id = 107)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (107, 7460, 'script:project=>ensemblgenomes,', \N, now(), now(), 'InterproFromCoreParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 7460, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::apis_mellifera' (id = 179)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (179, 7460, 'ftp://ftp.ncbi.nih.gov/genomes/Apis_mellifera/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_peptide::apis_mellifera' (id = 247)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (247, 7460, 'ftp://ftp.ncbi.nih.gov/genomes/Apis_mellifera/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::apis_mellifera' (id = 312)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (312, 7460, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Apis_mellifera/Ame.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Apis_mellifera/Ame.data.gz', 'ftp://ftp.ncbi.nih.gov/repository/UniGene/*.LOG', now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 7460, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 7460, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 7460, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 7460, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'ncRNA_EG::EG' (id = 428)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (428, 7460, 'script:logic_name=>ncrna_eg,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

#-------------------------------------------------------------------------------
# Data for species 'arabidopsis_lyrata' (id = 81972)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 81972, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 81972, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 81972, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 81972, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 81972, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (167, 81972, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_mRNA_predicted::MULTI' (id = 229)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (229, 81972, '', \N, now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (235, 81972, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 81972, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 81972, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 81972, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 81972, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 81972, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'arabidopsis_thaliana' (id = 3702)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 3702, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'CommunityGO::arabidopsis_thaliana' (id = 32)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (32, 3702, 'http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz ftp://ftp.geneontology.org/pub/go/gene-associations/gene_association.tair.gz', \N, now(), now(), 'TAIROntologyParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 3702, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'Gramene_Pathway::arabidopsis_thaliana' (id = 72)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (72, 3702, 'ftp://ftp.gramene.org/pub/gramene/ensemblgenomes/xrefs/pathways/aracyc.tab.gz', \N, now(), now(), 'GramenePathwayParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 3702, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 3702, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'PO_GROW::arabidopsis_thaliana' (id = 142)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (142, 3702, 'http://palea.cgrb.oregonstate.edu/viewsvn/Poc/trunk/ontology/OBO_format/plant_ontology.obo?view=co ftp://ftp.arabidopsis.org/home/tair/Ontologies/Plant_Ontology/po_temporal_gene_arabidopsis_tair.assoc', \N, now(), now(), 'TAIROntologyParser');

# Data from source 'PO_STRU::arabidopsis_thaliana' (id = 143)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (143, 3702, 'http://palea.cgrb.oregonstate.edu/viewsvn/Poc/trunk/ontology/OBO_format/plant_ontology.obo?view=co ftp://ftp.arabidopsis.org/home/tair/Ontologies/Plant_Ontology/po_anatomy_gene_arabidopsis_tair.assoc', \N, now(), now(), 'TAIROntologyParser');

# Data from source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (167, 3702, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_mRNA_predicted::MULTI' (id = 229)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (229, 3702, '', \N, now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (235, 3702, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TAIR_LOCUS::arabidopsis_thaliana' (id = 298)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (298, 3702, 'ftp://ftp.arabidopsis.org/Sequences/blast_datasets/TAIR10_blastsets/TAIR10_cdna_20101214_updated', \N, now(), now(), 'TAIRIDParser');

# Data from source 'UniGene::arabidopsis_thaliana' (id = 313)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (313, 3702, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Arabidopsis_thaliana/At.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Arabidopsis_thaliana/At.data.gz', \N, now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 3702, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 3702, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 3702, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 3702, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'ashbya_gossypii' (id = 284811)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 284811, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 284811, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 284811, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 284811, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 284811, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'PHIbase::MULTI' (id = 136)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (136, 284811, 'http://www.phi-base.org/PHI-export.php', \N, now(), now(), 'PHIbaseParser');

# Data from source 'RefSeq_dna::MULTI-fungi' (id = 169)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (169, 284811, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-fungi' (id = 237)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (237, 284811, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 284811, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 284811, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 284811, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'aspergillus_clavatus' (id = 5057)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 5057, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 5057, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 5057, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 5057, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 5057, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 5057, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 5057, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RefSeq_dna::MULTI-fungi' (id = 169)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (169, 5057, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-fungi' (id = 237)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (237, 5057, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 5057, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 5057, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 5057, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 5057, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'aspergillus_flavus' (id = 5059)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 5059, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 5059, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 5059, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 5059, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 5059, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 5059, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 5059, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RefSeq_dna::MULTI-fungi' (id = 169)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (169, 5059, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-fungi' (id = 237)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (237, 5059, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 5059, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 5059, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 5059, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 5059, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'aspergillus_fumigatus' (id = 330879)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 330879, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 330879, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 330879, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 330879, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 330879, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RefSeq_dna::MULTI-fungi' (id = 169)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (169, 330879, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-fungi' (id = 237)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (237, 330879, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 330879, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 330879, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 330879, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 330879, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'aspergillus_nidulans' (id = 162425)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 162425, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 162425, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 162425, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 162425, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 162425, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RefSeq_dna::MULTI-fungi' (id = 169)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (169, 162425, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-fungi' (id = 237)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (237, 162425, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 162425, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 162425, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 162425, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 162425, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'aspergillus_niger' (id = 425011)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 425011, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 425011, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 425011, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 425011, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 425011, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::EG' (id = 152)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (152, 425011, 'script:object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 425011, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RefSeq_dna::MULTI-fungi' (id = 169)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (169, 425011, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-fungi' (id = 237)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (237, 425011, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 425011, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 425011, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 425011, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 425011, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 425011, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

# Data from source 'misc_EG::EG' (id = 427)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (427, 425011, 'script:host=>mysql-eg-pan-1,port=>4276,dbname=>EG_Xrefs,', \N, now(), now(), 'EG_DBParser');

#-------------------------------------------------------------------------------
# Data for species 'aspergillus_oryzae' (id = 5062)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 5062, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 5062, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 5062, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 5062, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 5062, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 5062, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 5062, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RefSeq_dna::MULTI-fungi' (id = 169)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (169, 5062, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-fungi' (id = 237)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (237, 5062, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 5062, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 5062, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 5062, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 5062, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'aspergillus_terreus' (id = 33178)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 33178, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 33178, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 33178, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 33178, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 33178, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 33178, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 33178, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RefSeq_dna::MULTI-fungi' (id = 169)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (169, 33178, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-fungi' (id = 237)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (237, 33178, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 33178, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 33178, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 33178, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 33178, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'astyanax_mexicanus' (id = 7994)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 7994, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 7994, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 7994, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 7994, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 7994, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 7994, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 7994, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 7994, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::astyanax_mexicanus' (id = 180)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (180, 7994, 'ftp://ftp.ncbi.nih.gov/genomes/Astyanax_mexicanus/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 7994, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::astyanax_mexicanus' (id = 248)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (248, 7994, 'ftp://ftp.ncbi.nih.gov/genomes/Astyanax_mexicanus/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 7994, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 7994, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 7994, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 7994, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 7994, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'atta_cephalotes' (id = 12957)
#-------------------------------------------------------------------------------

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 12957, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 12957, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::CORE' (id = 107)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (107, 12957, 'script:project=>ensemblgenomes,', \N, now(), now(), 'InterproFromCoreParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 12957, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 12957, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 12957, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 12957, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 12957, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 12957, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 12957, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'ncRNA_EG::EG' (id = 428)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (428, 12957, 'script:logic_name=>ncrna_eg,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

#-------------------------------------------------------------------------------
# Data for species 'bombyx_mori' (id = 7091)
#-------------------------------------------------------------------------------

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 7091, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 7091, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::CORE' (id = 107)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (107, 7091, 'script:project=>ensemblgenomes,', \N, now(), now(), 'InterproFromCoreParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 7091, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 7091, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 7091, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::bombyx_mori' (id = 314)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (314, 7091, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Bombyx_mori/Bmo.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Bombyx_mori/Bmo.data.gz', \N, now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 7091, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 7091, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 7091, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 7091, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'ncRNA_EG::EG' (id = 428)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (428, 7091, 'script:logic_name=>ncrna_eg,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

#-------------------------------------------------------------------------------
# Data for species 'bos_taurus' (id = 9913)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 9913, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 9913, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 9913, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 9913, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 9913, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 9913, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 9913, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 9913, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::bos_taurus' (id = 181)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (181, 9913, 'ftp://ftp.ncbi.nih.gov/genomes/Bos_taurus/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 9913, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::bos_taurus' (id = 249)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (249, 9913, 'ftp://ftp.ncbi.nih.gov/genomes/Bos_taurus/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::bos_taurus' (id = 315)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (315, 9913, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Bos_taurus/Bt.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Bos_taurus/Bt.data.gz', \N, now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 9913, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 9913, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 9913, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 9913, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 9913, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'brachypodium_distachyon' (id = 15368)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 15368, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 15368, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 15368, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Gramene_Pathway::brachypodium_distachyon' (id = 73)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (73, 15368, 'ftp://ftp.gramene.org/pub/gramene/ensemblgenomes/xrefs/pathways/brachycyc.tab.gz', \N, now(), now(), 'GramenePathwayParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 15368, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 15368, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (167, 15368, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_mRNA_predicted::MULTI' (id = 229)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (229, 15368, '', \N, now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (235, 15368, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::brachypodium_distachyon' (id = 316)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (316, 15368, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Brachypodium_distachyon/Bdi.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Brachypodium_distachyon/Bdi.data.gz', \N, now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 15368, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 15368, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 15368, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 15368, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'brassica_rapa' (id = 51351)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 51351, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 51351, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 51351, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 51351, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 51351, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (167, 51351, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (235, 51351, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 51351, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 51351, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 51351, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 51351, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'ncRNA_EG::EG' (id = 428)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (428, 51351, 'script:logic_name=>ncrna_eg,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

#-------------------------------------------------------------------------------
# Data for species 'brugia_malayi' (id = 6279)
#-------------------------------------------------------------------------------

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 6279, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 6279, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 6279, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 6279, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 6279, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI-invertebrate' (id = 369)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (369, 6279, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/uniprot_trembl_invertebrates.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI-invertebrate' (id = 374)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (374, 6279, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/uniprot_sprot_invertebrates.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'wormbase::bmalayi' (id = 440)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (440, 6279, 'ftp://ftp.ebi.ac.uk/pub/databases/wormbase/releases/RELEASE/species/b_malayi/PRJNA10729/b_malayi.PRJNA10729.RELEASE.xrefs.txt.gz', \N, now(), now(), 'WormbaseDirectParser');

#-------------------------------------------------------------------------------
# Data for species 'caenorhabditis_brenneri' (id = 135651)
#-------------------------------------------------------------------------------

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 135651, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 135651, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 135651, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 135651, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 135651, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI-invertebrate' (id = 369)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (369, 135651, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/uniprot_trembl_invertebrates.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI-invertebrate' (id = 374)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (374, 135651, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/uniprot_sprot_invertebrates.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'wormbase::cbrenneri' (id = 441)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (441, 135651, 'ftp://ftp.ebi.ac.uk/pub/databases/wormbase/releases/RELEASE/species/c_brenneri/PRJNA20035/c_brenneri.PRJNA20035.RELEASE.xrefs.txt.gz', \N, now(), now(), 'WormbaseDirectParser');

#-------------------------------------------------------------------------------
# Data for species 'caenorhabditis_briggsae' (id = 6238)
#-------------------------------------------------------------------------------

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 6238, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 6238, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 6238, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 6238, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 6238, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI-invertebrate' (id = 369)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (369, 6238, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/uniprot_trembl_invertebrates.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI-invertebrate' (id = 374)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (374, 6238, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/uniprot_sprot_invertebrates.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'wormbase::cbriggsae' (id = 442)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (442, 6238, 'ftp://ftp.ebi.ac.uk/pub/databases/wormbase/releases/RELEASE/species/c_briggsae/PRJNA10731/c_briggsae.PRJNA10731.RELEASE.xrefs.txt.gz', \N, now(), now(), 'WormbaseDirectParser');

#-------------------------------------------------------------------------------
# Data for species 'caenorhabditis_elegans' (id = 6239)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 6239, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 6239, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 6239, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 6239, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 6239, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::caenorhabditis_elegans' (id = 318)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (318, 6239, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Caenorhabditis_elegans/Cel.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Caenorhabditis_elegans/Cel.data.gz', 'ftp://ftp.ncbi.nih.gov/repository/UniGene/*.LOG', now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 6239, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI-invertebrate' (id = 369)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (369, 6239, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/uniprot_trembl_invertebrates.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI-invertebrate' (id = 374)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (374, 6239, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/uniprot_sprot_invertebrates.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'wormbase::celegans' (id = 443)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (443, 6239, 'ftp://ftp.ebi.ac.uk/pub/databases/wormbase/releases/RELEASE/species/c_elegans/PRJNA13758/c_elegans.PRJNA13758.RELEASE.xrefs.txt.gz ', \N, now(), now(), 'WormbaseDirectParser');

#-------------------------------------------------------------------------------
# Data for species 'caenorhabditis_japonica' (id = 281687)
#-------------------------------------------------------------------------------

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 281687, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 281687, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 281687, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 281687, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 281687, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI-invertebrate' (id = 369)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (369, 281687, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/uniprot_trembl_invertebrates.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI-invertebrate' (id = 374)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (374, 281687, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/uniprot_sprot_invertebrates.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'wormbase::cjaponica' (id = 444)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (444, 281687, 'ftp://ftp.ebi.ac.uk/pub/databases/wormbase/releases/RELEASE/species/c_japonica/PRJNA12591/c_japonica.PRJNA12591.RELEASE.xrefs.txt.gz', \N, now(), now(), 'WormbaseDirectParser');

#-------------------------------------------------------------------------------
# Data for species 'caenorhabditis_remanei' (id = 31234)
#-------------------------------------------------------------------------------

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 31234, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 31234, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 31234, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 31234, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 31234, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI-invertebrate' (id = 369)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (369, 31234, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/uniprot_trembl_invertebrates.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI-invertebrate' (id = 374)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (374, 31234, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/uniprot_sprot_invertebrates.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'wormbase::cremanei' (id = 445)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (445, 31234, 'ftp://ftp.ebi.ac.uk/pub/databases/wormbase/releases/RELEASE/species/c_remanei/PRJNA53967/c_remanei.PRJNA53967.RELEASE.xrefs.txt.gz', \N, now(), now(), 'WormbaseDirectParser');

#-------------------------------------------------------------------------------
# Data for species 'callithrix_jacchus' (id = 9483)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 9483, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 9483, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 9483, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 9483, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 9483, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 9483, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 9483, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 9483, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::callithrix_jacchus' (id = 183)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (183, 9483, 'ftp://ftp.ncbi.nih.gov/genomes/Callithrix_jacchus/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 9483, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::callithrix_jacchus' (id = 251)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (251, 9483, 'ftp://ftp.ncbi.nih.gov/genomes/Callithrix_jacchus/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 9483, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 9483, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 9483, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 9483, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 9483, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'canis_familiaris' (id = 9615)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 9615, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 9615, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 9615, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 9615, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 9615, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 9615, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 9615, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 9615, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::canis_familiaris' (id = 184)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (184, 9615, 'ftp://ftp.ncbi.nih.gov/genomes/Canis_lupus_familiaris/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 9615, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::canis_familiaris' (id = 252)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (252, 9615, 'ftp://ftp.ncbi.nih.gov/genomes/Canis_lupus_familiaris/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::canis_familiaris' (id = 319)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (319, 9615, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Canis_lupus_familiaris/Cfa.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Canis_lupus_familiaris/Cfa.data.gz', 'ftp://ftp.ncbi.nih.gov/repository/UniGene/*.LOG', now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 9615, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 9615, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 9615, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 9615, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 9615, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'capitella_teleta' (id = 283909)
#-------------------------------------------------------------------------------

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 283909, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 283909, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::CORE' (id = 107)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (107, 283909, 'script:project=>ensemblgenomes,', \N, now(), now(), 'InterproFromCoreParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 283909, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 283909, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 283909, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 283909, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 283909, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 283909, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 283909, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'ncRNA_EG::EG' (id = 428)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (428, 283909, 'script:logic_name=>ncrna_eg,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

#-------------------------------------------------------------------------------
# Data for species 'cavia_porcellus' (id = 10141)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 10141, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 10141, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 10141, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 10141, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 10141, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 10141, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 10141, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 10141, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::cavia_porcellus' (id = 185)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (185, 10141, 'ftp://ftp.ncbi.nih.gov/genomes/Cavia_porcellus/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 10141, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::cavia_porcellus' (id = 253)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (253, 10141, 'ftp://ftp.ncbi.nih.gov/genomes/Cavia_porcellus/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 10141, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 10141, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 10141, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 10141, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 10141, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'chlamydomonas_reinhardtii' (id = 3055)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 3055, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 3055, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 3055, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 3055, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 3055, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (167, 3055, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_mRNA_predicted::MULTI' (id = 229)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (229, 3055, '', \N, now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (235, 3055, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::chlamydomonas_reinhardtii' (id = 320)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (320, 3055, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Chlamydomonas_reinhardtii/Cre.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Chlamydomonas_reinhardtii/Cre.data.gz', \N, now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 3055, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 3055, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 3055, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 3055, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'chlorocebus_sabaeus' (id = 60711)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 60711, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 60711, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 60711, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 60711, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 60711, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 60711, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 60711, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 60711, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::MULTI-vertebrate_mammalian' (id = 173)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (173, 60711, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_mammalian/vertebrate_mammalian*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 60711, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::MULTI-vertebrate_mammalian' (id = 241)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (241, 60711, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_mammalian/vertebrate_mammalian*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 60711, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 60711, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 60711, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 60711, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 60711, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'choloepus_hoffmanni' (id = 9358)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 9358, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 9358, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 9358, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 9358, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 9358, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 9358, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 9358, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 9358, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::MULTI-vertebrate_mammalian' (id = 173)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (173, 9358, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_mammalian/vertebrate_mammalian*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 9358, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::MULTI-vertebrate_mammalian' (id = 241)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (241, 9358, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_mammalian/vertebrate_mammalian*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 9358, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'UniProt::protein_id' (id = 364)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (364, 9358, '', \N, now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 9358, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 9358, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot::EMBL' (id = 377)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (377, 9358, '', \N, now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 9358, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 9358, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'ciona_intestinalis' (id = 7719)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 7719, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 7719, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 7719, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 7719, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 7719, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 7719, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 7719, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 7719, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::MULTI-vertebrate_other' (id = 174)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (174, 7719, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_other/vertebrate_other*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 7719, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::ciona_intestinalis' (id = 254)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (254, 7719, 'ftp://ftp.ncbi.nih.gov/genomes/Ciona_intestinalis/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_peptide::ciona_intestinalis' (id = 254)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (254, 7719, 'ftp://ftp.ncbi.nih.gov/genomes/Ciona_intestinalis/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::ciona_intestinalis' (id = 321)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (321, 7719, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Ciona_intestinalis/Cin.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Ciona_intestinalis/Cin.data.gz', 'ftp://ftp.ncbi.nih.gov/repository/UniGene/*.LOG', now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 7719, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 7719, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 7719, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'cint_aniseed_v1::ciona_intestinalis' (id = 403)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (403, 7719, 'ftp://ftp.jgi-psf.org/pub/JGI_data/Ciona/v1.0/ciona.prot.fasta.gz', \N, now(), now(), 'JGI_ProteinParser');

# Data from source 'cint_jgi_v1::ciona_intestinalis' (id = 404)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (404, 7719, 'ftp://ftp.jgi-psf.org/pub/JGI_data/Ciona/v1.0/ciona.prot.fasta.gz', \N, now(), now(), 'JGI_ProteinParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 7719, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 7719, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'ciona_savignyi' (id = 51511)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 51511, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 51511, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 51511, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 51511, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 51511, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 51511, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 51511, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 51511, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::MULTI-vertebrate_other' (id = 174)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (174, 51511, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_other/vertebrate_other*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 51511, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 51511, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_peptide::MULTI-vertebrate_other' (id = 242)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (242, 51511, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_other/vertebrate_other*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::ciona_savignyi' (id = 322)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (322, 51511, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Ciona_savignyi/Csa.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Ciona_savignyi/Csa.data.gz', 'ftp://ftp.ncbi.nih.gov/repository/UniGene/*.LOG', now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 51511, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 51511, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 51511, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 51511, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 51511, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'crassostrea_gigas' (id = 29159)
#-------------------------------------------------------------------------------

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 29159, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 29159, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::CORE' (id = 107)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (107, 29159, 'script:project=>ensemblgenomes,', \N, now(), now(), 'InterproFromCoreParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 29159, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 29159, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 29159, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 29159, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 29159, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 29159, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 29159, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'ncRNA_EG::EG' (id = 428)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (428, 29159, 'script:logic_name=>ncrna_eg,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

#-------------------------------------------------------------------------------
# Data for species 'culex_quinquefasciatus' (id = 7176)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 7176, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 7176, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 7176, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'ImmunoDB::culex_quinquefasciatus' (id = 105)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (105, 7176, 'file:ImmunoDB/Culex_ImmunoDB.txt', \N, now(), now(), 'ImmunoDBParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 7176, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'PUBMED::culex_quinquefasciatus' (id = 146)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (146, 7176, 'file:PubMed/PubMed_citations_Culex_quinquefasciatus_JHB.txt', \N, now(), now(), 'VBPubMedParser');

# Data from source 'UniGene::culex_quinquefasciatus' (id = 323)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (323, 7176, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Culex_quinquefasciatus/Cpi.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Culex_quinquefasciatus/Cpi.data.gz', 'ftp://ftp.ncbi.nih.gov/repository/UniGene/*.LOG', now(), now(), 'UniGeneParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 7176, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 7176, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'VB_Community_Annotation::culex_quinquefasciatus' (id = 384)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (384, 7176, 'file:VB_Community_Annotation/Culex_quinquefasciatus_symbols.txt', \N, now(), now(), 'VBCommunitySymbolParser');

# Data from source 'VB_Community_Annotation_Synonym::culex_quinquefasciatus' (id = 388)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (388, 7176, 'file:VB_Community_Annotation/Culex_quinquefasciatus_synonyms.txt', \N, now(), now(), 'VBCommunitySynonymParser');

# Data from source 'VB_External_Description::culex_quinquefasciatus' (id = 390)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (390, 7176, 'file:VB_External_Description/CpipJ1-2_gene_description.txt', \N, now(), now(), 'VBExternalDescriptionParser');

# Data from source 'VB_RNA_Description::culex_quinquefasciatus' (id = 394)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (394, 7176, 'file:VB_RNA_Description/CpipJ1-2_RNA_description.txt', \N, now(), now(), 'VBRNADescriptionParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 7176, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'cyanidioschyzon_merolae' (id = 45157)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 45157, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 45157, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 45157, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 45157, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 45157, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (167, 45157, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_mRNA_predicted::MULTI' (id = 229)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (229, 45157, '', \N, now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (235, 45157, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 45157, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 45157, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 45157, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 45157, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'danaus_plexippus' (id = 13037)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 13037, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 13037, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 13037, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 13037, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 13037, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 13037, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 13037, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 13037, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 13037, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'danio_rerio' (id = 7955)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 7955, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 7955, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::danio_rerio' (id = 59)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (59, 7955, 'ftp://ftp.geneontology.org/pub/go/gene-associations/gene_association.zfin.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 7955, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 7955, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 7955, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 7955, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 7955, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::danio_rerio' (id = 187)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (187, 7955, 'ftp://ftp.ncbi.nih.gov/refseq/D_rerio/mRNA_Prot/zebrafish*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 7955, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::danio_rerio' (id = 255)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (255, 7955, 'ftp://ftp.ncbi.nih.gov/refseq/D_rerio/mRNA_Prot/zebrafish*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::danio_rerio' (id = 324)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (324, 7955, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Danio_rerio/Dr.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Danio_rerio/Dr.data.gz', 'ftp://ftp.ncbi.nih.gov/repository/UniGene/*.LOG', now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 7955, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 7955, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 7955, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'ZFIN_ID::danio_rerio#01' (id = 397)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (397, 7955, 'http://zfin.org/data_transfer/Downloads/refseq.txt http://zfin.org/data_transfer/Downloads/uniprot.txt http://zfin.org/data_transfer/Downloads/aliases.txt http://zfin.org/data_transfer/Downloads/gene_seq.txt', \N, now(), now(), 'ZFINParser');

# Data from source 'ZFIN_ID::danio_rerio#02' (id = 398)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (398, 7955, 'script:source=>ZFIN_ID,host=>ens-staging1,', \N, now(), now(), 'VegaOfficialNameParser');

# Data from source 'ZFIN_ID::danio_rerio#03' (id = 399)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (399, 7955, 'ftp://zfin.org/pub/transfer/MEOW/zfin_genes.txt', \N, now(), now(), 'ZFINDescParser');

# Data from source 'ZFIN_ID_curated_transcript::danio_rerio' (id = 400)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (400, 7955, 'script:source=>ZFIN_ID,host=>ens-staging1,', \N, now(), now(), 'curated_transcriptParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 7955, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 7955, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'daphnia_pulex' (id = 6669)
#-------------------------------------------------------------------------------

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 6669, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 6669, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::CORE' (id = 107)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (107, 6669, 'script:project=>ensemblgenomes,', \N, now(), now(), 'InterproFromCoreParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 6669, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 6669, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 6669, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 6669, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 6669, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 6669, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 6669, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'ncRNA_EG::EG' (id = 428)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (428, 6669, 'script:logic_name=>ncrna_eg,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

#-------------------------------------------------------------------------------
# Data for species 'dasypus_novemcinctus' (id = 9361)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 9361, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 9361, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 9361, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 9361, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 9361, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 9361, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 9361, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 9361, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::dasypus_novemcinctus' (id = 188)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (188, 9361, 'ftp://ftp.ncbi.nih.gov/genomes/Dasypus_novemcinctus/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 9361, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::dasypus_novemcinctus' (id = 256)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (256, 9361, 'ftp://ftp.ncbi.nih.gov/genomes/Dasypus_novemcinctus/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 9361, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 9361, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 9361, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 9361, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 9361, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'dendroctonus_ponderosae' (id = 77166)
#-------------------------------------------------------------------------------

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 77166, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 77166, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::CORE' (id = 107)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (107, 77166, 'script:project=>ensemblgenomes,', \N, now(), now(), 'InterproFromCoreParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 77166, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 77166, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 77166, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 77166, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 77166, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 77166, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 77166, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'ncRNA_EG::EG' (id = 428)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (428, 77166, 'script:logic_name=>ncrna_eg,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

#-------------------------------------------------------------------------------
# Data for species 'dictyostelium_discoideum' (id = 366501)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 366501, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 366501, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 366501, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 366501, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 366501, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::EG' (id = 152)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (152, 366501, 'script:object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 366501, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RefSeq_dna::MULTI-protozoa' (id = 172)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (172, 366501, 'ftp://ftp.ncbi.nih.gov/refseq/release/protozoa/protozoa*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-protozoa' (id = 240)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (240, 366501, 'ftp://ftp.ncbi.nih.gov/refseq/release/protozoa/protozoa*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 366501, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 366501, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 366501, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 366501, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 366501, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

# Data from source 'misc_EG::EG' (id = 427)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (427, 366501, 'script:host=>mysql-eg-pan-1,port=>4276,dbname=>EG_Xrefs,', \N, now(), now(), 'EG_DBParser');

#-------------------------------------------------------------------------------
# Data for species 'dipodomys_ordii' (id = 10020)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 10020, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 10020, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 10020, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 10020, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 10020, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 10020, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 10020, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 10020, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::MULTI-vertebrate_mammalian' (id = 173)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (173, 10020, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_mammalian/vertebrate_mammalian*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 10020, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::MULTI-vertebrate_mammalian' (id = 241)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (241, 10020, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_mammalian/vertebrate_mammalian*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 10020, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 10020, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 10020, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 10020, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 10020, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'drosophila_ananassae' (id = 7217)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 7217, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 7217, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 7217, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 7217, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 7217, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 7217, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 7217, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 7217, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 7217, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 7217, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'flybase_gene_id::drosophila_ananassae' (id = 406)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (406, 7217, 'ftp://ftp.flybase.org/releases/current/dana_r1.3/gff/dana-all-r1.3.gff.gz', \N, now(), now(), 'FlybaseParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 7217, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'drosophila_erecta' (id = 7220)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 7220, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 7220, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 7220, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 7220, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 7220, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 7220, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 7220, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 7220, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 7220, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 7220, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'flybase_gene_id::drosophila_erecta' (id = 407)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (407, 7220, 'ftp://ftp.flybase.org/releases/current/dere_r1.3/gff/dere-all-r1.3.gff.gz', \N, now(), now(), 'FlybaseParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 7220, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'drosophila_grimshawi' (id = 7222)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 7222, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 7222, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 7222, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 7222, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 7222, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 7222, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 7222, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 7222, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 7222, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 7222, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'flybase_gene_id::drosophila_grimshawi' (id = 408)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (408, 7222, 'ftp://ftp.flybase.org/releases/current/dgri_r1.3/gff/dgri-all-r1.3.gff.gz', \N, now(), now(), 'FlybaseParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 7222, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'drosophila_melanogaster' (id = 7227)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 7227, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 7227, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::drosophila_melanogaster' (id = 60)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (60, 7227, 'ftp://ftp.geneontology.org/pub/go/gene-associations/gene_association.fb.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', \N, now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 7227, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 7227, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::drosophila_melanogaster' (id = 189)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (189, 7227, 'ftp://ftp.ncbi.nih.gov/genomes/Drosophila_melanogaster/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_peptide::drosophila_melanogaster' (id = 257)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (257, 7227, 'ftp://ftp.ncbi.nih.gov/genomes/Drosophila_melanogaster/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 7227, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 7227, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 7227, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'flybase_gene_id::drosophila_melanogaster' (id = 409)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (409, 7227, 'ftp://ftp.flybase.net/genomes/Drosophila_melanogaster/dmel_r5.46_FB2012_04/gff/dmel-all-r5.46.gff.gz', \N, now(), now(), 'FlybaseParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 7227, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'drosophila_mojavensis' (id = 7230)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 7230, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 7230, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 7230, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 7230, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 7230, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 7230, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 7230, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 7230, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 7230, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 7230, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'flybase_gene_id::drosophila_mojavensis' (id = 410)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (410, 7230, 'ftp://ftp.flybase.org/releases/current/dmoj_r1.3/gff/dmoj-all-r1.3.gff.gz', \N, now(), now(), 'FlybaseParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 7230, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'drosophila_persimilis' (id = 7234)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 7234, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 7234, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 7234, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 7234, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 7234, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 7234, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 7234, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 7234, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 7234, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 7234, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'flybase_gene_id::drosophila_persimilis' (id = 411)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (411, 7234, 'ftp://ftp.flybase.org/releases/current/dper_r1.3/gff/dper-all-r1.3.gff.gz', \N, now(), now(), 'FlybaseParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 7234, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'drosophila_pseudoobscura' (id = 46245)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 46245, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 46245, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 46245, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 46245, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 46245, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 46245, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 46245, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 46245, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 46245, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 46245, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'flybase_gene_id::drosophila_pseudoobscura' (id = 412)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (412, 46245, 'ftp://ftp.flybase.net/genomes/Drosophila_pseudoobscura/dpse/gff/dpse-all-r2.30.gff.gz', \N, now(), now(), 'FlybaseParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 46245, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'drosophila_sechellia' (id = 7238)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 7238, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 7238, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 7238, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 7238, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 7238, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 7238, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 7238, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 7238, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 7238, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 7238, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'flybase_gene_id::drosophila_sechellia' (id = 413)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (413, 7238, 'ftp://ftp.flybase.org/releases/current/dsec_r1.3/gff/dsec-all-r1.3.gff.gz', \N, now(), now(), 'FlybaseParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 7238, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'drosophila_simulans' (id = 7240)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 7240, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 7240, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 7240, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 7240, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 7240, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 7240, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 7240, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 7240, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 7240, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 7240, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'flybase_gene_id::drosophila_simulans' (id = 414)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (414, 7240, 'ftp://ftp.flybase.org/genomes/dsim/current/gff/dsim-all-r1.4.gff.gz', \N, now(), now(), 'FlybaseParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 7240, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'drosophila_virilis' (id = 7244)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 7244, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 7244, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 7244, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 7244, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 7244, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 7244, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 7244, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 7244, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 7244, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 7244, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'flybase_gene_id::drosophila_virilis' (id = 415)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (415, 7244, 'ftp://ftp.flybase.org/releases/current/dvir_r1.2/gff/dvir-all-r1.2.gff.gz', \N, now(), now(), 'FlybaseParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 7244, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'drosophila_willistoni' (id = 7260)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 7260, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 7260, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 7260, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 7260, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 7260, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 7260, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 7260, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 7260, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 7260, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 7260, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'flybase_gene_id::drosophila_willistoni' (id = 416)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (416, 7260, 'ftp://ftp.flybase.org/releases/current/dwil_r1.3/gff/dwil-all-r1.3.gff.gz', \N, now(), now(), 'FlybaseParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 7260, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'drosophila_yakuba' (id = 7245)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 7245, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 7245, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 7245, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 7245, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 7245, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 7245, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 7245, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 7245, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 7245, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 7245, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'flybase_gene_id::drosophila_yakuba' (id = 417)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (417, 7245, 'ftp://ftp.flybase.org/releases/current/dyak_r1.3/gff/dyak-all-r1.3.gff.gz', \N, now(), now(), 'FlybaseParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 7245, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'echinops_telfairi' (id = 9371)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 9371, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 9371, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 9371, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 9371, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 9371, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 9371, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 9371, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 9371, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::echinops_telfairi' (id = 190)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (190, 9371, 'ftp://ftp.ncbi.nih.gov/genomes/Echinops_telfairi/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 9371, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::echinops_telfairi' (id = 258)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (258, 9371, 'ftp://ftp.ncbi.nih.gov/genomes/Echinops_telfairi/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 9371, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 9371, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 9371, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 9371, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 9371, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'equus_caballus' (id = 9796)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 9796, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 9796, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 9796, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 9796, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 9796, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 9796, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 9796, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 9796, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::equus_caballus' (id = 191)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (191, 9796, 'ftp://ftp.ncbi.nih.gov/genomes/Equus_caballus/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 9796, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::equus_caballus' (id = 259)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (259, 9796, 'ftp://ftp.ncbi.nih.gov/genomes/Equus_caballus/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::equus_caballus' (id = 325)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (325, 9796, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Equus_caballus/Eca.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Equus_caballus/Eca.data.gz', 'ftp://ftp.ncbi.nih.gov/repository/UniGene/*.LOG', now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 9796, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 9796, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 9796, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 9796, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 9796, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'erinaceus_europaeus' (id = 9365)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 9365, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 9365, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 9365, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 9365, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 9365, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 9365, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 9365, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 9365, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::MULTI-vertebrate_mammalian' (id = 173)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (173, 9365, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_mammalian/vertebrate_mammalian*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 9365, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::MULTI-vertebrate_mammalian' (id = 241)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (241, 9365, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_mammalian/vertebrate_mammalian*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 9365, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 9365, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 9365, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 9365, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 9365, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'felis_catus' (id = 9685)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 9685, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 9685, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 9685, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 9685, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 9685, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 9685, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 9685, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 9685, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::felis_catus' (id = 192)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (192, 9685, 'ftp://ftp.ncbi.nih.gov/genomes/Felis_catus/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 9685, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::felis_catus' (id = 260)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (260, 9685, 'ftp://ftp.ncbi.nih.gov/genomes/Felis_catus/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 9685, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 9685, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 9685, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 9685, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 9685, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'ficedula_albicollis' (id = 59894)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 59894, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 59894, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 59894, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 59894, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 59894, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 59894, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 59894, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 59894, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::MULTI-vertebrate_other' (id = 174)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (174, 59894, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_other/vertebrate_other*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 59894, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::MULTI-vertebrate_other' (id = 242)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (242, 59894, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_other/vertebrate_other*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 59894, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 59894, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 59894, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 59894, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 59894, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'fusarium_oxysporum' (id = 426428)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 426428, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 426428, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 426428, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 426428, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 426428, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'PHIbase::MULTI' (id = 136)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (136, 426428, 'http://www.phi-base.org/PHI-export.php', \N, now(), now(), 'PHIbaseParser');

# Data from source 'RFAM::EG' (id = 152)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (152, 426428, 'script:object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 426428, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RefSeq_dna::MULTI-fungi' (id = 169)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (169, 426428, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-fungi' (id = 237)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (237, 426428, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 426428, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 426428, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 426428, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 426428, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 426428, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

# Data from source 'misc_EG::EG' (id = 427)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (427, 426428, 'script:host=>mysql-eg-pan-1,port=>4276,dbname=>EG_Xrefs,', \N, now(), now(), 'EG_DBParser');

#-------------------------------------------------------------------------------
# Data for species 'gadus_morhua' (id = 8049)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 8049, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 8049, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 8049, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 8049, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 8049, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 8049, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 8049, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 8049, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::MULTI-vertebrate_other' (id = 174)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (174, 8049, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_other/vertebrate_other*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 8049, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::MULTI-vertebrate_other' (id = 242)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (242, 8049, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_other/vertebrate_other*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::gadus_morhua' (id = 326)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (326, 8049, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Gadus_morhua/Gmr.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Gadus_morhua/Gmr.data.gz', 'ftp://ftp.ncbi.nih.gov/repository/UniGene/*.LOG', now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 8049, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 8049, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 8049, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 8049, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 8049, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'gaeumannomyces_graminis' (id = 644352)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 644352, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'BROAD_Magnaporthe_DB::MULTI' (id = 10)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (10, 644352, 'http://www.broadinstitute.org/annotation/genome/magnaporthe_comparative/download/?sp=EAGenomeSummary&sp=SGae_graminis_V2&sp=S.gz', \N, now(), now(), 'Broad_MagnaportheDBParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 644352, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 644352, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 644352, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 644352, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'PHIbase::MULTI' (id = 136)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (136, 644352, 'http://www.phi-base.org/PHI-export.php', \N, now(), now(), 'PHIbaseParser');

# Data from source 'RefSeq_dna::MULTI-fungi' (id = 169)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (169, 644352, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-fungi' (id = 237)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (237, 644352, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 644352, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 644352, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 644352, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'gallus_gallus' (id = 9031)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 9031, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'CGNC::gallus_gallus' (id = 21)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (21, 9031, 'http://birdgenenames.org/cgnc/downloads.jsp?file=standard', \N, now(), now(), 'CGNCParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 9031, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 9031, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 9031, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 9031, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 9031, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 9031, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 9031, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::gallus_gallus' (id = 193)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (193, 9031, 'ftp://ftp.ncbi.nih.gov/genomes/Gallus_gallus/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 9031, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::gallus_gallus' (id = 261)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (261, 9031, 'ftp://ftp.ncbi.nih.gov/genomes/Gallus_gallus/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::gallus_gallus' (id = 327)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (327, 9031, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Gallus_gallus/Gga.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Gallus_gallus/Gga.data.gz', 'ftp://ftp.ncbi.nih.gov/repository/UniGene/*.LOG', now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 9031, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 9031, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 9031, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 9031, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 9031, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'gasterosteus_aculeatus' (id = 69293)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 69293, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 69293, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 69293, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 69293, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 69293, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 69293, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 69293, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 69293, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::MULTI-vertebrate_other' (id = 174)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (174, 69293, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_other/vertebrate_other*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 69293, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::MULTI-vertebrate_other' (id = 242)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (242, 69293, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_other/vertebrate_other*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::gasterosteus_aculeatus' (id = 328)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (328, 69293, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Gasterosteus_aculeatus/Gac.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Gasterosteus_aculeatus/Gac.data.gz', 'ftp://ftp.ncbi.nih.gov/repository/UniGene/*.LOG', now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 69293, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 69293, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 69293, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 69293, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 69293, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'gibberella_moniliformis' (id = 117187)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 117187, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 117187, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 117187, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 117187, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 117187, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'PHIbase::MULTI' (id = 136)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (136, 117187, 'http://www.phi-base.org/PHI-export.php', \N, now(), now(), 'PHIbaseParser');

# Data from source 'RFAM::EG' (id = 152)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (152, 117187, 'script:object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 117187, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RefSeq_dna::MULTI-fungi' (id = 169)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (169, 117187, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-fungi' (id = 237)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (237, 117187, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 117187, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 117187, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 117187, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 117187, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 117187, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

# Data from source 'misc_EG::EG' (id = 427)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (427, 117187, 'script:host=>mysql-eg-pan-1,port=>4276,dbname=>EG_Xrefs,', \N, now(), now(), 'EG_DBParser');

#-------------------------------------------------------------------------------
# Data for species 'gibberella_zeae' (id = 229533)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 229533, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 229533, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 229533, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 229533, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 229533, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'PHIbase::MULTI' (id = 136)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (136, 229533, 'http://www.phi-base.org/PHI-export.php', \N, now(), now(), 'PHIbaseParser');

# Data from source 'RFAM::EG' (id = 152)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (152, 229533, 'script:object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 229533, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RefSeq_dna::MULTI-fungi' (id = 169)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (169, 229533, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-fungi' (id = 237)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (237, 229533, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 229533, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 229533, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 229533, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 229533, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 229533, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

# Data from source 'misc_EG::EG' (id = 427)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (427, 229533, 'script:host=>mysql-eg-pan-1,port=>4276,dbname=>EG_Xrefs,', \N, now(), now(), 'EG_DBParser');

#-------------------------------------------------------------------------------
# Data for species 'glycine_max' (id = 3847)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 3847, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 3847, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 3847, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 3847, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 3847, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'PHYTOZOME_GMAX_GENE::glycine_max' (id = 137)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (137, 3847, 'ftp://ftp.jgi-psf.org/pub/compgen/phytozome/v8.0/early_release/Gmax_v1.1/annotation/Gmax_v1.1_189_annotation_info.txt.gz', \N, now(), now(), 'PhytozomeGmaxParser');

# Data from source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (167, 3847, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (235, 3847, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 3847, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 3847, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 3847, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 3847, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'gorilla_gorilla' (id = 9595)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 9595, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 9595, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 9595, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 9595, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 9595, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 9595, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 9595, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 9595, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::gorilla_gorilla' (id = 194)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (194, 9595, 'ftp://ftp.ncbi.nih.gov/genomes/Gorilla_gorilla/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 9595, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::gorilla_gorilla' (id = 262)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (262, 9595, 'ftp://ftp.ncbi.nih.gov/genomes/Gorilla_gorilla/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 9595, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 9595, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 9595, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 9595, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 9595, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'heliconius_melpomene' (id = 34740)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 34740, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 34740, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 34740, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 34740, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 34740, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 34740, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 34740, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 34740, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 34740, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'helobdella_robusta' (id = 6412)
#-------------------------------------------------------------------------------

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 6412, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 6412, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::CORE' (id = 107)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (107, 6412, 'script:project=>ensemblgenomes,', \N, now(), now(), 'InterproFromCoreParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 6412, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 6412, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 6412, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 6412, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 6412, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 6412, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 6412, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'ncRNA_EG::EG' (id = 428)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (428, 6412, 'script:logic_name=>ncrna_eg,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

#-------------------------------------------------------------------------------
# Data for species 'homo_sapiens' (id = 9606)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 9606, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'CCDS::homo_sapiens' (id = 19)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (19, 9606, 'script:host=>ens-livemirror,dbname=>ccds_human_83,tran_name=>ENST,', \N, now(), now(), 'CCDSParser');

# Data from source 'DBASS3::homo_sapiens' (id = 33)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (33, 9606, 'http://www.dbass.soton.ac.uk/dbass3/download.aspx?item=genes', \N, now(), now(), 'DBASSParser');

# Data from source 'DBASS5::homo_sapiens' (id = 34)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (34, 9606, 'http://www.dbass.soton.ac.uk/dbass5/download.aspx?item=genes', \N, now(), now(), 'DBASSParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 9606, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::homo_sapiens' (id = 61)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (61, 9606, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/HUMAN/gene_association.goa_human.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'HGNC::homo_sapiens' (id = 80)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (80, 9606, 'script:source=>HGNC,host=>ens-staging1,', \N, now(), now(), 'VegaOfficialNameParser');

# Data from source 'HGNC::homo_sapiens#01' (id = 81)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (81, 9606, 'script:wget=>http://www.genenames.org/cgi-bin/download?title=HGNC+output+data&hgnc_dbtag=on&col=gd_hgnc_id&col=gd_status&col=gd_ccds_ids&status=Approved&status_opt=2&level=pri&=on&where=&order_by=gd_app_sym_sort&limit=&format=text&submit=submit&.cgifields=&.cgifields=level&.cgifields=chr&.cgifields=status&.cgifields=hgnc_dbtag,host=>ens-livemirror,dbname=>ccds_human_83,', \N, now(), now(), 'HGNC_CCDSParser');

# Data from source 'HGNC::homo_sapiens#02' (id = 82)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (82, 9606, 'http://www.genenames.org/cgi-bin/download?col=gd_hgnc_id&col=gd_app_sym&col=gd_app_name&col=gd_prev_sym&col=gd_aliases&col=gd_pub_eg_id&col=gd_pub_refseq_ids&col=md_eg_id&col=md_refseq_id&col=gd_pub_ensembl_id&col=md_prot_id&col=gd_lsdb_links&status=Approved&status_opt=3&=on&where=&order_by=gd_hgnc_id&limit=&format=text&hgnc_dbtag=on&submit=submit', \N, now(), now(), 'HGNCParser');

# Data from source 'HGNC_curated_transcript::homo_sapiens' (id = 90)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (90, 9606, 'script:source=>HGNC,host=>ens-staging1,', \N, now(), now(), 'curated_transcriptParser');

# Data from source 'HPA::homo_sapiens' (id = 93)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (93, 9606, 'http://www.proteinatlas.org/download/xref.php', \N, now(), now(), 'HPAParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 9606, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 9606, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'MIM2GENE::homo_sapiens' (id = 123)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (123, 9606, 'ftp://ftp.omim.org/OMIM/mim2gene.txt', \N, now(), now(), 'Mim2GeneParser');

# Data from source 'MIM::homo_sapiens' (id = 124)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (124, 9606, 'ftp://ftp.omim.org/OMIM/omim.txt.Z', \N, now(), now(), 'MIMParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 9606, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 9606, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 9606, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::homo_sapiens' (id = 195)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (195, 9606, 'script:host=>ens-livemirror,dbname=>ccds_human_83,', \N, now(), now(), 'RefSeq_CCDSParser');

# Data from source 'RefSeq_dna::homo_sapiens#01' (id = 196)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (196, 9606, 'ftp://ftp.ncbi.nih.gov/refseq/H_sapiens/mRNA_Prot/human*.rna.gbff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 9606, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::homo_sapiens' (id = 263)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (263, 9606, 'ftp://ftp.ncbi.nih.gov/refseq/H_sapiens/mRNA_Prot/human*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UCSC::homo_sapiens' (id = 306)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (306, 9606, 'ftp://hgdownload.cse.ucsc.edu/goldenPath/hg38/database/knownGene.txt.gz', 'ftp://hgdownload.cse.ucsc.edu/goldenPath/hg38/database/README.txt', now(), now(), 'UCSCParser');

# Data from source 'UniGene::homo_sapiens' (id = 330)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (330, 9606, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Homo_sapiens/Hs.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Homo_sapiens/Hs.data.gz', 'ftp://ftp.ncbi.nih.gov/repository/UniGene/*.LOG', now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 9606, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 9606, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 9606, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 9606, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 9606, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'hordeum_vulgare' (id = 112509)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 112509, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 112509, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 112509, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 112509, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 112509, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (167, 112509, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (235, 112509, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 112509, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 112509, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 112509, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 112509, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'ncRNA_EG::EG' (id = 428)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (428, 112509, 'script:logic_name=>ncrna_eg,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

#-------------------------------------------------------------------------------
# Data for species 'hyaloperonospora_arabidopsidis' (id = 559515)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 559515, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 559515, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 559515, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 559515, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 559515, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'PHIbase::MULTI' (id = 136)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (136, 559515, 'http://www.phi-base.org/PHI-export.php', \N, now(), now(), 'PHIbaseParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 559515, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 559515, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RefSeq_dna::MULTI-protozoa' (id = 172)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (172, 559515, 'ftp://ftp.ncbi.nih.gov/refseq/release/protozoa/protozoa*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-protozoa' (id = 240)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (240, 559515, 'ftp://ftp.ncbi.nih.gov/refseq/release/protozoa/protozoa*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 559515, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 559515, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 559515, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 559515, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'ictidomys_tridecemlineatus' (id = 43179)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 43179, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 43179, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 43179, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 43179, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 43179, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 43179, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 43179, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 43179, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::ictidomys_tridecemlineatus' (id = 197)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (197, 43179, 'ftp://ftp.ncbi.nih.gov/genomes/Ictidomys_tridecemlineatus/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 43179, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::ictidomys_tridecemlineatus' (id = 264)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (264, 43179, 'ftp://ftp.ncbi.nih.gov/genomes/Ictidomys_tridecemlineatus/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 43179, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 43179, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 43179, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 43179, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 43179, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'ixodes_scapularis' (id = 6945)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 6945, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 6945, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 6945, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 6945, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 6945, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'PUBMED::ixodes_scapularis' (id = 147)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (147, 6945, 'file:PubMed/PubMed_citations_Ixodes_scapularis_Wikel.txt', \N, now(), now(), 'VBPubMedParser');

# Data from source 'UniGene::ixodes_scapularis' (id = 332)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (332, 6945, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Ixodes_scapularis/Isc.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Ixodes_scapularis/Isc.data.gz', 'ftp://ftp.ncbi.nih.gov/repository/UniGene/*.LOG', now(), now(), 'UniGeneParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 6945, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 6945, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'VB_Community_Annotation::ixodes_scapularis' (id = 385)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (385, 6945, 'file:VB_Community_Annotation/Ixodes_scapularis_symbols.txt', \N, now(), now(), 'VBCommunitySymbolParser');

# Data from source 'VB_External_Description::ixodes_scapularis' (id = 391)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (391, 6945, 'file:VB_External_Description/IscaW1-1_gene_description.txt', \N, now(), now(), 'VBExternalDescriptionParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 6945, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'latimeria_chalumnae' (id = 7897)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 7897, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 7897, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 7897, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 7897, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 7897, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 7897, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 7897, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 7897, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::MULTI-vertebrate_other' (id = 174)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (174, 7897, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_other/vertebrate_other*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 7897, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::MULTI-vertebrate_other' (id = 242)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (242, 7897, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_other/vertebrate_other*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 7897, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 7897, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 7897, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 7897, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 7897, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'leersia_perrieri' (id = 77586)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 77586, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 77586, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 77586, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 77586, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 77586, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::EG' (id = 152)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (152, 77586, 'script:object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 77586, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (167, 77586, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_mRNA_predicted::MULTI' (id = 229)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (229, 77586, '', \N, now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (235, 77586, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 77586, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 77586, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 77586, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 77586, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 77586, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 77586, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'leishmania_major' (id = 5664)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 5664, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 5664, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 5664, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 5664, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 5664, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 5664, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 5664, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RefSeq_dna::MULTI-protozoa' (id = 172)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (172, 5664, 'ftp://ftp.ncbi.nih.gov/refseq/release/protozoa/protozoa*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-protozoa' (id = 240)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (240, 5664, 'ftp://ftp.ncbi.nih.gov/refseq/release/protozoa/protozoa*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 5664, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 5664, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 5664, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 5664, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'lepisosteus_oculatus' (id = 7918)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 7918, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 7918, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 7918, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 7918, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 7918, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 7918, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 7918, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 7918, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::lepisosteus_oculatus' (id = 198)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (198, 7918, 'ftp://ftp.ncbi.nih.gov/genomes/Lepisosteus_oculatus/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 7918, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::lepisosteus_oculatus' (id = 265)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (265, 7918, 'ftp://ftp.ncbi.nih.gov/genomes/Lepisosteus_oculatus/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 7918, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 7918, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 7918, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 7918, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 7918, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'lottia_gigantea' (id = 225164)
#-------------------------------------------------------------------------------

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 225164, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 225164, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::CORE' (id = 107)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (107, 225164, 'script:project=>ensemblgenomes,', \N, now(), now(), 'InterproFromCoreParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 225164, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 225164, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 225164, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 225164, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 225164, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 225164, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 225164, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'ncRNA_EG::EG' (id = 428)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (428, 225164, 'script:logic_name=>ncrna_eg,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

#-------------------------------------------------------------------------------
# Data for species 'loxodonta_africana' (id = 9785)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 9785, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 9785, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 9785, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 9785, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 9785, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 9785, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 9785, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 9785, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::loxodonta_africana' (id = 199)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (199, 9785, 'ftp://ftp.ncbi.nih.gov/genomes/Loxodonta_africana/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 9785, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::loxodonta_africana' (id = 266)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (266, 9785, 'ftp://ftp.ncbi.nih.gov/genomes/Loxodonta_africana/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 9785, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 9785, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 9785, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 9785, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 9785, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'macaca_mulatta' (id = 9544)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 9544, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 9544, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 9544, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 9544, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 9544, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 9544, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 9544, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 9544, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::macaca_mulatta' (id = 200)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (200, 9544, 'ftp://ftp.ncbi.nih.gov/genomes/Macaca_mulatta/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 9544, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::macaca_mulatta' (id = 267)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (267, 9544, 'ftp://ftp.ncbi.nih.gov/genomes/Macaca_mulatta/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::macaca_mulatta' (id = 333)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (333, 9544, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Macaca_mulatta/Mmu.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Macaca_mulatta/Mmu.data.gz', 'ftp://ftp.ncbi.nih.gov/repository/UniGene/*.LOG', now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 9544, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 9544, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 9544, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 9544, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 9544, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'macropus_eugenii' (id = 9315)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 9315, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 9315, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 9315, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 9315, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 9315, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 9315, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 9315, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 9315, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::MULTI-vertebrate_mammalian' (id = 173)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (173, 9315, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_mammalian/vertebrate_mammalian*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 9315, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::MULTI-vertebrate_mammalian' (id = 241)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (241, 9315, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_mammalian/vertebrate_mammalian*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 9315, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 9315, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 9315, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 9315, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 9315, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'magnaporthe_oryzae' (id = 242507)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 242507, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 242507, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 242507, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 242507, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 242507, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'PHIbase::MULTI' (id = 136)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (136, 242507, 'http://www.phi-base.org/PHI-export.php', \N, now(), now(), 'PHIbaseParser');

# Data from source 'RFAM::EG' (id = 152)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (152, 242507, 'script:object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 242507, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RefSeq_dna::MULTI-fungi' (id = 169)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (169, 242507, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-fungi' (id = 237)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (237, 242507, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 242507, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 242507, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 242507, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 242507, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 242507, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

# Data from source 'misc_EG::EG' (id = 427)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (427, 242507, 'script:host=>mysql-eg-pan-1,port=>4276,dbname=>EG_Xrefs,', \N, now(), now(), 'EG_DBParser');

#-------------------------------------------------------------------------------
# Data for species 'medicago_truncatula' (id = 3880)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 3880, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 3880, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 3880, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 3880, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 3880, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (167, 3880, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_mRNA_predicted::MULTI' (id = 229)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (229, 3880, '', \N, now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (235, 3880, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 3880, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 3880, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 3880, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 3880, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'ncRNA_EG::EG' (id = 428)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (428, 3880, 'script:logic_name=>ncrna_eg,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

#-------------------------------------------------------------------------------
# Data for species 'megaselia_scalaris' (id = 36166)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 36166, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 36166, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 36166, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 36166, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 36166, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::EG' (id = 152)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (152, 36166, 'script:object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 36166, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 36166, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 36166, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 36166, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 36166, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 36166, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 36166, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'meleagris_gallopavo' (id = 9103)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 9103, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 9103, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 9103, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 9103, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 9103, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 9103, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 9103, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 9103, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::meleagris_gallopavo' (id = 201)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (201, 9103, 'ftp://ftp.ncbi.nih.gov/genomes/Meleagris_gallopavo/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 9103, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::meleagris_gallopavo' (id = 268)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (268, 9103, 'ftp://ftp.ncbi.nih.gov/genomes/Meleagris_gallopavo/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::meleagris_gallopavo' (id = 334)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (334, 9103, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Meleagris_gallopavo/Mga.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Meleagris_gallopavo/Mga.data.gz', 'ftp://ftp.ncbi.nih.gov/repository/UniGene/*.LOG', now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 9103, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 9103, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 9103, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 9103, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 9103, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'melitaea_cinxia' (id = 113334)
#-------------------------------------------------------------------------------

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 113334, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 113334, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 113334, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 113334, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 113334, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 113334, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'ncRNA_EG::EG' (id = 428)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (428, 113334, 'script:logic_name=>ncrna_eg,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

#-------------------------------------------------------------------------------
# Data for species 'microcebus_murinus' (id = 30608)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 30608, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 30608, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 30608, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 30608, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 30608, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 30608, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 30608, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 30608, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::MULTI-vertebrate_mammalian' (id = 173)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (173, 30608, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_mammalian/vertebrate_mammalian*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 30608, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::MULTI-vertebrate_mammalian' (id = 241)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (241, 30608, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_mammalian/vertebrate_mammalian*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 30608, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 30608, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 30608, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 30608, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 30608, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'mnemiopsis_leidyi' (id = 27923)
#-------------------------------------------------------------------------------

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 27923, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 27923, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::CORE' (id = 107)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (107, 27923, 'script:project=>ensemblgenomes,', \N, now(), now(), 'InterproFromCoreParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 27923, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 27923, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 27923, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 27923, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 27923, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 27923, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 27923, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'ncRNA_EG::EG' (id = 428)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (428, 27923, 'script:logic_name=>ncrna_eg,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

#-------------------------------------------------------------------------------
# Data for species 'monodelphis_domestica' (id = 13616)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 13616, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 13616, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 13616, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 13616, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 13616, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 13616, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 13616, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 13616, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::monodelphis_domestica' (id = 202)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (202, 13616, 'ftp://ftp.ncbi.nih.gov/genomes/Monodelphis_domestica/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 13616, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::monodelphis_domestica' (id = 269)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (269, 13616, 'ftp://ftp.ncbi.nih.gov/genomes/Monodelphis_domestica/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::monodelphis_domestica' (id = 335)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (335, 13616, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Monodelphis_domestica/Mdm.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Monodelphis_domestica/Mdm.data.gz', 'ftp://ftp.ncbi.nih.gov/repository/UniGene/*.LOG', now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 13616, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 13616, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 13616, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 13616, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 13616, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'mus_musculus' (id = 10090)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 10090, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'CCDS::mus_musculus' (id = 20)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (20, 10090, 'script:host=>ens-livemirror,dbname=>ccds_mouse_83,tran_name=>ENSMUST,', \N, now(), now(), 'CCDSParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 10090, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::mus_musculus' (id = 62)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (62, 10090, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/MOUSE/gene_association.goa_mouse.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'IKMCs_KOs::mus_musculus' (id = 95)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (95, 10090, 'script:', \N, now(), now(), 'IKMCParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 10090, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 10090, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'MGI::mus_musculus#01' (id = 114)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (114, 10090, 'ftp://ftp.informatics.jax.org/pub/reports/MRK_ENSEMBL.rpt', \N, now(), now(), 'MGIParser');

# Data from source 'MGI::mus_musculus#02' (id = 115)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (115, 10090, 'script:source=>MGI,host=>ens-staging2,', \N, now(), now(), 'VegaOfficialNameParser');

# Data from source 'MGI::mus_musculus#03' (id = 116)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (116, 10090, 'script:wget=>ftp://ftp.ncbi.nlm.nih.gov/pub/CCDS/current_mouse/CCDS.current.txt,', \N, now(), now(), 'MGI_CCDS_Parser');

# Data from source 'MGI::mus_musculus#05' (id = 118)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (118, 10090, 'ftp://ftp.informatics.jax.org/pub/reports/MRK_List2.rpt', \N, now(), now(), 'MGI_Desc_Parser');

# Data from source 'MGI_curated_transcript::mus_musculus' (id = 120)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (120, 10090, 'script:source=>MGI,host=>ens-staging2,', \N, now(), now(), 'curated_transcriptParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 10090, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 10090, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 10090, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::mus_musculus' (id = 203)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (203, 10090, 'script:host=>ens-livemirror,dbname=>ccds_mouse_83,', \N, now(), now(), 'RefSeq_CCDSParser');

# Data from source 'RefSeq_dna::mus_musculus#01' (id = 204)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (204, 10090, 'ftp://ftp.ncbi.nih.gov/refseq/M_musculus/mRNA_Prot/mouse*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 10090, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::mus_musculus' (id = 270)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (270, 10090, 'ftp://ftp.ncbi.nih.gov/refseq/M_musculus/mRNA_Prot/mouse*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UCSC::mus_musculus' (id = 307)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (307, 10090, 'ftp://hgdownload.cse.ucsc.edu/goldenPath/mm10/database/knownGene.txt.gz', 'ftp://hgdownload.cse.ucsc.edu/goldenPath/mm10/database/README.txt', now(), now(), 'UCSCParser');

# Data from source 'UniGene::mus_musculus' (id = 336)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (336, 10090, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Mus_musculus/Mm.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Mus_musculus/Mm.data.gz', 'ftp://ftp.ncbi.nih.gov/repository/UniGene/*.LOG', now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 10090, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 10090, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 10090, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 10090, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 10090, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'musa_acuminata' (id = 214687)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 214687, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 214687, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 214687, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 214687, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 214687, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (167, 214687, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (235, 214687, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 214687, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 214687, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 214687, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 214687, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'mustela_putorius_furo' (id = 9669)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 9669, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 9669, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 9669, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 9669, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 9669, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 9669, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 9669, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 9669, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::mustela_putorius_furo' (id = 205)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (205, 9669, 'ftp://ftp.ncbi.nih.gov/genomes/Mustela_putorius_furo/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 9669, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::mustela_putorius_furo' (id = 271)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (271, 9669, 'ftp://ftp.ncbi.nih.gov/genomes/Mustela_putorius_furo/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 9669, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 9669, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 9669, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 9669, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 9669, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'mycosphaerella_graminicola' (id = 336722)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 336722, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 336722, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 336722, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 336722, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 336722, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'PHIbase::MULTI' (id = 136)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (136, 336722, 'http://www.phi-base.org/PHI-export.php', \N, now(), now(), 'PHIbaseParser');

# Data from source 'RFAM::EG' (id = 152)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (152, 336722, 'script:object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 336722, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RefSeq_dna::MULTI-fungi' (id = 169)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (169, 336722, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-fungi' (id = 237)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (237, 336722, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 336722, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 336722, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 336722, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 336722, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 336722, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

# Data from source 'misc_EG::EG' (id = 427)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (427, 336722, 'script:host=>mysql-eg-pan-1,port=>4276,dbname=>EG_Xrefs,', \N, now(), now(), 'EG_DBParser');

#-------------------------------------------------------------------------------
# Data for species 'myotis_lucifugus' (id = 59463)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 59463, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 59463, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 59463, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 59463, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 59463, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 59463, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 59463, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 59463, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::MULTI-vertebrate_mammalian' (id = 173)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (173, 59463, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_mammalian/vertebrate_mammalian*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 59463, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::MULTI-vertebrate_mammalian' (id = 241)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (241, 59463, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_mammalian/vertebrate_mammalian*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 59463, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 59463, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 59463, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 59463, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 59463, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'nasonia_vitripennis' (id = 7425)
#-------------------------------------------------------------------------------

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 7425, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 7425, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::CORE' (id = 107)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (107, 7425, 'script:project=>ensemblgenomes,', \N, now(), now(), 'InterproFromCoreParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 7425, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 7425, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 7425, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 7425, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 7425, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 7425, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 7425, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'ncRNA_EG::EG' (id = 428)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (428, 7425, 'script:logic_name=>ncrna_eg,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

#-------------------------------------------------------------------------------
# Data for species 'nectria_haematococca' (id = 140110)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 140110, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 140110, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 140110, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 140110, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 140110, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'PHIbase::MULTI' (id = 136)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (136, 140110, 'http://www.phi-base.org/PHI-export.php', \N, now(), now(), 'PHIbaseParser');

# Data from source 'RFAM::EG' (id = 152)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (152, 140110, 'script:object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 140110, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RefSeq_dna::MULTI-fungi' (id = 169)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (169, 140110, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-fungi' (id = 237)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (237, 140110, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 140110, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 140110, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 140110, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 140110, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 140110, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'nematostella_vectensis' (id = 45351)
#-------------------------------------------------------------------------------

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 45351, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 45351, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::CORE' (id = 107)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (107, 45351, 'script:project=>ensemblgenomes,', \N, now(), now(), 'InterproFromCoreParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 45351, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 45351, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 45351, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::nematostella_vectensis' (id = 337)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (337, 45351, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Nematostella_vectensis/Nve.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Nematostella_vectensis/Nve.data.gz', \N, now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 45351, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 45351, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 45351, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 45351, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'ncRNA_EG::EG' (id = 428)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (428, 45351, 'script:logic_name=>ncrna_eg,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

#-------------------------------------------------------------------------------
# Data for species 'neosartorya_fischeri' (id = 36630)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 36630, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 36630, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 36630, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 36630, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 36630, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 36630, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 36630, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RefSeq_dna::MULTI-fungi' (id = 169)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (169, 36630, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-fungi' (id = 237)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (237, 36630, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 36630, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 36630, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 36630, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 36630, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'neurospora_crassa' (id = 5141)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 5141, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 5141, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 5141, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 5141, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 5141, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::EG' (id = 152)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (152, 5141, 'script:object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 5141, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RefSeq_dna::MULTI-fungi' (id = 169)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (169, 5141, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-fungi' (id = 237)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (237, 5141, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 5141, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 5141, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 5141, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 5141, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 5141, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

# Data from source 'misc_EG::EG' (id = 427)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (427, 5141, 'script:host=>mysql-eg-pan-1,port=>4276,dbname=>EG_Xrefs,', \N, now(), now(), 'EG_DBParser');

#-------------------------------------------------------------------------------
# Data for species 'nomascus_leucogenys' (id = 61853)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 61853, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 61853, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 61853, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 61853, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 61853, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 61853, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 61853, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 61853, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::MULTI-vertebrate_mammalian' (id = 173)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (173, 61853, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_mammalian/vertebrate_mammalian*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 61853, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::MULTI-vertebrate_mammalian' (id = 241)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (241, 61853, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_mammalian/vertebrate_mammalian*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 61853, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'UniProt::protein_id' (id = 364)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (364, 61853, '', \N, now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 61853, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 61853, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot::EMBL' (id = 377)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (377, 61853, '', \N, now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 61853, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 61853, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'ochotona_princeps' (id = 9978)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 9978, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 9978, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 9978, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 9978, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 9978, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 9978, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 9978, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 9978, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::ochotona_princeps' (id = 206)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (206, 9978, 'ftp://ftp.ncbi.nih.gov/genomes/Ochotona_princeps/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 9978, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::ochotona_princeps' (id = 272)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (272, 9978, 'ftp://ftp.ncbi.nih.gov/genomes/Ochotona_princeps/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 9978, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 9978, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 9978, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 9978, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 9978, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'onchocerca_volvulus' (id = 6282)
#-------------------------------------------------------------------------------

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 6282, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 6282, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 6282, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 6282, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 6282, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 6282, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI-invertebrate' (id = 369)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (369, 6282, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/uniprot_trembl_invertebrates.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI-invertebrate' (id = 374)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (374, 6282, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/uniprot_sprot_invertebrates.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'wormbase::ovolvulus' (id = 446)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (446, 6282, 'ftp://ftp.ebi.ac.uk/pub/databases/wormbase/releases/RELEASE/species/o_volvulus/PRJEB513/o_volvulus.PRJEB513.RELEASE.xrefs.txt.gz', \N, now(), now(), 'WormbaseDirectParser');

#-------------------------------------------------------------------------------
# Data for species 'oreochromis_niloticus' (id = 8128)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 8128, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 8128, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 8128, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 8128, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 8128, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 8128, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 8128, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 8128, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::oreochromis_niloticus' (id = 207)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (207, 8128, 'ftp://ftp.ncbi.nih.gov/genomes/Oreochromis_niloticus/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 8128, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::oreochromis_niloticus' (id = 273)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (273, 8128, 'ftp://ftp.ncbi.nih.gov/genomes/Oreochromis_niloticus/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::oreochromis_niloticus' (id = 338)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (338, 8128, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Oreochromis_niloticus/Oni.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Oreochromis_niloticus/Oni.data.gz', 'ftp://ftp.ncbi.nih.gov/repository/UniGene/*.LOG', now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 8128, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 8128, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 8128, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 8128, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 8128, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'ornithorhynchus_anatinus' (id = 9258)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 9258, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 9258, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 9258, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 9258, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 9258, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 9258, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 9258, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 9258, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::ornithorhynchus_anatinus' (id = 208)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (208, 9258, 'ftp://ftp.ncbi.nih.gov/genomes/Ornithorhynchus_anatinus/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 9258, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::ornithorhynchus_anatinus' (id = 274)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (274, 9258, 'ftp://ftp.ncbi.nih.gov/genomes/Ornithorhynchus_anatinus/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::ornithorhynchus_anatinus' (id = 339)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (339, 9258, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Ornithorhynchus_anatinus/Oan.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Ornithorhynchus_anatinus/Oan.data.gz', 'ftp://ftp.ncbi.nih.gov/repository/UniGene/*.LOG', now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 9258, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 9258, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 9258, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 9258, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 9258, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'oryctolagus_cuniculus' (id = 9986)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 9986, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 9986, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 9986, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 9986, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 9986, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 9986, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 9986, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 9986, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::oryctolagus_cuniculus' (id = 209)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (209, 9986, 'ftp://ftp.ncbi.nih.gov/genomes/Oryctolagus_cuniculus/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 9986, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::oryctolagus_cuniculus' (id = 275)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (275, 9986, 'ftp://ftp.ncbi.nih.gov/genomes/Oryctolagus_cuniculus/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::oryctolagus_cuniculus' (id = 340)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (340, 9986, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Oryctolagus_cuniculus/Ocu.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Oryctolagus_cuniculus/Ocu.data.gz', \N, now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 9986, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 9986, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 9986, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 9986, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 9986, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'oryza_barthii' (id = 65489)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 65489, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 65489, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 65489, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 65489, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 65489, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (167, 65489, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_mRNA_predicted::MULTI' (id = 229)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (229, 65489, '', \N, now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (235, 65489, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 65489, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 65489, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 65489, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 65489, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'oryza_brachyantha' (id = 4533)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 4533, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 4533, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 4533, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 4533, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 4533, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (167, 4533, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (235, 4533, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 4533, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 4533, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 4533, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 4533, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'oryza_glaberrima' (id = 4538)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 4538, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 4538, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 4538, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 4538, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 4538, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (167, 4538, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (235, 4538, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 4538, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 4538, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 4538, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 4538, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'oryza_glumaepatula' (id = 40148)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 40148, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 40148, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 40148, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 40148, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 40148, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (167, 40148, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_mRNA_predicted::MULTI' (id = 229)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (229, 40148, '', \N, now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (235, 40148, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 40148, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 40148, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 40148, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 40148, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'oryza_granulata' (id = 110450)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 110450, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 110450, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 110450, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 110450, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 110450, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (167, 110450, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_mRNA_predicted::MULTI' (id = 229)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (229, 110450, '', \N, now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (235, 110450, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 110450, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 110450, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 110450, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 110450, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'oryza_indica' (id = 39946)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 39946, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 39946, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 39946, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 39946, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 39946, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (167, 39946, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_mRNA_predicted::MULTI' (id = 229)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (229, 39946, '', \N, now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (235, 39946, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 39946, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 39946, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 39946, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 39946, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'oryza_longistaminata' (id = 4528)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 4528, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 4528, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 4528, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 4528, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 4528, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (167, 4528, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_mRNA_predicted::MULTI' (id = 229)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (229, 4528, '', \N, now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (235, 4528, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 4528, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 4528, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 4528, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 4528, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'oryza_meridionalis' (id = 40149)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 40149, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 40149, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 40149, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 40149, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 40149, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (167, 40149, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_mRNA_predicted::MULTI' (id = 229)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (229, 40149, '', \N, now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (235, 40149, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 40149, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 40149, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 40149, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 40149, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'oryza_nivara' (id = 4536)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 4536, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 4536, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 4536, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 4536, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 4536, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (167, 4536, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_mRNA_predicted::MULTI' (id = 229)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (229, 4536, '', \N, now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (235, 4536, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 4536, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 4536, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 4536, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 4536, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'oryza_officinalis' (id = 4535)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 4535, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 4535, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 4535, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 4535, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 4535, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (167, 4535, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_mRNA_predicted::MULTI' (id = 229)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (229, 4535, '', \N, now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (235, 4535, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 4535, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 4535, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 4535, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 4535, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'oryza_punctata' (id = 4537)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 4537, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 4537, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 4537, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 4537, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 4537, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (167, 4537, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_mRNA_predicted::MULTI' (id = 229)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (229, 4537, '', \N, now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (235, 4537, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 4537, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 4537, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 4537, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 4537, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'oryza_sativa' (id = 39947)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 39947, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 39947, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 39947, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Gramene_Pathway::oryza_sativa' (id = 74)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (74, 39947, 'ftp://ftp.gramene.org/pub/gramene/ensemblgenomes/xrefs/pathways/ricecyc.tab.gz', \N, now(), now(), 'GramenePathwayParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 39947, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 39947, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (167, 39947, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_mRNA_predicted::MULTI' (id = 229)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (229, 39947, '', \N, now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (235, 39947, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 39947, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 39947, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 39947, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 39947, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'oryzias_latipes' (id = 8090)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 8090, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 8090, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 8090, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 8090, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 8090, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 8090, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 8090, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 8090, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::oryzias_latipes' (id = 210)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (210, 8090, 'ftp://ftp.ncbi.nih.gov/genomes/Oryzias_latipes/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 8090, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::oryzias_latipes' (id = 276)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (276, 8090, 'ftp://ftp.ncbi.nih.gov/genomes/Oryzias_latipes/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::oryzias_latipes' (id = 341)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (341, 8090, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Oryzias_latipes/Ola.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Oryzias_latipes/Ola.data.gz', 'ftp://ftp.ncbi.nih.gov/repository/UniGene/*.LOG', now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 8090, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 8090, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 8090, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 8090, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 8090, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'otolemur_garnettii' (id = 30611)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 30611, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 30611, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 30611, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 30611, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 30611, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 30611, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 30611, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 30611, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::otolemur_garnettii' (id = 211)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (211, 30611, 'ftp://ftp.ncbi.nih.gov/genomes/Otolemur_garnettii/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 30611, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::otolemur_garnettii' (id = 278)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (278, 30611, 'ftp://ftp.ncbi.nih.gov/genomes/Otolemur_garnettii/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 30611, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 30611, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 30611, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 30611, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 30611, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'ovis_aries' (id = 9940)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 9940, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 9940, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 9940, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 9940, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 9940, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 9940, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 9940, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 9940, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::ovis_aries' (id = 212)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (212, 9940, 'ftp://ftp.ncbi.nih.gov/genomes/Ovis_aries/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 9940, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::ovis_aries' (id = 279)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (279, 9940, 'ftp://ftp.ncbi.nih.gov/genomes/Ovis_aries/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::ovis_aries' (id = 342)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (342, 9940, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Ovis_aries/Oar.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Ovis_aries/Oar.data.gz', 'ftp://ftp.ncbi.nih.gov/repository/UniGene/*.LOG', now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 9940, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 9940, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 9940, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 9940, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 9940, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'pan_troglodytes' (id = 9598)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 9598, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 9598, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 9598, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 9598, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 9598, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 9598, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 9598, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 9598, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::pan_troglodytes' (id = 213)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (213, 9598, 'ftp://ftp.ncbi.nih.gov/genomes/Pan_troglodytes/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 9598, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::pan_troglodytes' (id = 280)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (280, 9598, 'ftp://ftp.ncbi.nih.gov/genomes/Pan_troglodytes/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::pan_troglodytes' (id = 343)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (343, 9598, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Pan_troglodytes/Ptr.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Pan_troglodytes/Ptr.data.gz', 'ftp://ftp.ncbi.nih.gov/repository/UniGene/*.LOG', now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 9598, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 9598, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 9598, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 9598, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 9598, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'papio_anubis' (id = 9555)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 9555, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 9555, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 9555, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 9555, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 9555, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 9555, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 9555, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 9555, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::papio_anubis' (id = 214)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (214, 9555, 'ftp://ftp.ncbi.nih.gov/genomes/Papio_anubis/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 9555, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::papio_anubis' (id = 281)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (281, 9555, 'ftp://ftp.ncbi.nih.gov/genomes/Papio_anubis/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::papio_anubis' (id = 344)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (344, 9555, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Papio_anubis/Pan.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Papio_anubis/Pan.data.gz', 'ftp://ftp.ncbi.nih.gov/repository/UniGene/*.LOG', now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 9555, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 9555, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 9555, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 9555, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 9555, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'pediculus_humanus' (id = 121225)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 121225, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 121225, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 121225, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 121225, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 121225, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'UniGene::aedes_aegypti' (id = 309)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (309, 121225, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Aedes_aegypti/Aae.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Aedes_aegypti/Aae.data.gz', \N, now(), now(), 'UniGeneParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 121225, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 121225, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 121225, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'pelodiscus_sinensis' (id = 13735)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 13735, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 13735, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 13735, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 13735, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 13735, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 13735, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 13735, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 13735, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::MULTI-vertebrate_other' (id = 174)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (174, 13735, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_other/vertebrate_other*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 13735, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::MULTI-vertebrate_other' (id = 242)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (242, 13735, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_other/vertebrate_other*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 13735, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 13735, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 13735, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 13735, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 13735, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'petromyzon_marinus' (id = 7757)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 7757, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 7757, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 7757, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 7757, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 7757, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 7757, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 7757, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 7757, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::MULTI-vertebrate_other' (id = 174)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (174, 7757, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_other/vertebrate_other*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 7757, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::MULTI-vertebrate_other' (id = 242)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (242, 7757, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_other/vertebrate_other*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::petromyzon_marinus' (id = 345)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (345, 7757, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Petromyzon_marinus/Pma.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Petromyzon_marinus/Pma.data.gz', 'ftp://ftp.ncbi.nih.gov/repository/UniGene/*.LOG', now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 7757, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 7757, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 7757, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 7757, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 7757, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'phaeodactylum_tricornutum' (id = 2850)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 2850, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 2850, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 2850, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 2850, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 2850, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::EG' (id = 152)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (152, 2850, 'script:object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 2850, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RefSeq_dna::MULTI-complete' (id = 168)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (168, 2850, 'ftp://ftp.ncbi.nih.gov/refseq/release/complete/complete*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-complete' (id = 236)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (236, 2850, 'ftp://ftp.ncbi.nih.gov/refseq/release/complete/complete*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 2850, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 2850, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 2850, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 2850, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 2850, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

# Data from source 'misc_EG::EG' (id = 427)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (427, 2850, 'script:host=>mysql-eg-pan-1,port=>4276,dbname=>EG_Xrefs,', \N, now(), now(), 'EG_DBParser');

#-------------------------------------------------------------------------------
# Data for species 'phaeosphaeria_nodorum' (id = 321614)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 321614, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 321614, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 321614, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 321614, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 321614, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'PHIbase::MULTI' (id = 136)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (136, 321614, 'http://www.phi-base.org/PHI-export.php', \N, now(), now(), 'PHIbaseParser');

# Data from source 'RefSeq_dna::MULTI-fungi' (id = 169)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (169, 321614, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-fungi' (id = 237)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (237, 321614, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 321614, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 321614, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 321614, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'physcomitrella_patens' (id = 145481)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 145481, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 145481, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 145481, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 145481, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 145481, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (167, 145481, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_mRNA_predicted::MULTI' (id = 229)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (229, 145481, '', \N, now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (235, 145481, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::physcomitrella_patens' (id = 346)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (346, 145481, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Physcomitrella_patens/Ppa.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Physcomitrella_patens/Ppa.data.gz', \N, now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 145481, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 145481, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 145481, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 145481, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'physeter_macrocephalus' (id = 9755)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 9755, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 9755, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 9755, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 9755, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 9755, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 9755, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RefSeq_dna::MULTI-vertebrate_other' (id = 174)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (174, 9755, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_other/vertebrate_other*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 9755, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::MULTI-vertebrate_other' (id = 242)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (242, 9755, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_other/vertebrate_other*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 9755, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 9755, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 9755, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 9755, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 9755, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'phytophthora_infestans' (id = 403677)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 403677, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 403677, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 403677, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 403677, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 403677, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'PHIbase::MULTI' (id = 136)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (136, 403677, 'http://www.phi-base.org/PHI-export.php', \N, now(), now(), 'PHIbaseParser');

# Data from source 'RFAM::EG' (id = 152)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (152, 403677, 'script:object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 403677, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RefSeq_dna::MULTI-complete' (id = 168)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (168, 403677, 'ftp://ftp.ncbi.nih.gov/refseq/release/complete/complete*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-complete' (id = 236)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (236, 403677, 'ftp://ftp.ncbi.nih.gov/refseq/release/complete/complete*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 403677, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 403677, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 403677, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 403677, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 403677, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

# Data from source 'misc_EG::EG' (id = 427)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (427, 403677, 'script:host=>mysql-eg-pan-1,port=>4276,dbname=>EG_Xrefs,', \N, now(), now(), 'EG_DBParser');

#-------------------------------------------------------------------------------
# Data for species 'phytophthora_ramorum' (id = 164328)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 164328, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 164328, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 164328, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 164328, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 164328, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'PHIbase::MULTI' (id = 136)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (136, 164328, 'http://www.phi-base.org/PHI-export.php', \N, now(), now(), 'PHIbaseParser');

# Data from source 'RFAM::EG' (id = 152)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (152, 164328, 'script:object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 164328, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RefSeq_dna::MULTI-complete' (id = 168)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (168, 164328, 'ftp://ftp.ncbi.nih.gov/refseq/release/complete/complete*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-complete' (id = 236)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (236, 164328, 'ftp://ftp.ncbi.nih.gov/refseq/release/complete/complete*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 164328, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 164328, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 164328, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 164328, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 164328, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

# Data from source 'misc_EG::EG' (id = 427)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (427, 164328, 'script:host=>mysql-eg-pan-1,port=>4276,dbname=>EG_Xrefs,', \N, now(), now(), 'EG_DBParser');

#-------------------------------------------------------------------------------
# Data for species 'phytophthora_sojae' (id = 67593)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 67593, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 67593, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 67593, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 67593, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 67593, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'PHIbase::MULTI' (id = 136)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (136, 67593, 'http://www.phi-base.org/PHI-export.php', \N, now(), now(), 'PHIbaseParser');

# Data from source 'RFAM::EG' (id = 152)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (152, 67593, 'script:object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 67593, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RefSeq_dna::MULTI-complete' (id = 168)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (168, 67593, 'ftp://ftp.ncbi.nih.gov/refseq/release/complete/complete*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-complete' (id = 236)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (236, 67593, 'ftp://ftp.ncbi.nih.gov/refseq/release/complete/complete*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 67593, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 67593, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 67593, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 67593, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 67593, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

# Data from source 'misc_EG::EG' (id = 427)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (427, 67593, 'script:host=>mysql-eg-pan-1,port=>4276,dbname=>EG_Xrefs,', \N, now(), now(), 'EG_DBParser');

#-------------------------------------------------------------------------------
# Data for species 'plasmodium_berghei' (id = 5823)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 5823, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 5823, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 5823, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 5823, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 5823, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::EG' (id = 152)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (152, 5823, 'script:object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 5823, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RefSeq_dna::MULTI-complete' (id = 168)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (168, 5823, 'ftp://ftp.ncbi.nih.gov/refseq/release/complete/complete*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-complete' (id = 236)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (236, 5823, 'ftp://ftp.ncbi.nih.gov/refseq/release/complete/complete*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 5823, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 5823, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 5823, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 5823, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 5823, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

# Data from source 'misc_EG::EG' (id = 427)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (427, 5823, 'script:host=>mysql-eg-pan-1,port=>4276,dbname=>EG_Xrefs,', \N, now(), now(), 'EG_DBParser');

#-------------------------------------------------------------------------------
# Data for species 'plasmodium_chabaudi' (id = 31271)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 31271, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 31271, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 31271, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 31271, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 31271, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::EG' (id = 152)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (152, 31271, 'script:object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 31271, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RefSeq_dna::MULTI-complete' (id = 168)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (168, 31271, 'ftp://ftp.ncbi.nih.gov/refseq/release/complete/complete*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-complete' (id = 236)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (236, 31271, 'ftp://ftp.ncbi.nih.gov/refseq/release/complete/complete*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 31271, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 31271, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 31271, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 31271, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 31271, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

# Data from source 'misc_EG::EG' (id = 427)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (427, 31271, 'script:host=>mysql-eg-pan-1,port=>4276,dbname=>EG_Xrefs,', \N, now(), now(), 'EG_DBParser');

#-------------------------------------------------------------------------------
# Data for species 'plasmodium_falciparum' (id = 5833)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 5833, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 5833, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 5833, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 5833, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 5833, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::EG' (id = 152)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (152, 5833, 'script:object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 5833, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RefSeq_dna::MULTI-complete' (id = 168)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (168, 5833, 'ftp://ftp.ncbi.nih.gov/refseq/release/complete/complete*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-complete' (id = 236)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (236, 5833, 'ftp://ftp.ncbi.nih.gov/refseq/release/complete/complete*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 5833, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 5833, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 5833, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 5833, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 5833, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

# Data from source 'misc_EG::EG' (id = 427)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (427, 5833, 'script:host=>mysql-eg-pan-1,port=>4276,dbname=>EG_Xrefs,', \N, now(), now(), 'EG_DBParser');

#-------------------------------------------------------------------------------
# Data for species 'plasmodium_knowlesi' (id = 5851)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 5851, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 5851, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 5851, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 5851, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 5851, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::EG' (id = 152)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (152, 5851, 'script:object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 5851, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RefSeq_dna::MULTI-complete' (id = 168)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (168, 5851, 'ftp://ftp.ncbi.nih.gov/refseq/release/complete/complete*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-complete' (id = 236)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (236, 5851, 'ftp://ftp.ncbi.nih.gov/refseq/release/complete/complete*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 5851, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 5851, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 5851, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 5851, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 5851, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

# Data from source 'misc_EG::EG' (id = 427)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (427, 5851, 'script:host=>mysql-eg-pan-1,port=>4276,dbname=>EG_Xrefs,', \N, now(), now(), 'EG_DBParser');

#-------------------------------------------------------------------------------
# Data for species 'plasmodium_vivax' (id = 5855)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 5855, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 5855, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 5855, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 5855, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 5855, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::EG' (id = 152)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (152, 5855, 'script:object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 5855, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RefSeq_dna::MULTI-complete' (id = 168)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (168, 5855, 'ftp://ftp.ncbi.nih.gov/refseq/release/complete/complete*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-complete' (id = 236)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (236, 5855, 'ftp://ftp.ncbi.nih.gov/refseq/release/complete/complete*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 5855, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 5855, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 5855, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 5855, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 5855, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

# Data from source 'misc_EG::EG' (id = 427)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (427, 5855, 'script:host=>mysql-eg-pan-1,port=>4276,dbname=>EG_Xrefs,', \N, now(), now(), 'EG_DBParser');

#-------------------------------------------------------------------------------
# Data for species 'poecilia_formosa' (id = 48698)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 48698, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 48698, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 48698, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 48698, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 48698, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 48698, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 48698, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 48698, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::MULTI-vertebrate_other' (id = 174)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (174, 48698, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_other/vertebrate_other*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 48698, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::MULTI-vertebrate_other' (id = 242)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (242, 48698, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_other/vertebrate_other*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 48698, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 48698, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 48698, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 48698, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 48698, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'pongo_abelii' (id = 9601)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 9601, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 9601, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 9601, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 9601, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 9601, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 9601, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 9601, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 9601, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::pongo_abelii' (id = 215)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (215, 9601, 'ftp://ftp.ncbi.nih.gov/genomes/Pongo_abelii/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 9601, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::pongo_abelii' (id = 282)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (282, 9601, 'ftp://ftp.ncbi.nih.gov/genomes/Pongo_abelii/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::pongo_abelii' (id = 347)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (347, 9601, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Pongo_abelii/Pab.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Pongo_abelii/Pab.data.gz', 'ftp://ftp.ncbi.nih.gov/repository/UniGene/*.LOG', now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 9601, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 9601, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 9601, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 9601, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 9601, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'populus_trichocarpa' (id = 3694)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 3694, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 3694, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 3694, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Gramene_Pathway::populus_trichocarpa' (id = 75)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (75, 3694, 'ftp://ftp.gramene.org/pub/gramene/ensemblgenomes/xrefs/pathways/poplarcyc.tab.gz', \N, now(), now(), 'GramenePathwayParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 3694, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 3694, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (167, 3694, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_mRNA_predicted::MULTI' (id = 229)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (229, 3694, '', \N, now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (235, 3694, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::populus_trichocarpa' (id = 348)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (348, 3694, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Populus_trichocarpa/Pth.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Populus_trichocarpa/Pth.data.gz', \N, now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 3694, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 3694, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 3694, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 3694, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'pristionchus_pacificus' (id = 54126)
#-------------------------------------------------------------------------------

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 54126, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 54126, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 54126, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 54126, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 54126, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI-invertebrate' (id = 369)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (369, 54126, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/uniprot_trembl_invertebrates.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI-invertebrate' (id = 374)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (374, 54126, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/uniprot_sprot_invertebrates.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'wormbase::ppacificus' (id = 447)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (447, 54126, 'ftp://ftp.ebi.ac.uk/pub/databases/wormbase/releases/RELEASE/species/p_pacificus/PRJNA12644/p_pacificus.PRJNA12644.RELEASE.xrefs.txt.gz', \N, now(), now(), 'WormbaseDirectParser');

#-------------------------------------------------------------------------------
# Data for species 'procavia_capensis' (id = 9813)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 9813, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 9813, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 9813, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 9813, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 9813, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 9813, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 9813, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 9813, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::MULTI-vertebrate_mammalian' (id = 173)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (173, 9813, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_mammalian/vertebrate_mammalian*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 9813, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::MULTI-vertebrate_mammalian' (id = 241)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (241, 9813, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_mammalian/vertebrate_mammalian*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 9813, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 9813, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 9813, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 9813, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 9813, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'pteropus_vampyrus' (id = 132908)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 132908, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 132908, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 132908, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 132908, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 132908, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 132908, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 132908, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 132908, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::MULTI-vertebrate_mammalian' (id = 173)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (173, 132908, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_mammalian/vertebrate_mammalian*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 132908, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::MULTI-vertebrate_mammalian' (id = 241)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (241, 132908, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_mammalian/vertebrate_mammalian*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 132908, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 132908, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 132908, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 132908, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 132908, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'puccinia_graministritici' (id = 418459)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 418459, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 418459, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 418459, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 418459, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 418459, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'PHIbase::MULTI' (id = 136)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (136, 418459, 'http://www.phi-base.org/PHI-export.php', \N, now(), now(), 'PHIbaseParser');

# Data from source 'RFAM::EG' (id = 152)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (152, 418459, 'script:object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 418459, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RefSeq_dna::MULTI-fungi' (id = 169)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (169, 418459, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-fungi' (id = 237)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (237, 418459, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 418459, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 418459, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 418459, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 418459, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 418459, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

# Data from source 'misc_EG::EG' (id = 427)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (427, 418459, 'script:host=>mysql-eg-pan-1,port=>4276,dbname=>EG_Xrefs,', \N, now(), now(), 'EG_DBParser');

#-------------------------------------------------------------------------------
# Data for species 'puccinia_triticina' (id = 208348)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 208348, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 208348, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 208348, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 208348, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 208348, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'PHIbase::MULTI' (id = 136)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (136, 208348, 'http://www.phi-base.org/PHI-export.php', \N, now(), now(), 'PHIbaseParser');

# Data from source 'RFAM::EG' (id = 152)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (152, 208348, 'script:object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 208348, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RefSeq_dna::MULTI-fungi' (id = 169)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (169, 208348, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-fungi' (id = 237)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (237, 208348, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 208348, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 208348, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 208348, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 208348, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 208348, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

# Data from source 'misc_EG::EG' (id = 427)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (427, 208348, 'script:host=>mysql-eg-pan-1,port=>4276,dbname=>EG_Xrefs,', \N, now(), now(), 'EG_DBParser');

#-------------------------------------------------------------------------------
# Data for species 'pythium_ultimum' (id = 65071)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 65071, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 65071, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 65071, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 65071, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 65071, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'PHIbase::MULTI' (id = 136)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (136, 65071, 'http://www.phi-base.org/PHI-export.php', \N, now(), now(), 'PHIbaseParser');

# Data from source 'RFAM::EG' (id = 152)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (152, 65071, 'script:object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 65071, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RefSeq_dna::MULTI-protozoa' (id = 172)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (172, 65071, 'ftp://ftp.ncbi.nih.gov/refseq/release/protozoa/protozoa*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-protozoa' (id = 240)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (240, 65071, 'ftp://ftp.ncbi.nih.gov/refseq/release/protozoa/protozoa*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 65071, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 65071, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 65071, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 65071, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 65071, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

# Data from source 'misc_EG::EG' (id = 427)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (427, 65071, 'script:host=>mysql-eg-pan-1,port=>4276,dbname=>EG_Xrefs,', \N, now(), now(), 'EG_DBParser');

#-------------------------------------------------------------------------------
# Data for species 'rattus_norvegicus' (id = 10116)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 10116, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 10116, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::rattus_norvegicus' (id = 63)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (63, 10116, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/RAT/gene_association.goa_rat.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 10116, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 10116, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 10116, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RGD::rattus_norvegicus' (id = 156)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (156, 10116, 'ftp://rgd.mcw.edu/pub/data_release/GENES_RAT.txt', \N, now(), now(), 'RGDParser');

# Data from source 'RGD::rattus_norvegicus#01' (id = 157)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (157, 10116, 'script:source=>RGD,chost=>genebuild3,cdbname=>db8_rattus_norvegicus_core_77_5,vhost=>genebuild6,vdbname=>db8_rat_vega_77,', \N, now(), now(), 'VegaOfficialNameParser');

# Data from source 'RGD::rattus_norvegicus#02' (id = 158)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (158, 10116, '', \N, now(), now(), 'done_in_RGDParser');

# Data from source 'RGD_curated_transcript::mus_musculus' (id = 159)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (159, 10116, 'script:source=>RGD,host=>ens-staging2,', \N, now(), now(), 'curated_transcriptParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 10116, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 10116, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::rattus_norvegicus' (id = 216)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (216, 10116, 'ftp://ftp.ncbi.nih.gov/refseq/R_norvegicus/mRNA_Prot/rat*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 10116, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::rattus_norvegicus' (id = 283)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (283, 10116, 'ftp://ftp.ncbi.nih.gov/refseq/R_norvegicus/mRNA_Prot/rat*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::rattus_norvegicus' (id = 349)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (349, 10116, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Rattus_norvegicus/Rn.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Rattus_norvegicus/Rn.data.gz', 'ftp://ftp.ncbi.nih.gov/repository/UniGene/*.LOG', now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 10116, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 10116, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 10116, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 10116, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 10116, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'rhodnius_prolixus' (id = 13249)
#-------------------------------------------------------------------------------

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 13249, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 13249, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::CORE' (id = 107)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (107, 13249, 'script:project=>ensemblgenomes,', \N, now(), now(), 'InterproFromCoreParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 13249, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 13249, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 13249, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 13249, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 13249, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 13249, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 13249, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'ncRNA_EG::EG' (id = 428)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (428, 13249, 'script:logic_name=>ncrna_eg,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

#-------------------------------------------------------------------------------
# Data for species 'saccharomyces_cerevisiae' (id = 4932)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 4932, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 4932, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::saccharomyces_cerevisiae' (id = 64)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (64, 4932, 'ftp://ftp.geneontology.org/pub/go/gene-associations/gene_association.sgd.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', \N, now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 4932, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 4932, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-fungi' (id = 169)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (169, 4932, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-fungi' (id = 237)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (237, 4932, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'SGD::saccharomyces_cerevisiae' (id = 294)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (294, 4932, 'http://downloads.yeastgenome.org/curation/chromosomal_feature/SGD_features.tab', \N, now(), now(), 'SGDParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 4932, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 4932, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 4932, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'sarcophilus_harrisii' (id = 9305)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 9305, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 9305, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 9305, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 9305, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 9305, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 9305, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 9305, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 9305, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::MULTI-vertebrate_mammalian' (id = 173)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (173, 9305, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_mammalian/vertebrate_mammalian*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 9305, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::MULTI-vertebrate_mammalian' (id = 241)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (241, 9305, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_mammalian/vertebrate_mammalian*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 9305, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 9305, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 9305, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 9305, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 9305, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'schistosoma_mansoni' (id = 6183)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 6183, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 6183, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 6183, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 6183, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 6183, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 6183, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 6183, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 6183, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 6183, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 6183, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 6183, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'misc_EG::EG' (id = 427)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (427, 6183, 'script:host=>mysql-eg-pan-1,port=>4276,dbname=>EG_Xrefs,', \N, now(), now(), 'EG_DBParser');

#-------------------------------------------------------------------------------
# Data for species 'schizosaccharomyces_pombe' (id = 4896)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 4896, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 4896, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 4896, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 4896, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'PomBase::schizosaccharomyces_pombe' (id = 149)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (149, 4896, 'http://www.ebi.ac.uk/~mcdowall/PomBaseFeatureDescriptions_For-v65.txt', \N, now(), now(), 'PomBaseParser');

# Data from source 'RefSeq_dna::MULTI-fungi' (id = 169)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (169, 4896, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-fungi' (id = 237)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (237, 4896, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 4896, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 4896, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 4896, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'selaginella_moellendorffii' (id = 88036)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 88036, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 88036, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 88036, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 88036, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 88036, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (167, 88036, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_mRNA_predicted::MULTI' (id = 229)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (229, 88036, '', \N, now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (235, 88036, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::selaginella_moellendorffii' (id = 350)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (350, 88036, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Selaginella_moellendorffii/Smo.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Selaginella_moellendorffii/Smo.data.gz', \N, now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 88036, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 88036, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 88036, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 88036, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'misc_EG::EG' (id = 427)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (427, 88036, 'script:host=>mysql-eg-pan-1,port=>4276,dbname=>EG_Xrefs,', \N, now(), now(), 'EG_DBParser');

#-------------------------------------------------------------------------------
# Data for species 'setaria_italica' (id = 4555)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 4555, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 4555, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 4555, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 4555, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 4555, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (167, 4555, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (235, 4555, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 4555, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 4555, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 4555, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 4555, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'solanum_lycopersicum' (id = 4081)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 4081, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 4081, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 4081, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Gramene_Pathway::solanum_lycopersicum' (id = 76)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (76, 4081, 'ftp://ftp.gramene.org/pub/gramene/ensemblgenomes/xrefs/pathways/lycocyc.tab.gz', \N, now(), now(), 'GramenePathwayParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 4081, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 4081, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (167, 4081, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (235, 4081, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 4081, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 4081, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 4081, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 4081, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'solanum_tuberosum' (id = 4113)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 4113, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 4113, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 4113, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Gramene_Pathway::solanum_tuberosum' (id = 77)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (77, 4113, 'ftp://ftp.gramene.org/pub/gramene/ensemblgenomes/xrefs/pathways/potatocyc.tab.gz', \N, now(), now(), 'GramenePathwayParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 4113, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 4113, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'PGSC_GENE::solanum_tuberosum' (id = 135)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (135, 4113, 'http://solanaceae.plantbiology.msu.edu/data/PGSC_DM_v3.4_gene_func.txt.zip', \N, now(), now(), 'PGSCParser');

# Data from source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (167, 4113, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (235, 4113, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::solanum_tuberosum' (id = 352)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (352, 4113, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Solanum_tuberosum/Stu.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Solanum_tuberosum/Stu.data.gz', \N, now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 4113, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 4113, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 4113, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 4113, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'ncRNA_EG::EG' (id = 428)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (428, 4113, 'script:logic_name=>ncrna_eg,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

#-------------------------------------------------------------------------------
# Data for species 'solenopsis_invicta' (id = 13686)
#-------------------------------------------------------------------------------

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 13686, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 13686, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::CORE' (id = 107)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (107, 13686, 'script:project=>ensemblgenomes,', \N, now(), now(), 'InterproFromCoreParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 13686, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 13686, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 13686, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 13686, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 13686, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 13686, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 13686, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'ncRNA_EG::EG' (id = 428)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (428, 13686, 'script:logic_name=>ncrna_eg,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

#-------------------------------------------------------------------------------
# Data for species 'sorex_araneus' (id = 42254)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 42254, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 42254, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 42254, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 42254, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 42254, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 42254, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 42254, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 42254, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::sorex_araneus' (id = 217)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (217, 42254, 'ftp://ftp.ncbi.nih.gov/genomes/Sorex_araneus/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 42254, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::sorex_araneus' (id = 284)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (284, 42254, 'ftp://ftp.ncbi.nih.gov/genomes/Sorex_araneus/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 42254, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 42254, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 42254, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 42254, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 42254, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'sorghum_bicolor' (id = 4558)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 4558, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 4558, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 4558, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Gramene_Pathway::sorghum_bicolor' (id = 78)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (78, 4558, 'ftp://ftp.gramene.org/pub/gramene/ensemblgenomes/xrefs/pathways/sorghumcyc.tab.gz', \N, now(), now(), 'GramenePathwayParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 4558, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 4558, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (167, 4558, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_mRNA_predicted::MULTI' (id = 229)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (229, 4558, '', \N, now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (235, 4558, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::sorghum_bicolor' (id = 353)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (353, 4558, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Sorghum_bicolor/Sbi.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Sorghum_bicolor/Sbi.data.gz', \N, now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 4558, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 4558, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 4558, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 4558, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'strigamia_maritima' (id = 126957)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 126957, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 126957, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 126957, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 126957, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 126957, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::EG' (id = 152)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (152, 126957, 'script:object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 126957, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 126957, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 126957, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 126957, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 126957, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 126957, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 126957, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'strongylocentrotus_purpuratus' (id = 7668)
#-------------------------------------------------------------------------------

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 7668, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 7668, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::CORE' (id = 107)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (107, 7668, 'script:project=>ensemblgenomes,', \N, now(), now(), 'InterproFromCoreParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 7668, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::strongylocentrotus_purpuratus' (id = 218)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (218, 7668, 'ftp://ftp.ncbi.nih.gov/genomes/Strongylocentrotus_purpuratus/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_peptide::strongylocentrotus_purpuratus' (id = 285)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (285, 7668, 'ftp://ftp.ncbi.nih.gov/genomes/Strongylocentrotus_purpuratus/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::strongylocentrotus_purpuratus' (id = 354)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (354, 7668, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Strongylocentrotus_purpuratus/Spu.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Strongylocentrotus_purpuratus/Spu.data.gz', \N, now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 7668, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 7668, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 7668, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 7668, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'misc_EG::EG' (id = 427)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (427, 7668, 'script:host=>mysql-eg-pan-1,port=>4276,dbname=>EG_Xrefs,', \N, now(), now(), 'EG_DBParser');

# Data from source 'ncRNA_EG::EG' (id = 428)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (428, 7668, 'script:logic_name=>ncrna_eg,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

#-------------------------------------------------------------------------------
# Data for species 'strongyloides_ratti' (id = 34506)
#-------------------------------------------------------------------------------

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 34506, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 34506, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 34506, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 34506, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 34506, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI-invertebrate' (id = 369)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (369, 34506, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/uniprot_trembl_invertebrates.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI-invertebrate' (id = 374)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (374, 34506, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/uniprot_sprot_invertebrates.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/current_release/knowledgebase/taxonomic_divisions/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'wormbase::sratti' (id = 448)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (448, 34506, 'ftp://ftp.ebi.ac.uk/pub/databases/wormbase/releases/RELEASE/species/s_ratti/PRJEB125/s_ratti.PRJEB125.RELEASE.xrefs.txt.gz', \N, now(), now(), 'WormbaseDirectParser');

#-------------------------------------------------------------------------------
# Data for species 'sus_scrofa' (id = 9823)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 9823, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 9823, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 9823, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 9823, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 9823, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'PIGGY::sus_scrofa' (id = 138)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (138, 9823, 'script:host=>ens-staging2,source=>PIGGY,', \N, now(), now(), 'VegaOfficialNameParser');

# Data from source 'PIGGY_curated_transcript::sus_scrofa' (id = 139)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (139, 9823, 'script:source=>PIGGY,host=>ens-staging2,', \N, now(), now(), 'curated_transcriptParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 9823, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 9823, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 9823, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::sus_scrofa' (id = 219)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (219, 9823, 'ftp://ftp.ncbi.nih.gov/genomes/Sus_scrofa/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 9823, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::sus_scrofa' (id = 286)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (286, 9823, 'ftp://ftp.ncbi.nih.gov/genomes/Sus_scrofa/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::sus_scrofa' (id = 355)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (355, 9823, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Sus_scrofa/Ssc.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Sus_scrofa/Ssc.data.gz', \N, now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 9823, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 9823, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 9823, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 9823, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 9823, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'taeniopygia_guttata' (id = 59729)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 59729, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 59729, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 59729, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 59729, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 59729, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 59729, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 59729, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 59729, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::taeniopygia_guttata' (id = 220)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (220, 59729, 'ftp://ftp.ncbi.nih.gov/genomes/Taeniopygia_guttata/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 59729, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::taeniopygia_guttata' (id = 287)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (287, 59729, 'ftp://ftp.ncbi.nih.gov/genomes/Taeniopygia_guttata/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::taeniopygia_guttata' (id = 356)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (356, 59729, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Taeniopygia_guttata/Tgu.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Taeniopygia_guttata/Tgu.data.gz', 'ftp://ftp.ncbi.nih.gov/repository/UniGene/*.LOG', now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 59729, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 59729, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 59729, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 59729, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 59729, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'takifugu_rubripes' (id = 31033)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 31033, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 31033, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 31033, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 31033, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 31033, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 31033, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 31033, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 31033, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::takifugu_rubripes' (id = 221)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (221, 31033, 'ftp://ftp.ncbi.nih.gov/genomes/Takifugu_rubripes/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 31033, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::takifugu_rubripes' (id = 288)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (288, 31033, 'ftp://ftp.ncbi.nih.gov/genomes/Takifugu_rubripes/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::takifugu_rubripes' (id = 357)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (357, 31033, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Takifugu_rubripes/Tru.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Takifugu_rubripes/Tru.data.gz', 'ftp://ftp.ncbi.nih.gov/repository/UniGene/*.LOG', now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 31033, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 31033, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 31033, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 31033, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 31033, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'tarsius_syrichta' (id = 9478)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 9478, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 9478, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 9478, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 9478, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 9478, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 9478, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 9478, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 9478, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::MULTI-vertebrate_mammalian' (id = 173)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (173, 9478, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_mammalian/vertebrate_mammalian*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 9478, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::MULTI-vertebrate_mammalian' (id = 241)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (241, 9478, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_mammalian/vertebrate_mammalian*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 9478, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 9478, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 9478, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 9478, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 9478, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'tetranychus_urticae' (id = 32264)
#-------------------------------------------------------------------------------

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 32264, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 32264, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::CORE' (id = 107)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (107, 32264, 'script:project=>ensemblgenomes,', \N, now(), now(), 'InterproFromCoreParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 32264, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 32264, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 32264, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 32264, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 32264, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 32264, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 32264, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'ncRNA_EG::EG' (id = 428)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (428, 32264, 'script:logic_name=>ncrna_eg,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

#-------------------------------------------------------------------------------
# Data for species 'tetraodon_nigroviridis' (id = 99883)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 99883, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 99883, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 99883, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 99883, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 99883, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 99883, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 99883, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 99883, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::MULTI-vertebrate_other' (id = 174)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (174, 99883, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_other/vertebrate_other*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 99883, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::MULTI-vertebrate_other' (id = 242)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (242, 99883, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_other/vertebrate_other*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 99883, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 99883, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 99883, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 99883, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 99883, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'thalassiosira_pseudonana' (id = 35128)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 35128, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 35128, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 35128, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 35128, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 35128, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::EG' (id = 152)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (152, 35128, 'script:object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 35128, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RefSeq_dna::MULTI-complete' (id = 168)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (168, 35128, 'ftp://ftp.ncbi.nih.gov/refseq/release/complete/complete*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-complete' (id = 236)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (236, 35128, 'ftp://ftp.ncbi.nih.gov/refseq/release/complete/complete*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 35128, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 35128, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 35128, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 35128, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 35128, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

# Data from source 'misc_EG::EG' (id = 427)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (427, 35128, 'script:host=>mysql-eg-pan-1,port=>4276,dbname=>EG_Xrefs,', \N, now(), now(), 'EG_DBParser');

#-------------------------------------------------------------------------------
# Data for species 'tribolium_castaneum' (id = 7070)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 7070, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 7070, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 7070, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 7070, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 7070, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::tribolium_castaneum' (id = 222)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (222, 7070, 'ftp://ftp.ncbi.nih.gov/genomes/Tribolium_castaneum/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_peptide::tribolium_castaneum' (id = 289)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (289, 7070, 'ftp://ftp.ncbi.nih.gov/genomes/Tribolium_castaneum/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::tribolium_castaneum' (id = 358)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (358, 7070, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Tribolium_castaneum/Tca.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Tribolium_castaneum/Tca.data.gz', \N, now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 7070, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 7070, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 7070, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 7070, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'trichinella_spiralis' (id = 6334)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 6334, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 6334, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 6334, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 6334, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-complete' (id = 168)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (168, 6334, 'ftp://ftp.ncbi.nih.gov/refseq/release/complete/complete*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-complete' (id = 236)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (236, 6334, 'ftp://ftp.ncbi.nih.gov/refseq/release/complete/complete*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 6334, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 6334, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 6334, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'trichoplax_adhaerens' (id = 10228)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 10228, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 10228, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 10228, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 10228, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 10228, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::EG' (id = 152)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (152, 10228, 'script:object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 10228, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RefSeq_dna::MULTI-invertebrate' (id = 170)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (170, 10228, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-invertebrate' (id = 238)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (238, 10228, 'ftp://ftp.ncbi.nih.gov/refseq/release/invertebrate/invertebrate*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 10228, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 10228, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 10228, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 10228, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 10228, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

# Data from source 'misc_EG::EG' (id = 427)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (427, 10228, 'script:host=>mysql-eg-pan-1,port=>4276,dbname=>EG_Xrefs,', \N, now(), now(), 'EG_DBParser');

#-------------------------------------------------------------------------------
# Data for species 'triticum_aestivum' (id = 4565)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 4565, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 4565, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 4565, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 4565, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 4565, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (167, 4565, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (235, 4565, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 4565, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 4565, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 4565, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 4565, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'ncRNA_EG::EG' (id = 428)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (428, 4565, 'script:logic_name=>ncrna_eg,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

#-------------------------------------------------------------------------------
# Data for species 'triticum_urartu' (id = 4572)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 4572, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 4572, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 4572, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 4572, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 4572, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (167, 4572, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (235, 4572, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 4572, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 4572, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 4572, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 4572, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'ncRNA_EG::EG' (id = 428)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (428, 4572, 'script:logic_name=>ncrna_eg,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

#-------------------------------------------------------------------------------
# Data for species 'tuber_melanosporum' (id = 656061)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 656061, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 656061, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 656061, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 656061, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 656061, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-fungi' (id = 169)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (169, 656061, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-fungi' (id = 237)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (237, 656061, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 656061, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 656061, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 656061, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'tupaia_belangeri' (id = 37347)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 37347, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 37347, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 37347, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 37347, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 37347, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 37347, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 37347, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 37347, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::MULTI-vertebrate_mammalian' (id = 173)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (173, 37347, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_mammalian/vertebrate_mammalian*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 37347, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::MULTI-vertebrate_mammalian' (id = 241)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (241, 37347, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_mammalian/vertebrate_mammalian*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 37347, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 37347, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 37347, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 37347, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 37347, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'tursiops_truncatus' (id = 9739)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 9739, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 9739, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 9739, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 9739, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 9739, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 9739, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 9739, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 9739, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::tursiops_truncatus' (id = 223)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (223, 9739, 'ftp://ftp.ncbi.nih.gov/genomes/Tursiops_truncatus/RNA/rna.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 9739, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::tursiops_truncatus' (id = 290)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (290, 9739, 'ftp://ftp.ncbi.nih.gov/genomes/Tursiops_truncatus/protein/protein.gbk.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 9739, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 9739, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 9739, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 9739, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 9739, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'ustilago_maydis' (id = 237631)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 237631, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 237631, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 237631, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 237631, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 237631, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'PHIbase::MULTI' (id = 136)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (136, 237631, 'http://www.phi-base.org/PHI-export.php', \N, now(), now(), 'PHIbaseParser');

# Data from source 'RFAM::EG' (id = 152)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (152, 237631, 'script:object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RNAMMER::MULTI' (id = 162)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (162, 237631, 'script:biotype=>rRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'RefSeq_dna::MULTI-fungi' (id = 169)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (169, 237631, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-fungi' (id = 237)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (237, 237631, 'ftp://ftp.ncbi.nih.gov/refseq/release/fungi/fungi*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'TRNASCAN_SE::MULTI' (id = 302)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (302, 237631, 'script:biotype=>tRNA,object_type=>gene,project=>ensemblgenomes,copy_description_from_object=>1,', \N, now(), now(), 'CoreXrefParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 237631, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 237631, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 237631, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 237631, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

# Data from source 'misc_EG::EG' (id = 427)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (427, 237631, 'script:host=>mysql-eg-pan-1,port=>4276,dbname=>EG_Xrefs,', \N, now(), now(), 'EG_DBParser');

#-------------------------------------------------------------------------------
# Data for species 'vicugna_pacos' (id = 30538)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 30538, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 30538, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 30538, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 30538, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 30538, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 30538, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 30538, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 30538, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::MULTI-vertebrate_mammalian' (id = 173)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (173, 30538, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_mammalian/vertebrate_mammalian*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 30538, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::MULTI-vertebrate_mammalian' (id = 241)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (241, 30538, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_mammalian/vertebrate_mammalian*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 30538, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 30538, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 30538, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 30538, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 30538, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'vitis_vinifera' (id = 29760)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 29760, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 29760, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 29760, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 29760, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 29760, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (167, 29760, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (235, 29760, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 29760, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 29760, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 29760, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 29760, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

#-------------------------------------------------------------------------------
# Data for species 'xenopus_tropicalis' (id = 8364)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 8364, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 8364, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 8364, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 8364, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 8364, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 8364, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 8364, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 8364, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::MULTI-vertebrate_other' (id = 174)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (174, 8364, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_other/vertebrate_other*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 8364, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::MULTI-vertebrate_other' (id = 242)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (242, 8364, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_other/vertebrate_other*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::xenopus_tropicalis' (id = 360)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (360, 8364, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Xenopus_tropicalis/Str.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Xenopus_tropicalis/Str.data.gz', 'ftp://ftp.ncbi.nih.gov/repository/UniGene/*.LOG', now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 8364, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'UniProt::protein_id' (id = 364)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (364, 8364, '', \N, now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 8364, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 8364, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot::EMBL' (id = 377)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (377, 8364, '', \N, now(), now(), 'UniProtParser');

# Data from source 'Xenopus_Jamboree::xenopus_tropicalis' (id = 396)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (396, 8364, 'ftp://ftp.xenbase.org/pub/GenePageReports/GenePageEnsemblModelMapping.txt', \N, now(), now(), 'XenopusJamboreeParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 8364, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 8364, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'xiphophorus_maculatus' (id = 8083)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::MULTI' (id = 5)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (5, 8083, 'script:project=>ensembl,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 8083, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 8083, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 8083, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 8083, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RFAM::MULTI' (id = 153)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (153, 8083, 'script:wget=>ftp://ftp.sanger.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz,', \N, now(), now(), 'RFAMParser');

# Data from source 'RNACentral::MULTI' (id = 161)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (161, 8083, 'ftp://ftp.ebi.ac.uk/pub/databases/RNAcentral/current_release/md5/md5.tsv.gz', \N, now(), now(), 'ChecksumParser');

# Data from source 'Reactome::MULTI' (id = 163)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (163, 8083, 'http://www.reactome.org/download/current/Ensembl2Reactome_All_Levels.txt http://www.reactome.org/download/current/UniProt2Reactome_All_Levels.txt', 'http://www.reactome.org/ReactomeRESTfulAPI/RESTfulWS/version', now(), now(), 'ReactomeParser');

# Data from source 'RefSeq_dna::MULTI-vertebrate_other' (id = 174)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (174, 8083, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_other/vertebrate_other*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_import::otherfeatures' (id = 224)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (224, 8083, 'script:project=>ensembl,', \N, now(), now(), 'RefSeqCoordinateParser');

# Data from source 'RefSeq_peptide::MULTI-vertebrate_other' (id = 242)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (242, 8083, 'ftp://ftp.ncbi.nih.gov/refseq/release/vertebrate_other/vertebrate_other*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 8083, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'UniProt::protein_id' (id = 364)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (364, 8083, '', \N, now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 8083, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 8083, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot::EMBL' (id = 377)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (377, 8083, '', \N, now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::MULTI' (id = 421)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (421, 8083, 'script:host=>ens-livemirror,dbname=>ensembl_ontology_83,', \N, now(), now(), 'GOSlimParser');

# Data from source 'miRBase::MULTI' (id = 422)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (422, 8083, 'ftp://mirbase.org/pub/mirbase/CURRENT/miRNA.dat.gz', \N, now(), now(), 'miRBaseParser');

#-------------------------------------------------------------------------------
# Data for species 'zea_mays' (id = 4577)
#-------------------------------------------------------------------------------

# Data from source 'ArrayExpress::EG' (id = 4)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (4, 4577, 'script:project=>ensemblgenomes,', \N, now(), now(), 'ArrayExpressParser');

# Data from source 'EntrezGene::MULTI' (id = 44)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (44, 4577, 'ftp://ftp.ncbi.nlm.nih.gov/gene/DATA/gene_info.gz', \N, now(), now(), 'EntrezGeneParser');

# Data from source 'GO::MULTI' (id = 56)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (56, 4577, 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/UNIPROT/gene_association.goa_uniprot.gz http://archive.geneontology.org/latest-termdb/go_daily-termdb.obo-xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/GO/goa/current_release_numbers.txt', now(), now(), 'GOParser');

# Data from source 'Gramene_Pathway::zea_mays' (id = 79)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (79, 4577, 'ftp://ftp.gramene.org/pub/gramene/ensemblgenomes/xrefs/pathways/maizecyc.tab.gz', \N, now(), now(), 'GramenePathwayParser');

# Data from source 'Interpro::MULTI' (id = 108)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (108, 4577, 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/interpro.xml.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/interpro/release_notes.txt', now(), now(), 'InterproParser');

# Data from source 'InterproGO::MULTI' (id = 109)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (109, 4577, 'http://www.geneontology.org/external2go/interpro2go', \N, now(), now(), 'InterproGoParser');

# Data from source 'RefSeq_dna::MULTI-Plants' (id = 167)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (167, 4577, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.rna.fna.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_mRNA_predicted::MULTI' (id = 229)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (229, 4577, '', \N, now(), now(), 'RefSeqParser');

# Data from source 'RefSeq_peptide::MULTI-Plants' (id = 235)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (235, 4577, 'ftp://ftp.ncbi.nih.gov/refseq/release/plant/plant*.protein.gpff.gz', 'ftp://ftp.ncbi.nih.gov/refseq/release/release-notes/RefSeq-release*.txt', now(), now(), 'RefSeqGPFFParser');

# Data from source 'UniGene::zea_mays' (id = 361)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (361, 4577, 'ftp://ftp.ncbi.nih.gov/repository/UniGene/Zea_mays/Zm.seq.uniq.gz ftp://ftp.ncbi.nih.gov/repository/UniGene/Zea_mays/Zm.data.gz', \N, now(), now(), 'UniGeneParser');

# Data from source 'UniParc::MULTI' (id = 362)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (362, 4577, 'ftp://ftp.ebi.ac.uk/pub/contrib/uniparc/upidump.lis', \N, now(), now(), 'ChecksumParser');

# Data from source 'Uniprot/SPTREMBL::MULTI' (id = 367)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (367, 4577, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_trembl.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'Uniprot/SWISSPROT::MULTI' (id = 373)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (373, 4577, 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/uniprot_sprot.dat.gz', 'ftp://ftp.ebi.ac.uk/pub/databases/uniprot/knowledgebase/reldate.txt', now(), now(), 'UniProtParser');

# Data from source 'goslim_goa::EG' (id = 420)
INSERT INTO source_url (source_id, species_id, url, release_url, file_modified_date, upload_date, parser)
VALUES (420, 4577, 'script:host=>mysql-eg-staging-1.ebi.ac.uk,port=>4160,dbname=>ensemblgenomes_ontology_21_74,', \N, now(), now(), 'GOSlimParser');

# FINISHED SUCCESSFULLY
